\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop}{}\section{negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop Class Reference}
\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop}\index{negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop@{negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop}}
Inheritance diagram for negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop\+:\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[height=2.000000cm]{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop}
\end{center}
\end{figure}
\subsection*{Public Member Functions}
\begin{DoxyCompactItemize}
\item 
def {\bfseries \+\_\+\+\_\+init\+\_\+\+\_\+} (self, s\+\_\+output\+\_\+files\+\_\+prefix)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a8f42fc0afc9024177960bb6287e45493}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a8f42fc0afc9024177960bb6287e45493}

\item 
def {\bfseries open\+Out} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a7cea46f9212d05e98ce14b7647781590}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a7cea46f9212d05e98ce14b7647781590}

\item 
def {\bfseries open\+Err} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ae3f0c28a1a9afbef82fddd8d26d3ea3b}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ae3f0c28a1a9afbef82fddd8d26d3ea3b}

\item 
def {\bfseries open\+Mega\+DB} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_afac367cb22b4498a2c5170173de1bdc1}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_afac367cb22b4498a2c5170173de1bdc1}

\item 
def {\bfseries open\+Genepop} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a0c69fcc0a83881fe8ea9956979de342a}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a0c69fcc0a83881fe8ea9956979de342a}

\item 
def {\bfseries copy\+Me} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a5ab021b0ff9d452891a52313b9402fbf}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a5ab021b0ff9d452891a52313b9402fbf}

\item 
def {\bfseries open\+Conf} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_abd8a699e5a3adf2b746bdb778602fe60}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_abd8a699e5a3adf2b746bdb778602fe60}

\item 
def \hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a3046e1eac1fbd82acf47c0f7c0d06b70}{bz2\+Compress\+All\+Files} (self, ls\+\_\+files\+\_\+to\+\_\+skip=\mbox{[}$\,$\mbox{]})
\item 
def \hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a99bc161de61efea5751a230b9b4c9950}{write\+Genepop\+File\+Header\+And\+Loci\+List} (self, o\+\_\+genepopfile, i\+\_\+num\+\_\+loci, f\+\_\+nbne\+\_\+ratio=None, b\+\_\+do\+\_\+compress=False)
\item 
def \hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a97a29bbf73ef3b22a5de110d473611b6}{gen2\+Genepop} (self, idx\+\_\+allele\+\_\+start, idx\+\_\+allele\+\_\+stop, b\+\_\+do\+\_\+compress=True, b\+\_\+pop\+\_\+per\+\_\+gen=False, f\+\_\+nbne\+\_\+ratio=None)
\item 
def {\bfseries remove\+Output\+File\+By\+Ext} (self, s\+\_\+extension)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a27af4e615fcb17d2343dc27f614c1359}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a27af4e615fcb17d2343dc27f614c1359}

\item 
def {\bfseries basename} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_af67ba734fbe5d4240a8cef38e63586d1}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_af67ba734fbe5d4240a8cef38e63586d1}

\item 
def {\bfseries basename} (self, s\+\_\+basename)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_aafca25668d51a1beec732306b7973884}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_aafca25668d51a1beec732306b7973884}

\item 
def {\bfseries confname} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a67795b5f8245e4e921615ac4cc79a5be}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a67795b5f8245e4e921615ac4cc79a5be}

\item 
def {\bfseries simname} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ac3e82333b61e2c3e9545c9ca8e9cf64c}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ac3e82333b61e2c3e9545c9ca8e9cf64c}

\item 
def {\bfseries genname} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a43b4446563d323091900f8b6bcfd6044}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a43b4446563d323091900f8b6bcfd6044}

\item 
def {\bfseries dbname} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a6159d8c414cea2420a3de80ab00222bc}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a6159d8c414cea2420a3de80ab00222bc}

\item 
def {\bfseries genepopname} (self)\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a19133bb8e7bcd272da350abbb5427fd4}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a19133bb8e7bcd272da350abbb5427fd4}

\end{DoxyCompactItemize}
\subsection*{Public Attributes}
\begin{DoxyCompactItemize}
\item 
{\bfseries out}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a4198008f7f7f029a29f8441ea7619170}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a4198008f7f7f029a29f8441ea7619170}

\item 
{\bfseries err}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a5eeccbf099cd8be3e8ab5906565d7b79}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a5eeccbf099cd8be3e8ab5906565d7b79}

\item 
{\bfseries mega\+DB}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a76379971c8aa1a3b175f171821b0b8f1}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a76379971c8aa1a3b175f171821b0b8f1}

\item 
{\bfseries conf}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ac6e385cca72ff0073251b511eea1103f}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ac6e385cca72ff0073251b511eea1103f}

\item 
{\bfseries genepop}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a0f1141b3cdaea508a0302a61dacdabf6}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a0f1141b3cdaea508a0302a61dacdabf6}

\end{DoxyCompactItemize}
\subsection*{Static Public Attributes}
\begin{DoxyCompactItemize}
\item 
dictionary {\bfseries D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS}
\item 
string {\bfseries C\+O\+M\+P\+R\+E\+S\+S\+I\+O\+N\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+ON} = \char`\"{}bz2\char`\"{}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_aaf295d98adb01d86781790b5a57c5348}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_aaf295d98adb01d86781790b5a57c5348}

\end{DoxyCompactItemize}


\subsection{Detailed Description}
\begin{DoxyVerb}Object meant to fetch parameter values and prepare them for 
use in a simuPop simulation.  

Object to be passed to a PGOpSimuPop object, which is, in turn,
is passed to a PGGuiSimuPop object, or created in pgutilities,
to run parallel replicates.  This object writes and manages the
output of the simupop simulation. 

2017_05_16.  Adding to the output file extensions the two files,
listing age-counts and the PWOP Nb estimates file extensions.
These are written by PGOpSimuPop instances, on the fist replicate
only, initially to have been for testing only, but have become
part of the standard output.  Their extensions are included here
so that, if pgutilities def, remove_simulation_replicate_output_files
is called on a cancelled simulation run, these files will also
be removed.


2017_08_08.  Adding to output file etendsions a file created when
the new PGOpSimuPop.OUTPUT_GENEPOP_HET_FILTERED output mode is used.
\end{DoxyVerb}
 

Definition at line 51 of file pgoutputsimupop.\+py.



\subsection{Member Function Documentation}
\index{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}!bz2\+Compress\+All\+Files@{bz2\+Compress\+All\+Files}}
\index{bz2\+Compress\+All\+Files@{bz2\+Compress\+All\+Files}!negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}}
\subsubsection[{\texorpdfstring{bz2\+Compress\+All\+Files(self, ls\+\_\+files\+\_\+to\+\_\+skip=[])}{bz2CompressAllFiles(self, ls_files_to_skip=[])}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop.\+bz2\+Compress\+All\+Files (
\begin{DoxyParamCaption}
\item[{}]{self, }
\item[{}]{ls\+\_\+files\+\_\+to\+\_\+skip = {\ttfamily \mbox{[}\mbox{]}}}
\end{DoxyParamCaption}
)}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a3046e1eac1fbd82acf47c0f7c0d06b70}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a3046e1eac1fbd82acf47c0f7c0d06b70}
\begin{DoxyVerb}used code and advice in, 
http://stackoverflow.com/questions/9518705/big-file-compression-with-python

Note: checked the shutil documentation at https://docs.python.org/2/library/shutil.html
which warns of loss of meta file info (owner/group ACLs) when using shutil.copy() or shutil.copy2().  Unclear
whether this applies to the copyfileobj, though my few tests showd all of these were retained.
\end{DoxyVerb}
 

Definition at line 242 of file pgoutputsimupop.\+py.


\begin{DoxyCode}
242     \textcolor{keyword}{def }\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a3046e1eac1fbd82acf47c0f7c0d06b70}{bz2CompressAllFiles}(self, ls\_files\_to\_skip=[] ):
243         \textcolor{stringliteral}{'''}
244 \textcolor{stringliteral}{        used code and advice in, }
245 \textcolor{stringliteral}{        http://stackoverflow.com/questions/9518705/big-file-compression-with-python}
246 \textcolor{stringliteral}{}
247 \textcolor{stringliteral}{        Note: checked the shutil documentation at https://docs.python.org/2/library/shutil.html}
248 \textcolor{stringliteral}{        which warns of loss of meta file info (owner/group ACLs) when using shutil.copy() or
       shutil.copy2().  Unclear}
249 \textcolor{stringliteral}{        whether this applies to the copyfileobj, though my few tests showd all of these were retained.}
250 \textcolor{stringliteral}{        '''}
251         \textcolor{keywordflow}{for} s\_myfile \textcolor{keywordflow}{in} [ self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a50f64be6bb990a0ac6898a44b4bf6fbc}{\_\_outname}, self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a26363a7bbc4ae53b86585e32f0360e2a}{\_\_errname}, self.
      \hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_aa89de70104807ce7e26bb17310f74b97}{\_\_megadbname}, self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_af5c9a72247505d2c2bb3d7400cbc09b3}{\_\_confname}, self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a05d91a106b2529b6cbae7777717a03b6}{\_\_genepopname} ]:
252             
253             \textcolor{keywordflow}{if} s\_myfile \textcolor{keywordflow}{in} ls\_files\_to\_skip:
254                 \textcolor{keywordflow}{pass}
255             \textcolor{keywordflow}{elif} \textcolor{keywordflow}{not} self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a1fdfb988619bd4a9d377a1e20f52839c}{\_\_file\_exists}( s\_myfile ):
256                 self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_abf6aab4f7982c2b0f9e4544343bf241b}{\_\_raise\_file\_not\_found\_error}( s\_myfile, \textcolor{stringliteral}{"compress file
       with bz2"}  )
257             \textcolor{keywordflow}{else}:
258                 with open( s\_myfile, \textcolor{stringliteral}{'rb'} ) \textcolor{keyword}{as} o\_input:
259                     with bz2.BZ2File( s\_myfile + \textcolor{stringliteral}{'.bz2'}, \textcolor{stringliteral}{'wb'}, compresslevel=9 ) \textcolor{keyword}{as} o\_output:
260                         \textcolor{comment}{#try/except overkill, but want logic to show}
261                         \textcolor{comment}{#we don't remove the input file}
262                         \textcolor{comment}{#if the copy op fails:}
263                         \textcolor{keywordflow}{try}: 
264                             shutil.copyfileobj( o\_input, o\_output )
265                         \textcolor{keywordflow}{except} Exception \textcolor{keyword}{as}  e:
266                             \textcolor{keywordflow}{raise} e
267                         \textcolor{comment}{#end try except}
268 
269                         \textcolor{comment}{#required by windows only, else error can't remove:}
270                         o\_input.close()
271                         os.remove( s\_myfile )   
272                     \textcolor{comment}{#end with bzfile for writing}
273                 \textcolor{comment}{#end with current file for reading}
274             \textcolor{comment}{#end if file absent, else exists}
275         \textcolor{comment}{#end for each file}
\end{DoxyCode}
\index{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}!gen2\+Genepop@{gen2\+Genepop}}
\index{gen2\+Genepop@{gen2\+Genepop}!negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}}
\subsubsection[{\texorpdfstring{gen2\+Genepop(self, idx\+\_\+allele\+\_\+start, idx\+\_\+allele\+\_\+stop, b\+\_\+do\+\_\+compress=\+True, b\+\_\+pop\+\_\+per\+\_\+gen=\+False, f\+\_\+nbne\+\_\+ratio=\+None)}{gen2Genepop(self, idx_allele_start, idx_allele_stop, b_do_compress=True, b_pop_per_gen=False, f_nbne_ratio=None)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop.\+gen2\+Genepop (
\begin{DoxyParamCaption}
\item[{}]{self, }
\item[{}]{idx\+\_\+allele\+\_\+start, }
\item[{}]{idx\+\_\+allele\+\_\+stop, }
\item[{}]{b\+\_\+do\+\_\+compress = {\ttfamily True}, }
\item[{}]{b\+\_\+pop\+\_\+per\+\_\+gen = {\ttfamily False}, }
\item[{}]{f\+\_\+nbne\+\_\+ratio = {\ttfamily None}}
\end{DoxyParamCaption}
)}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a97a29bbf73ef3b22a5de110d473611b6}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a97a29bbf73ef3b22a5de110d473611b6}
\begin{DoxyVerb}reads the *.gen file from the simuPop output
and produces a genepop file based on the gen file

param idx_allele_start: int, one-based index giving the
item (loci) number of the first allele entry in the line of gen-file input
Example, if the *gen file has 10 microsates and 40 SNPs, then its first
10 allele entries (that follow the indiv number and the gneration number)
will be the microsats, and the final 40 allele entries will be the SNPs
If we wanted only the Microsats to be included in the gen file, then
this param would be "1", and the idx_allele_stop value would be 10. 
If we want both msats and snps, we'd give 1 and 50 for these params. 
If we want only SNPs, wed enter 11 for start allele and 50 for stop.

param idx_allele_stop: int, one-based index giving the
item (loci) number of the last allele entry in the line of gen-file input

param optional b_do_compress, if true, will bzip2 the output genepop file

param optional b_pop_per_gen, if true, will insert a "Pop" line between the
generations, as given by the 2nd field in the gen file

2017_02_12.  In implementing a bias adjustment in the NeEstimator's ldne-based
Nb estimates, calculated in pgdriveneestimator.py, We add to this def an arg 
"f_nbne_ratio", defalut None.  If arg is non-None, we append to the header line 
text, "nbne=str(f_nbne_ratio)". When the GUI interface for the Nb estimation reads
in genepop files, it will check for they key=value string and then, if present,
pass the value on to the calls that eventually invoke the pgdriveneestimator.
\end{DoxyVerb}
 

Definition at line 327 of file pgoutputsimupop.\+py.


\begin{DoxyCode}
327             f\_nbne\_ratio=\textcolor{keywordtype}{None} ):
328 
329         \textcolor{stringliteral}{'''}
330 \textcolor{stringliteral}{        reads the *.gen file from the simuPop output}
331 \textcolor{stringliteral}{        and produces a genepop file based on the gen file}
332 \textcolor{stringliteral}{}
333 \textcolor{stringliteral}{        param idx\_allele\_start: int, one-based index giving the}
334 \textcolor{stringliteral}{        item (loci) number of the first allele entry in the line of gen-file input}
335 \textcolor{stringliteral}{        Example, if the *gen file has 10 microsates and 40 SNPs, then its first}
336 \textcolor{stringliteral}{        10 allele entries (that follow the indiv number and the gneration number)}
337 \textcolor{stringliteral}{        will be the microsats, and the final 40 allele entries will be the SNPs}
338 \textcolor{stringliteral}{        If we wanted only the Microsats to be included in the gen file, then}
339 \textcolor{stringliteral}{        this param would be "1", and the idx\_allele\_stop value would be 10. }
340 \textcolor{stringliteral}{        If we want both msats and snps, we'd give 1 and 50 for these params. }
341 \textcolor{stringliteral}{        If we want only SNPs, wed enter 11 for start allele and 50 for stop.}
342 \textcolor{stringliteral}{}
343 \textcolor{stringliteral}{        param idx\_allele\_stop: int, one-based index giving the}
344 \textcolor{stringliteral}{        item (loci) number of the last allele entry in the line of gen-file input}
345 \textcolor{stringliteral}{}
346 \textcolor{stringliteral}{        param optional b\_do\_compress, if true, will bzip2 the output genepop file}
347 \textcolor{stringliteral}{}
348 \textcolor{stringliteral}{        param optional b\_pop\_per\_gen, if true, will insert a "Pop" line between the}
349 \textcolor{stringliteral}{        generations, as given by the 2nd field in the gen file}
350 \textcolor{stringliteral}{}
351 \textcolor{stringliteral}{        2017\_02\_12.  In implementing a bias adjustment in the NeEstimator's ldne-based}
352 \textcolor{stringliteral}{        Nb estimates, calculated in pgdriveneestimator.py, We add to this def an arg }
353 \textcolor{stringliteral}{        "f\_nbne\_ratio", defalut None.  If arg is non-None, we append to the header line }
354 \textcolor{stringliteral}{        text, "nbne=str(f\_nbne\_ratio)". When the GUI interface for the Nb estimation reads}
355 \textcolor{stringliteral}{        in genepop files, it will check for they key=value string and then, if present,}
356 \textcolor{stringliteral}{        pass the value on to the calls that eventually invoke the pgdriveneestimator.}
357 \textcolor{stringliteral}{        '''}
358 
359         o\_genfile=\textcolor{keywordtype}{None}
360 
361         o\_genepopfile=\textcolor{keywordtype}{None}
362     
363         \textcolor{stringliteral}{'''}
364 \textcolor{stringliteral}{        2017\_04\_21.  We change the way we name the temp file, so that it will be written}
365 \textcolor{stringliteral}{        int he output base path, rather than the currdir. Note that we define the var}
366 \textcolor{stringliteral}{        here, but create the name just before using, as the check for duplicate file names,}
367 \textcolor{stringliteral}{        which is perfomred in the call to our new def \_\_get\_temp\_file\_name, }
368 \textcolor{stringliteral}{        should be done just before we then open the file.}
369 \textcolor{stringliteral}{        '''}
370         s\_temp\_file\_name=\textcolor{keywordtype}{None}
371     
372         s\_genepop\_filename=self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a05d91a106b2529b6cbae7777717a03b6}{\_\_genepopname}  
373 
374         i\_num\_loci=(idx\_allele\_stop-idx\_allele\_start) + 1
375 
376         IDX\_GEN\_INDIV=0
377         IDX\_GEN\_GENERATION=1
378         IDX\_FIRST\_ALLELE=2
379 
380         i\_genepop\_already\_exists=self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a1fdfb988619bd4a9d377a1e20f52839c}{\_\_file\_exists}( s\_genepop\_filename )
381 
382         \textcolor{keywordflow}{if} i\_genepop\_already\_exists != FILE\_DOES\_NOT\_EXIST:
383 
384             self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_ad3f4c271b5b800d0137460e2349452e0}{\_\_raise\_file\_exists\_error}( s\_genepop\_filename, 
385                     \textcolor{stringliteral}{"create genepop file from gen file"} )
386         \textcolor{comment}{#end if genepop file already exists}
387 
388         i\_genfile\_exists=self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a1fdfb988619bd4a9d377a1e20f52839c}{\_\_file\_exists}( self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a30fb6b94af13efad6becfbe6fddc1d95}{\_\_genfile} )
389         
390         \textcolor{comment}{#if we have the uncompressed version available}
391         \textcolor{comment}{#we use it:}
392         \textcolor{keywordflow}{if} i\_genfile\_exists == FILE\_EXISTS\_UNCOMPRESSED \(\backslash\)
393             \textcolor{keywordflow}{or} i\_genfile\_exists == FILE\_EXISTS\_AS\_BOTH\_UNCOMPRESSED\_AND\_BZ2:
394 
395             o\_genfile=open( self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a30fb6b94af13efad6becfbe6fddc1d95}{\_\_genfile} )
396 
397         \textcolor{keywordflow}{elif} i\_genfile\_exists == FILE\_EXISTS\_AS\_BZ2:
398             o\_genfile=bz2.BZ2File( self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a30fb6b94af13efad6becfbe6fddc1d95}{\_\_genfile} + \textcolor{stringliteral}{".bz2"}, mode=\textcolor{stringliteral}{'r' )}
399 \textcolor{stringliteral}{        }\textcolor{keywordflow}{else}:
400             self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_abf6aab4f7982c2b0f9e4544343bf241b}{\_\_raise\_file\_not\_found\_error}( self.
      \hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a30fb6b94af13efad6becfbe6fddc1d95}{\_\_genfile}, \textcolor{stringliteral}{"convert gen file to genepop"} )
401         \textcolor{comment}{#end if uncompressed only or uncomp. and compressed, else compressed only, else no file}
402 
403         s\_temp\_file\_name=self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a719984e6f178fb39b6ab349674251738}{\_\_get\_temp\_file\_name}()
404 
405         \textcolor{keywordflow}{if} b\_do\_compress == \textcolor{keyword}{True}:
406             o\_genepopfile=bz2.BZ2File( s\_temp\_file\_name + \textcolor{stringliteral}{'.bz2'}, \textcolor{stringliteral}{'w'}, compresslevel=9 ) 
407         \textcolor{keywordflow}{else}:
408             o\_genepopfile=open( s\_temp\_file\_name, \textcolor{stringliteral}{'w'} )
409         \textcolor{comment}{#end if compress else don't}
410 
411         \textcolor{comment}{#write header and loci list:}
412         self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a99bc161de61efea5751a230b9b4c9950}{writeGenepopFileHeaderAndLociList}( o\_genepopfile, 
413                                                     i\_num\_loci, 
414                                                     f\_nbne\_ratio, 
415                                                     b\_do\_compress )
416         
417         \textcolor{comment}{#write indiv id and alleles, line by line:}
418         i\_currgen=\textcolor{keywordtype}{None}
419 
420         \textcolor{keywordflow}{for} s\_line \textcolor{keywordflow}{in} o\_genfile:
421 
422             \textcolor{stringliteral}{'''}
423 \textcolor{stringliteral}{            2017\_05\_02.  For python 3, because in a compressed file python 3 reads bytes}
424 \textcolor{stringliteral}{            instead of strings (despite opening bzw2 in in mode='r'), }
425 \textcolor{stringliteral}{            we need to decode }\textcolor{keywordflow}{if} we read \textcolor{keywordflow}{in} bytes.
426             \textcolor{stringliteral}{'''}
427 \textcolor{stringliteral}{            if type( s\_line ) == bytes:}
428 \textcolor{stringliteral}{                s\_line=s\_line.decode( SYSDEFAULTENCODING )}
429 \textcolor{stringliteral}{            #end if type is bytes}
430 \textcolor{stringliteral}{}
431 \textcolor{stringliteral}{            ls\_line = s\_line.rstrip().split(" ") }
432 \textcolor{stringliteral}{}
433 \textcolor{stringliteral}{            '''}
434             As of 2016\_09\_01, we are no longer typing the id \textcolor{keyword}{as} an int, since
435             we changed the indiv id string that \textcolor{keywordflow}{is} written to the gen file,
436             (see pgopsimupop.py \textcolor{keyword}{def }\_\_outputAge )
437             to be a (currenly semicolon) delimited list of the indiv age,sex,etc)
438             \textcolor{stringliteral}{'''}
439 \textcolor{stringliteral}{            }
440 \textcolor{stringliteral}{            s\_id = ls\_line[IDX\_GEN\_INDIV]}
441 \textcolor{stringliteral}{}
442 \textcolor{stringliteral}{            i\_gen = int(float(ls\_line[IDX\_GEN\_GENERATION]))}
443 \textcolor{stringliteral}{            idx\_first\_allele\_in\_line=( idx\_allele\_start + IDX\_FIRST\_ALLELE ) - 1}
444 \textcolor{stringliteral}{            idx\_last\_allele\_in\_line=( idx\_allele\_stop  + IDX\_FIRST\_ALLELE ) - 1 }
445 \textcolor{stringliteral}{            ls\_alleles = ls\_line[ idx\_first\_allele\_in\_line : idx\_last\_allele\_in\_line + 1 ]}
446 \textcolor{stringliteral}{}
447 \textcolor{stringliteral}{            if i\_currgen is None or ( b\_pop\_per\_gen == True and i\_currgen != i\_gen ):}
448 \textcolor{stringliteral}{                s\_pop\_line="pop\(\backslash\)n"}
449 \textcolor{stringliteral}{}
450 \textcolor{stringliteral}{                if b\_do\_compress:}
451 \textcolor{stringliteral}{                        s\_pop\_line=s\_pop\_line.encode( SYSDEFAULTENCODING )}
452 \textcolor{stringliteral}{                #end if b\_do\_compress}
453 \textcolor{stringliteral}{}
454 \textcolor{stringliteral}{                o\_genepopfile.write( s\_pop\_line )}
455 \textcolor{stringliteral}{                i\_currgen=i\_gen}
456 \textcolor{stringliteral}{}
457 \textcolor{stringliteral}{            #end if no curr gen or new gen and we write pop per gen}
458 \textcolor{stringliteral}{            }
459 \textcolor{stringliteral}{            '''}
460             2017\_03\_20.  Until now the default way to join the alleles with the
461             individual id was simply to separate with a comma.  NeEstimator2 was
462             robust to this \textcolor{keywordflow}{and} correctly parsed.  However, now that we are also
463             using LDNe2 to get ldne estimates, it \textcolor{keywordflow}{is} necessary to put a space
464             after the comma, \textcolor{keywordflow}{or} the latter executable throws \textcolor{keywordflow}{and} error.  Also,
465             note that the comma plus space \textcolor{keywordflow}{is} the usual way to separate, \textcolor{keywordflow}{and}
466             though \textcolor{keywordflow}{not} specified \textcolor{keywordflow}{in} the file spec, it \textcolor{keywordflow}{is} the only way, now that
467             I\textcolor{stringliteral}{'ve encountered the error, that other genepop files I'}ve looked at
468             are formatted.
469             \textcolor{stringliteral}{'''}
470 \textcolor{stringliteral}{            if IN\_POP\_SECTIONS\_SEPARATE\_FIRST\_LOCI\_FROM\_COMMA\_WITH\_SPACE==True:}
471 \textcolor{stringliteral}{                s\_delimit\_indiv\_from\_loci=", "}
472 \textcolor{stringliteral}{            else:}
473 \textcolor{stringliteral}{                s\_delimit\_indiv\_from\_loci=","}
474 \textcolor{stringliteral}{            #end if we aret to separate first loci from indiv with comma and space}
475 \textcolor{stringliteral}{}
476 \textcolor{stringliteral}{            s\_this\_entry=s\_id + s\_delimit\_indiv\_from\_loci +  " ".join( ls\_alleles ) + "\(\backslash\)n" }
477 \textcolor{stringliteral}{        }
478 \textcolor{stringliteral}{            if b\_do\_compress:}
479 \textcolor{stringliteral}{                s\_this\_entry=s\_this\_entry.encode( SYSDEFAULTENCODING )}
480 \textcolor{stringliteral}{            #end if writing compressed data}
481 \textcolor{stringliteral}{}
482 \textcolor{stringliteral}{            o\_genepopfile.write( s\_this\_entry )}
483 \textcolor{stringliteral}{}
484 \textcolor{stringliteral}{        #end for each line in gen file}
485 \textcolor{stringliteral}{}
486 \textcolor{stringliteral}{        o\_genfile.close()}
487 \textcolor{stringliteral}{        o\_genepopfile.close()}
488 \textcolor{stringliteral}{}
489 \textcolor{stringliteral}{        s\_final\_name=s\_temp\_file\_name.replace( s\_temp\_file\_name, s\_genepop\_filename )}
490 \textcolor{stringliteral}{}
491 \textcolor{stringliteral}{        #change to shutil.move from os.rename -- threw errors (in docker install) when}
492 \textcolor{stringliteral}{        #renaming across volumes:}
493 \textcolor{stringliteral}{        if b\_do\_compress:}
494 \textcolor{stringliteral}{            s\_temp\_file\_name=s\_temp\_file\_name + ".bz2"}
495 \textcolor{stringliteral}{            s\_final\_name=s\_final\_name + ".bz2"}
496 \textcolor{stringliteral}{        #end if we wrote a compressed file}
497 \textcolor{stringliteral}{}
498 \textcolor{stringliteral}{        shutil.move( s\_temp\_file\_name, s\_final\_name )}
499 \textcolor{stringliteral}{}
500 \textcolor{stringliteral}{        return}
\textcolor{stringliteral}{\end{DoxyCode}
\index{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}!write\+Genepop\+File\+Header\+And\+Loci\+List@{write\+Genepop\+File\+Header\+And\+Loci\+List}}
\index{write\+Genepop\+File\+Header\+And\+Loci\+List@{write\+Genepop\+File\+Header\+And\+Loci\+List}!negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}}
\subsubsection[{\texorpdfstring{write\+Genepop\+File\+Header\+And\+Loci\+List(self, o\+\_\+genepopfile, i\+\_\+num\+\_\+loci, f\+\_\+nbne\+\_\+ratio=\+None, b\+\_\+do\+\_\+compress=\+False)}{writeGenepopFileHeaderAndLociList(self, o_genepopfile, i_num_loci, f_nbne_ratio=None, b_do_compress=False)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop.\+write\+Genepop\+File\+Header\+And\+Loci\+List (
\begin{DoxyParamCaption}
\item[{}]{self, }
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{i\+\_\+num\+\_\+loci, }
\item[{}]{f\+\_\+nbne\+\_\+ratio = {\ttfamily None}, }
\item[{}]{b\+\_\+do\+\_\+compress = {\ttfamily False}}
\end{DoxyParamCaption}
)}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a99bc161de61efea5751a230b9b4c9950}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a99bc161de61efea5751a230b9b4c9950}
\begin{DoxyVerb}2017_08_04. Code to write the genpop file header
and loci list was moved here from def gen2Genepop
so that it can also be called from a PGOpSimuPop object
that is writing the genepop file using the new
mode that skips the intermediate *.gen, *.sim , etc.
files.
\end{DoxyVerb}
 

Definition at line 282 of file pgoutputsimupop.\+py.


\begin{DoxyCode}
282                                 b\_do\_compress=\textcolor{keyword}{False} ):
283     
284         \textcolor{stringliteral}{'''}
285 \textcolor{stringliteral}{        2017\_08\_04. Code to write the genpop file header}
286 \textcolor{stringliteral}{        and loci list was moved here from def gen2Genepop}
287 \textcolor{stringliteral}{        so that it can also be called from a PGOpSimuPop object}
288 \textcolor{stringliteral}{        that is writing the genepop file using the new}
289 \textcolor{stringliteral}{        mode that skips the intermediate *.gen, *.sim , etc.}
290 \textcolor{stringliteral}{        files.}
291 \textcolor{stringliteral}{        '''}
292 
293         \textcolor{comment}{#write header and loci list:}
294         s\_header=\textcolor{stringliteral}{"genepop from simupop sim run, data file "} + self.\hyperlink{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a30fb6b94af13efad6becfbe6fddc1d95}{\_\_genfile}
295 
296         \textcolor{keywordflow}{if} f\_nbne\_ratio \textcolor{keywordflow}{is} \textcolor{keywordflow}{not} \textcolor{keywordtype}{None}:
297             s\_header += \textcolor{stringliteral}{", nbne="} + str( f\_nbne\_ratio )
298         \textcolor{comment}{#end if caller passed and nb/ne ratio value}
299 
300         s\_header\_line=s\_header + \textcolor{stringliteral}{"\(\backslash\)n"}
301 
302         \textcolor{keywordflow}{if} b\_do\_compress:
303             s\_header\_line=s\_header\_line.encode( SYSDEFAULTENCODING )
304         \textcolor{comment}{#end if writing compressed data}
305 
306         o\_genepopfile.write(  s\_header\_line )
307 
308 
309         \textcolor{keywordflow}{for} i\_locus \textcolor{keywordflow}{in} range( i\_num\_loci ):
310             s\_locus\_name=\textcolor{stringliteral}{"l"} + str(i\_locus) + \textcolor{stringliteral}{"\(\backslash\)n"} 
311 
312             \textcolor{keywordflow}{if} b\_do\_compress:
313                 s\_locus\_name=s\_locus\_name.encode( SYSDEFAULTENCODING )
314             \textcolor{comment}{#end if do compress }
315 
316             o\_genepopfile.write(s\_locus\_name )
317 
318         \textcolor{comment}{#end for each loci number}
319 
320         \textcolor{keywordflow}{return}
\end{DoxyCode}


\subsection{Member Data Documentation}
\index{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}!D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS@{D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS}}
\index{D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS@{D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS}!negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop@{negui\+::pgoutputsimupop\+::\+P\+G\+Output\+Simu\+Pop}}
\subsubsection[{\texorpdfstring{D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS}{DICT_OUTPUT_FILE_EXTENSIONS}}]{\setlength{\rightskip}{0pt plus 5cm}dictionary negui.\+pgoutputsimupop.\+P\+G\+Output\+Simu\+Pop.\+D\+I\+C\+T\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+E\+X\+T\+E\+N\+S\+I\+O\+NS\hspace{0.3cm}{\ttfamily [static]}}\hypertarget{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a9c9afd31cd09eedc7187eaf6a48ef74a}{}\label{classnegui_1_1pgoutputsimupop_1_1PGOutputSimuPop_a9c9afd31cd09eedc7187eaf6a48ef74a}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = \{ \textcolor{stringliteral}{"simfile"}:\textcolor{stringliteral}{"sim"},
2                                     \textcolor{stringliteral}{"genfile"}:\textcolor{stringliteral}{"gen"},
3                                     \textcolor{stringliteral}{"dbfile"}:\textcolor{stringliteral}{"db"},
4                                     \textcolor{stringliteral}{"conffile"}:\textcolor{stringliteral}{"conf"},
5                                     \textcolor{stringliteral}{"genepop"}:\textcolor{stringliteral}{"genepop"},
6                                     \textcolor{stringliteral}{"age\_counts"}:\textcolor{stringliteral}{"\_age\_counts\_by\_gen.tsv"},
7                                     \textcolor{stringliteral}{"sim\_nb\_estimates"}:\textcolor{stringliteral}{"\_nb\_values\_calc\_by\_gen.tsv"},
8                                     \textcolor{stringliteral}{"het\_filter\_info"}: \textcolor{stringliteral}{"\_het\_value\_by\_cycle\_number.tsv"} \}
\end{DoxyCode}


Definition at line 75 of file pgoutputsimupop.\+py.



The documentation for this class was generated from the following file\+:\begin{DoxyCompactItemize}
\item 
pgoutputsimupop.\+py\end{DoxyCompactItemize}
