\hypertarget{namespacenegui_1_1pgdriveneestimator}{}\section{negui.\+pgdriveneestimator Namespace Reference}
\label{namespacenegui_1_1pgdriveneestimator}\index{negui.\+pgdriveneestimator@{negui.\+pgdriveneestimator}}
\subsection*{Classes}
\begin{DoxyCompactItemize}
\item 
class \hyperlink{classnegui_1_1pgdriveneestimator_1_1ArgSet}{Arg\+Set}
\item 
class \hyperlink{classnegui_1_1pgdriveneestimator_1_1Counter}{Counter}
\item 
class \hyperlink{classnegui_1_1pgdriveneestimator_1_1DebugMode}{Debug\+Mode}
\end{DoxyCompactItemize}
\subsection*{Functions}
\begin{DoxyCompactItemize}
\item 
def {\bfseries set\+\_\+indices\+\_\+ne\+\_\+estimator\+\_\+output\+\_\+fields\+\_\+to\+\_\+skip} ()\hypertarget{namespacenegui_1_1pgdriveneestimator_a59077d0d79daf6f8b1e39172a12b5a34}{}\label{namespacenegui_1_1pgdriveneestimator_a59077d0d79daf6f8b1e39172a12b5a34}

\item 
def {\bfseries get\+\_\+datetime\+\_\+as\+\_\+string} ()\hypertarget{namespacenegui_1_1pgdriveneestimator_a7c40584a0a1685f8175c6034422adf4c}{}\label{namespacenegui_1_1pgdriveneestimator_a7c40584a0a1685f8175c6034422adf4c}

\item 
def {\bfseries set\+\_\+debug\+\_\+mode} (s\+\_\+arg)\hypertarget{namespacenegui_1_1pgdriveneestimator_a2b44aee684d792b951e5787db7b69189}{}\label{namespacenegui_1_1pgdriveneestimator_a2b44aee684d792b951e5787db7b69189}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a8972bdaeb431f941619300e9e63a0732}{get\+\_\+invalid\+\_\+file\+\_\+names} (ls\+\_\+file\+\_\+names)
\item 
def {\bfseries file\+\_\+name\+\_\+list\+\_\+is\+\_\+all\+\_\+strings} (ls\+\_\+files)\hypertarget{namespacenegui_1_1pgdriveneestimator_a3bbea10cf2e57b63cac6fffdbabd2f37}{}\label{namespacenegui_1_1pgdriveneestimator_a3bbea10cf2e57b63cac6fffdbabd2f37}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_ad0c4544bb6ce5bc184615cfade0f8495}{get\+\_\+genepop\+\_\+file\+\_\+list} (s\+\_\+genepop\+\_\+files\+\_\+arg)
\item 
def {\bfseries parse\+\_\+args} (args)\hypertarget{namespacenegui_1_1pgdriveneestimator_a014177779b9f1991ca67336bd39160fa}{}\label{namespacenegui_1_1pgdriveneestimator_a014177779b9f1991ca67336bd39160fa}

\item 
def {\bfseries write\+\_\+parms\+\_\+to\+\_\+file} (lv\+\_\+params)\hypertarget{namespacenegui_1_1pgdriveneestimator_a0b74086907fdc54d2dbb0f9d797ec678}{}\label{namespacenegui_1_1pgdriveneestimator_a0b74086907fdc54d2dbb0f9d797ec678}

\item 
def {\bfseries get\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+sample\+\_\+name} (s\+\_\+sample\+\_\+name, s\+\_\+sample\+\_\+scheme)\hypertarget{namespacenegui_1_1pgdriveneestimator_a85f2f02c06f18fb4888a00d40ffa9361}{}\label{namespacenegui_1_1pgdriveneestimator_a85f2f02c06f18fb4888a00d40ffa9361}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a4551e387ca20ed229183818ac8ce733c}{raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file} (o\+\_\+genepopfile)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a06645f9b380f736368e27585331ec22f}{get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range} (o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a7df54bcca5c33c80e648c440527a85ce}{get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment} (o\+\_\+ne\+\_\+estimator, lv\+\_\+results\+\_\+list, f\+\_\+nbne\+\_\+ratio)
\item 
def {\bfseries get\+\_\+nbne\+\_\+ratio\+\_\+from\+\_\+genepop\+\_\+file\+\_\+header} (o\+\_\+genepopfile)\hypertarget{namespacenegui_1_1pgdriveneestimator_a94131d79681342c10290d15abb65c696}{}\label{namespacenegui_1_1pgdriveneestimator_a94131d79681342c10290d15abb65c696}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a99dfc5e3dc217054199cd4f693e26236}{do\+\_\+ldne\+\_\+bias\+\_\+adjustment} (f\+\_\+ldne\+\_\+estimate, f\+\_\+nbne\+\_\+ratio)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_aeb13bee850fbafd8116a6b7e8048b98c}{do\+\_\+estimate} (xxx\+\_\+todo\+\_\+changeme)
\item 
def {\bfseries write\+\_\+results} (ds\+\_\+results, o\+\_\+main\+\_\+outfile, o\+\_\+secondary\+\_\+outfile)\hypertarget{namespacenegui_1_1pgdriveneestimator_af2d5e050f9a707ed67728913d3ac620f}{}\label{namespacenegui_1_1pgdriveneestimator_af2d5e050f9a707ed67728913d3ac620f}

\item 
def {\bfseries update\+\_\+indiv\+\_\+list} (dddli\+\_\+indiv\+\_\+list, dv\+\_\+indiv\+\_\+info\+\_\+for\+\_\+replicate, lv\+\_\+sample\+\_\+vals)\hypertarget{namespacenegui_1_1pgdriveneestimator_a4185236461b3f8636d0ca816a84f28f8}{}\label{namespacenegui_1_1pgdriveneestimator_a4185236461b3f8636d0ca816a84f28f8}

\item 
def {\bfseries make\+\_\+indiv\+\_\+table\+\_\+name} ()\hypertarget{namespacenegui_1_1pgdriveneestimator_a2a22ddf30dc8ad29107131c4d4ca9be6}{}\label{namespacenegui_1_1pgdriveneestimator_a2a22ddf30dc8ad29107131c4d4ca9be6}

\item 
def {\bfseries get\+\_\+indiv\+\_\+table\+\_\+header} (lf\+\_\+proportions)\hypertarget{namespacenegui_1_1pgdriveneestimator_a5764313a499a7eb9fc99db74768cd10d}{}\label{namespacenegui_1_1pgdriveneestimator_a5764313a499a7eb9fc99db74768cd10d}

\item 
def {\bfseries write\+\_\+indiv\+\_\+table} (dddli\+\_\+indiv\+\_\+list, s\+\_\+file\+\_\+name, lf\+\_\+proportions)\hypertarget{namespacenegui_1_1pgdriveneestimator_a263379187199613b753db847b3105160}{}\label{namespacenegui_1_1pgdriveneestimator_a263379187199613b753db847b3105160}

\item 
def {\bfseries print\+\_\+test\+\_\+list\+\_\+replicate\+\_\+selection\+\_\+indices} (o\+\_\+genepopfile, s\+\_\+sample\+\_\+scheme, s\+\_\+indiv\+\_\+subsample\+\_\+tag, s\+\_\+loci\+\_\+sample\+\_\+value, s\+\_\+loci\+\_\+replicate\+\_\+number, o\+\_\+secondary\+\_\+outfile, s\+\_\+population\+\_\+subsample\+\_\+tag)\hypertarget{namespacenegui_1_1pgdriveneestimator_ad9f7e28c15ac7902cb795cb27997648e}{}\label{namespacenegui_1_1pgdriveneestimator_ad9f7e28c15ac7902cb795cb27997648e}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a8918df5a619173d4baa530e79dfe2385}{parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields} (ls\+\_\+sampling\+\_\+list\+\_\+items)
\item 
def {\bfseries get\+\_\+field\+\_\+names\+\_\+from\+\_\+criteria\+\_\+test} (ls\+\_\+field\+\_\+names, s\+\_\+test)\hypertarget{namespacenegui_1_1pgdriveneestimator_ab7cb1ccdd969e3d2cd3ba2b8dadaa4d9}{}\label{namespacenegui_1_1pgdriveneestimator_ab7cb1ccdd969e3d2cd3ba2b8dadaa4d9}

\item 
def {\bfseries get\+\_\+criteria\+\_\+sampling\+\_\+params} (lv\+\_\+sample\+\_\+values)\hypertarget{namespacenegui_1_1pgdriveneestimator_a08185cf9c5aff8db21ee0eefa04cd237}{}\label{namespacenegui_1_1pgdriveneestimator_a08185cf9c5aff8db21ee0eefa04cd237}

\item 
def {\bfseries get\+\_\+cohorts\+\_\+sampling\+\_\+params} (lv\+\_\+sample\+\_\+values)\hypertarget{namespacenegui_1_1pgdriveneestimator_ac5418cd5138d005dae534383305be65e}{}\label{namespacenegui_1_1pgdriveneestimator_ac5418cd5138d005dae534383305be65e}

\item 
def {\bfseries get\+\_\+relateds\+\_\+sampling\+\_\+params} (lv\+\_\+sample\+\_\+values)\hypertarget{namespacenegui_1_1pgdriveneestimator_ab57b50b97d036e0ac3cd5d62728312ba}{}\label{namespacenegui_1_1pgdriveneestimator_ab57b50b97d036e0ac3cd5d62728312ba}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a74f962e4a6815d8acc2e23c9ad146d0d}{do\+\_\+sample} (o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+range, i\+\_\+max\+\_\+pop\+\_\+range, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, i\+\_\+replicates, s\+\_\+loci\+\_\+sampling\+\_\+scheme, v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, i\+\_\+min\+\_\+total\+\_\+loci, i\+\_\+max\+\_\+total\+\_\+loci, i\+\_\+loci\+\_\+replicates)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_aa663cfeded8b8a3710d1a11192f12581}{do\+\_\+sample\+\_\+individuals} (o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+range, i\+\_\+max\+\_\+pop\+\_\+range, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, i\+\_\+replicates)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a408b2c39ad32bd720e35f2eace94509c}{do\+\_\+sample\+\_\+loci} (o\+\_\+genepopfile, s\+\_\+loci\+\_\+sampling\+\_\+scheme, v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, i\+\_\+min\+\_\+total\+\_\+loci, i\+\_\+max\+\_\+total\+\_\+loci, li\+\_\+population\+\_\+list, i\+\_\+loci\+\_\+replicates=1)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_af559a2d83516e3fb3fcebbfcdc96881b}{get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name} (s\+\_\+original\+\_\+genepop\+\_\+file\+\_\+name, s\+\_\+loci\+\_\+subsample\+\_\+tag, s\+\_\+this\+\_\+pop\+\_\+sample\+\_\+name, s\+\_\+temporary\+\_\+directory)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a9c650be4ed8ae5fa5e599523c22273bd}{make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name} (s\+\_\+original\+\_\+genepop\+\_\+file, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count, i\+\_\+indiv\+\_\+sample\+\_\+count, i\+\_\+loci\+\_\+subsample\+\_\+count)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a0f589286698490f5054d972daa7c9160}{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate} (o\+\_\+genepopfile, s\+\_\+sample\+\_\+scheme, f\+\_\+min\+\_\+allele\+\_\+freq, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, o\+\_\+debug\+\_\+mode, llv\+\_\+args\+\_\+each\+\_\+process, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, o\+\_\+secondary\+\_\+outfile, f\+\_\+nbne\+\_\+ratio, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}{get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag} (s\+\_\+loci\+\_\+subsample\+\_\+tag)
\item 
def {\bfseries get\+\_\+loci\+\_\+subsample\+\_\+tags\+\_\+for\+\_\+this\+\_\+indiv\+\_\+subsample} (o\+\_\+genepopfile, s\+\_\+indiv\+\_\+sample\+\_\+tag)\hypertarget{namespacenegui_1_1pgdriveneestimator_a3ad4e94d91a0c1c1757084b9cfe6f3af}{}\label{namespacenegui_1_1pgdriveneestimator_a3ad4e94d91a0c1c1757084b9cfe6f3af}

\item 
def {\bfseries write\+\_\+result\+\_\+sets} (lds\+\_\+results, lv\+\_\+sample\+\_\+values, o\+\_\+debug\+\_\+mode, o\+\_\+main\+\_\+outfile, o\+\_\+secondary\+\_\+outfile)\hypertarget{namespacenegui_1_1pgdriveneestimator_a9664c63b7ac063e43774e89e7c9674c2}{}\label{namespacenegui_1_1pgdriveneestimator_a9664c63b7ac063e43774e89e7c9674c2}

\item 
def {\bfseries execute\+\_\+ne\+\_\+for\+\_\+each\+\_\+sample} (llv\+\_\+args\+\_\+each\+\_\+process, o\+\_\+process\+\_\+pool, o\+\_\+debug\+\_\+mode, o\+\_\+multiprocessing\+\_\+event)\hypertarget{namespacenegui_1_1pgdriveneestimator_ad09b9b97dd438ce3e0f4fcf0206e6b22}{}\label{namespacenegui_1_1pgdriveneestimator_ad09b9b97dd438ce3e0f4fcf0206e6b22}

\item 
def {\bfseries write\+\_\+header\+\_\+main\+\_\+table} (I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, o\+\_\+main\+\_\+outfile)\hypertarget{namespacenegui_1_1pgdriveneestimator_a547e55a4d06bc1d5edf1b9d254dd34ca}{}\label{namespacenegui_1_1pgdriveneestimator_a547e55a4d06bc1d5edf1b9d254dd34ca}

\item 
def {\bfseries get\+\_\+sample\+\_\+abbrevs\+\_\+for\+\_\+criteria} (lds\+\_\+results)\hypertarget{namespacenegui_1_1pgdriveneestimator_a0b7ada08342688dc19f3be50db30ad0b}{}\label{namespacenegui_1_1pgdriveneestimator_a0b7ada08342688dc19f3be50db30ad0b}

\item 
def {\bfseries drive\+\_\+estimator} (args)\hypertarget{namespacenegui_1_1pgdriveneestimator_a75b4a299d8030d8bbca11e01fd3211da}{}\label{namespacenegui_1_1pgdriveneestimator_a75b4a299d8030d8bbca11e01fd3211da}

\item 
def {\bfseries set\+\_\+mod\+\_\+level\+\_\+var\+\_\+to\+\_\+ldne2\+\_\+executable\+\_\+path} ()\hypertarget{namespacenegui_1_1pgdriveneestimator_a5d449c8ab43f1e25695a00307cc55862}{}\label{namespacenegui_1_1pgdriveneestimator_a5d449c8ab43f1e25695a00307cc55862}

\item 
def {\bfseries did\+\_\+find\+\_\+ne\+\_\+estimator\+\_\+executable} ()\hypertarget{namespacenegui_1_1pgdriveneestimator_a18217c273c60b3009f24ab642d00bae5}{}\label{namespacenegui_1_1pgdriveneestimator_a18217c273c60b3009f24ab642d00bae5}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a4010f65f71a1bce539c5e15317632be2}{mymainlongfilelist} (q\+\_\+args)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a54013fc419eb559e07a5d278baa3fc6d}{set\+\_\+messaging\+\_\+procedures} (s\+\_\+gui\+\_\+flag)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a8344146a59d84da02da349abd68723ad}{mymain} (q\+\_\+args)
\item 
def {\bfseries get\+\_\+ranges\+\_\+for\+\_\+list\+\_\+slices} (i\+\_\+length\+\_\+of\+\_\+list, i\+\_\+slice\+\_\+size=P\+O\+O\+L\+\_\+\+B\+A\+T\+C\+H\+\_\+\+S\+I\+ZE)\hypertarget{namespacenegui_1_1pgdriveneestimator_a093ecb7ad7eaff69fd621e066fcdc29d}{}\label{namespacenegui_1_1pgdriveneestimator_a093ecb7ad7eaff69fd621e066fcdc29d}

\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a810fb4b95084358f69aafdc1e048451d}{get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs} (o\+\_\+genepopfile, s\+\_\+pop\+\_\+subsample\+\_\+tag=None, s\+\_\+individual\+\_\+subsample\+\_\+tag=None, s\+\_\+loci\+\_\+subsample\+\_\+tag=None)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_a4d13e12b51231deb46355be190102cbc}{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate} (o\+\_\+genepopfile, s\+\_\+sample\+\_\+scheme, f\+\_\+min\+\_\+allele\+\_\+freq, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, o\+\_\+debug\+\_\+mode, llv\+\_\+args\+\_\+each\+\_\+process, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, o\+\_\+secondary\+\_\+outfile, f\+\_\+nbne\+\_\+ratio, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count)
\item 
def \hyperlink{namespacenegui_1_1pgdriveneestimator_aaf461586686c9d7f0bf90188ab241dcc}{get\+\_\+count\+\_\+estimator\+\_\+fields} ()
\item 
def {\bfseries get\+\_\+total\+\_\+bytes\+\_\+needed\+\_\+for\+\_\+genepop\+\_\+object} (s\+\_\+filename)\hypertarget{namespacenegui_1_1pgdriveneestimator_ab41c87e91be7262bc61207e9e987fcd2}{}\label{namespacenegui_1_1pgdriveneestimator_ab41c87e91be7262bc61207e9e987fcd2}

\item 
def {\bfseries get\+\_\+total\+\_\+ram\+\_\+usage\+\_\+for\+\_\+this\+\_\+file} (s\+\_\+filename, o\+\_\+debug\+\_\+mode, i\+\_\+total\+\_\+processes)\hypertarget{namespacenegui_1_1pgdriveneestimator_ae85cd0fbf1a18d081c098a7f31a340d2}{}\label{namespacenegui_1_1pgdriveneestimator_ae85cd0fbf1a18d081c098a7f31a340d2}

\item 
def {\bfseries file\+\_\+can\+\_\+be\+\_\+added\+\_\+to\+\_\+current\+\_\+set} (s\+\_\+filename, o\+\_\+total\+\_\+bytes\+\_\+currently\+\_\+used\+\_\+for\+\_\+gp\+\_\+objects, i\+\_\+total\+\_\+processes, o\+\_\+debug\+\_\+mode, i\+\_\+bytes\+\_\+available\+\_\+virtual\+\_\+memory)\hypertarget{namespacenegui_1_1pgdriveneestimator_ac52a11ee572f6b104b8b2af3076b6479}{}\label{namespacenegui_1_1pgdriveneestimator_ac52a11ee572f6b104b8b2af3076b6479}

\item 
def {\bfseries add\+\_\+file\+\_\+to\+\_\+current\+\_\+set} (s\+\_\+filename, i\+\_\+min\+\_\+pop\+\_\+range, i\+\_\+max\+\_\+pop\+\_\+range, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, i\+\_\+total\+\_\+replicates, s\+\_\+loci\+\_\+sampling\+\_\+scheme, v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, i\+\_\+min\+\_\+total\+\_\+loci, i\+\_\+max\+\_\+total\+\_\+loci, i\+\_\+loci\+\_\+replicates, llv\+\_\+args\+\_\+each\+\_\+process, b\+\_\+do\+\_\+nb\+\_\+bias\+\_\+adjustment, f\+\_\+nbne\+\_\+ratio, f\+\_\+min\+\_\+allele\+\_\+freq, o\+\_\+debug\+\_\+mode, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, i\+\_\+genepop\+\_\+file\+\_\+count, s\+\_\+temporary\+\_\+directory, o\+\_\+secondary\+\_\+outfile)\hypertarget{namespacenegui_1_1pgdriveneestimator_ae45071494b5c39165cac87c992d6e5cd}{}\label{namespacenegui_1_1pgdriveneestimator_ae45071494b5c39165cac87c992d6e5cd}

\item 
def {\bfseries raise\+\_\+exception\+\_\+file\+\_\+size\+\_\+limit} (s\+\_\+filename, o\+\_\+debug\+\_\+mode, i\+\_\+total\+\_\+processes)\hypertarget{namespacenegui_1_1pgdriveneestimator_a4a2864d554435ebb337c4c519511d93e}{}\label{namespacenegui_1_1pgdriveneestimator_a4a2864d554435ebb337c4c519511d93e}

\item 
def {\bfseries process\+\_\+current\+\_\+set\+\_\+of\+\_\+gp\+\_\+files} (llv\+\_\+args\+\_\+each\+\_\+process, s\+\_\+filename, o\+\_\+debug\+\_\+mode, i\+\_\+total\+\_\+processes, o\+\_\+multiprocessing\+\_\+event, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, o\+\_\+main\+\_\+outfile, o\+\_\+secondary\+\_\+outfile, o\+\_\+total\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, s\+\_\+all\+\_\+interrupt\+\_\+messages)\hypertarget{namespacenegui_1_1pgdriveneestimator_a6abd32b781a8d2d442ae8b140b6b2c30}{}\label{namespacenegui_1_1pgdriveneestimator_a6abd32b781a8d2d442ae8b140b6b2c30}

\end{DoxyCompactItemize}
\subsection*{Variables}
\begin{DoxyCompactItemize}
\item 
bool {\bfseries V\+E\+R\+B\+O\+SE} = False\hypertarget{namespacenegui_1_1pgdriveneestimator_a1fff696a30e4f191608d9ab8333f68e5}{}\label{namespacenegui_1_1pgdriveneestimator_a1fff696a30e4f191608d9ab8333f68e5}

\item 
bool {\bfseries V\+E\+R\+Y\+\_\+\+V\+E\+R\+B\+O\+SE} = False\hypertarget{namespacenegui_1_1pgdriveneestimator_aeead01603a5ef9c29a3f7cd3cfb8dfce}{}\label{namespacenegui_1_1pgdriveneestimator_aeead01603a5ef9c29a3f7cd3cfb8dfce}

\item 
{\bfseries A\+C\+T\+I\+V\+A\+T\+E\+\_\+\+G\+U\+I\+\_\+\+M\+E\+S\+S\+A\+G\+I\+NG} = None\hypertarget{namespacenegui_1_1pgdriveneestimator_a1d259b1af1d5cf6385db6b499c6488cb}{}\label{namespacenegui_1_1pgdriveneestimator_a1d259b1af1d5cf6385db6b499c6488cb}

\item 
{\bfseries G\+U\+I\+E\+RR} = None\hypertarget{namespacenegui_1_1pgdriveneestimator_a6ee6d86b2aa6effad38fe001b81243a7}{}\label{namespacenegui_1_1pgdriveneestimator_a6ee6d86b2aa6effad38fe001b81243a7}

\item 
{\bfseries G\+U\+I\+I\+N\+FO} = None\hypertarget{namespacenegui_1_1pgdriveneestimator_ad8d363e31b05edc13b445663ee3ab01c}{}\label{namespacenegui_1_1pgdriveneestimator_ad8d363e31b05edc13b445663ee3ab01c}

\item 
{\bfseries N\+E\+E\+S\+T\+I\+M\+A\+T\+OR} = pgne.\+N\+E\+E\+S\+T\+I\+M\+A\+T\+OR\hypertarget{namespacenegui_1_1pgdriveneestimator_ae82abb35ceeccb8108458b21eacd93ce}{}\label{namespacenegui_1_1pgdriveneestimator_ae82abb35ceeccb8108458b21eacd93ce}

\item 
{\bfseries L\+D\+N\+E2} = pgne.\+L\+D\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+I\+ON\hypertarget{namespacenegui_1_1pgdriveneestimator_aafc38c434d06f90cae059ec1bfb55906}{}\label{namespacenegui_1_1pgdriveneestimator_aafc38c434d06f90cae059ec1bfb55906}

\item 
{\bfseries E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+T\+O\+\_\+\+U\+SE} = L\+D\+N\+E2\hypertarget{namespacenegui_1_1pgdriveneestimator_ac84cd9a16546927d64f69d5d8e666eac}{}\label{namespacenegui_1_1pgdriveneestimator_ac84cd9a16546927d64f69d5d8e666eac}

\item 
{\bfseries P\+A\+T\+H\+\_\+\+T\+O\+\_\+\+L\+D\+N\+E2} = None\hypertarget{namespacenegui_1_1pgdriveneestimator_af9d9edfba4ccb35868fcc92a0121ff77}{}\label{namespacenegui_1_1pgdriveneestimator_af9d9edfba4ccb35868fcc92a0121ff77}

\item 
string {\bfseries spacer} = \char`\"{} \char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a783795e063ad4c9d423fdb51c16ef2d2}{}\label{namespacenegui_1_1pgdriveneestimator_a783795e063ad4c9d423fdb51c16ef2d2}

\item 
string {\bfseries spacer2} = \char`\"{} \char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a67dd1eeed8a1a30f704b56250ccce873}{}\label{namespacenegui_1_1pgdriveneestimator_a67dd1eeed8a1a30f704b56250ccce873}

\item 
string {\bfseries spacer3} = \char`\"{} \char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_aea4fce536a2ded454b3545a03bf9ea69}{}\label{namespacenegui_1_1pgdriveneestimator_aea4fce536a2ded454b3545a03bf9ea69}

\item 
list {\bfseries L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+S\+H\+O\+R\+T\+\_\+\+R\+E\+Q\+U\+I\+R\+ED} = \mbox{[} \char`\"{}-\/f\char`\"{}, \char`\"{}-\/s\char`\"{}, \char`\"{}-\/p\char`\"{}, \char`\"{}-\/m\char`\"{}, \char`\"{}-\/a\char`\"{}, \char`\"{}-\/r\char`\"{}, \char`\"{}-\/e\char`\"{}, \char`\"{}-\/c\char`\"{}, \char`\"{}-\/l\char`\"{}, \char`\"{}-\/i\char`\"{}, \char`\"{}-\/n\char`\"{}, \char`\"{}-\/x\char`\"{}, \char`\"{}-\/g\char`\"{},\char`\"{}-\/q\char`\"{} \mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_aae439ccd49f1eb54994fba6301243d0c}{}\label{namespacenegui_1_1pgdriveneestimator_aae439ccd49f1eb54994fba6301243d0c}

\item 
list {\bfseries L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}
\item 
list {\bfseries L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}\hypertarget{namespacenegui_1_1pgdriveneestimator_a724a8227b1b9d47cf4cc6079615ac6f2}{}\label{namespacenegui_1_1pgdriveneestimator_a724a8227b1b9d47cf4cc6079615ac6f2}

\item 
list {\bfseries L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+S\+H\+O\+R\+T\+\_\+\+O\+P\+T\+I\+O\+N\+AL} = \mbox{[} \char`\"{}-\/o\char`\"{}, \char`\"{}-\/d\char`\"{}, \char`\"{}-\/b\char`\"{}, \char`\"{}-\/j\char`\"{} \mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8adf3309d69b960bd35c1516501e5b71}{}\label{namespacenegui_1_1pgdriveneestimator_a8adf3309d69b960bd35c1516501e5b71}

\item 
list {\bfseries L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+O\+P\+T\+I\+O\+N\+AL} = \mbox{[} \char`\"{}-\/-\/processes\char`\"{}, \char`\"{}-\/-\/mode\char`\"{}, \char`\"{}-\/-\/nbneratio\char`\"{}, \char`\"{}-\/-\/donbbiasadjust\char`\"{} \mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_ae135a1fb66d7c2ac6a85b5f1a515e918}{}\label{namespacenegui_1_1pgdriveneestimator_ae135a1fb66d7c2ac6a85b5f1a515e918}

\item 
list {\bfseries L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL}
\item 
int {\bfseries I\+D\+X\+\_\+\+G\+E\+N\+E\+P\+O\+P\+\_\+\+F\+I\+L\+ES} = 0\hypertarget{namespacenegui_1_1pgdriveneestimator_ab0e9eab0cba1678dbe8702c8638b22b2}{}\label{namespacenegui_1_1pgdriveneestimator_ab0e9eab0cba1678dbe8702c8638b22b2}

\item 
int {\bfseries I\+D\+X\+\_\+\+S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+ME} = 1\hypertarget{namespacenegui_1_1pgdriveneestimator_a31f8fcb68429a538163ab80b51c4ce8a}{}\label{namespacenegui_1_1pgdriveneestimator_a31f8fcb68429a538163ab80b51c4ce8a}

\item 
int {\bfseries I\+D\+X\+\_\+\+S\+A\+M\+P\+L\+E\+\_\+\+V\+A\+L\+U\+E\+\_\+\+L\+I\+ST} = 2\hypertarget{namespacenegui_1_1pgdriveneestimator_ad1fbe668ed6bb196e7247ae8d6358c92}{}\label{namespacenegui_1_1pgdriveneestimator_ad1fbe668ed6bb196e7247ae8d6358c92}

\item 
int {\bfseries I\+D\+X\+\_\+\+M\+I\+N\+\_\+\+P\+O\+P\+\_\+\+S\+I\+ZE} = 3\hypertarget{namespacenegui_1_1pgdriveneestimator_a0f7b02c2dab3acd91ba217a6cbdef4d8}{}\label{namespacenegui_1_1pgdriveneestimator_a0f7b02c2dab3acd91ba217a6cbdef4d8}

\item 
int {\bfseries I\+D\+X\+\_\+\+M\+A\+X\+\_\+\+P\+O\+P\+\_\+\+S\+I\+ZE} = 4\hypertarget{namespacenegui_1_1pgdriveneestimator_a3a779c3b4f3e5fd8db0224fd3fbdeba7}{}\label{namespacenegui_1_1pgdriveneestimator_a3a779c3b4f3e5fd8db0224fd3fbdeba7}

\item 
int {\bfseries I\+D\+X\+\_\+\+P\+O\+P\+\_\+\+R\+A\+N\+GE} = 5\hypertarget{namespacenegui_1_1pgdriveneestimator_a488e146036273d300b7f528cf22ad017}{}\label{namespacenegui_1_1pgdriveneestimator_a488e146036273d300b7f528cf22ad017}

\item 
int {\bfseries I\+D\+X\+\_\+\+M\+I\+N\+\_\+\+A\+L\+L\+E\+L\+E\+\_\+\+F\+R\+EQ} = 6\hypertarget{namespacenegui_1_1pgdriveneestimator_ac03385eb1adea263a4ac8eb0ba30af74}{}\label{namespacenegui_1_1pgdriveneestimator_ac03385eb1adea263a4ac8eb0ba30af74}

\item 
int {\bfseries I\+D\+X\+\_\+\+R\+E\+P\+L\+I\+C\+A\+T\+ES} = 7\hypertarget{namespacenegui_1_1pgdriveneestimator_a510a047ec241b52adde698c284dd9c01}{}\label{namespacenegui_1_1pgdriveneestimator_a510a047ec241b52adde698c284dd9c01}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+S\+A\+M\+P\+L\+I\+N\+G\+\_\+\+S\+C\+H\+E\+ME} = 8\hypertarget{namespacenegui_1_1pgdriveneestimator_a7261063d4a5ff2ffd7bd7d90d71eeb68}{}\label{namespacenegui_1_1pgdriveneestimator_a7261063d4a5ff2ffd7bd7d90d71eeb68}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+S\+C\+H\+E\+M\+E\+\_\+\+P\+A\+R\+AM} = 9\hypertarget{namespacenegui_1_1pgdriveneestimator_ac1d60673bf39741f3214e383a7e90d8c}{}\label{namespacenegui_1_1pgdriveneestimator_ac1d60673bf39741f3214e383a7e90d8c}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+M\+I\+N\+\_\+\+T\+O\+T\+AL} = 10\hypertarget{namespacenegui_1_1pgdriveneestimator_aafd567651957807eb1b986b1617fe8ee}{}\label{namespacenegui_1_1pgdriveneestimator_aafd567651957807eb1b986b1617fe8ee}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+M\+A\+X\+\_\+\+T\+O\+T\+AL} = 11\hypertarget{namespacenegui_1_1pgdriveneestimator_aa93c875ccfead4ec9c863057783227a6}{}\label{namespacenegui_1_1pgdriveneestimator_aa93c875ccfead4ec9c863057783227a6}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+R\+A\+N\+GE} = 12\hypertarget{namespacenegui_1_1pgdriveneestimator_a7cd2524d305c83a248b30e600cc0b1d8}{}\label{namespacenegui_1_1pgdriveneestimator_a7cd2524d305c83a248b30e600cc0b1d8}

\item 
int {\bfseries I\+D\+X\+\_\+\+L\+O\+C\+I\+\_\+\+S\+A\+M\+P\+L\+E\+\_\+\+R\+E\+P\+L\+I\+C\+A\+T\+ES} = 13\hypertarget{namespacenegui_1_1pgdriveneestimator_a3b479f5812e7feccdbf7edba5468b29e}{}\label{namespacenegui_1_1pgdriveneestimator_a3b479f5812e7feccdbf7edba5468b29e}

\item 
int {\bfseries I\+D\+X\+\_\+\+P\+R\+O\+C\+E\+S\+S\+ES} = 14\hypertarget{namespacenegui_1_1pgdriveneestimator_a7ae0fc5b19216f144e69c43d980891fb}{}\label{namespacenegui_1_1pgdriveneestimator_a7ae0fc5b19216f144e69c43d980891fb}

\item 
int {\bfseries I\+D\+X\+\_\+\+D\+E\+B\+U\+G\+\_\+\+M\+O\+DE} = 15\hypertarget{namespacenegui_1_1pgdriveneestimator_aa97a6de3c44f3552c26677ebda76de39}{}\label{namespacenegui_1_1pgdriveneestimator_aa97a6de3c44f3552c26677ebda76de39}

\item 
int {\bfseries I\+D\+X\+\_\+\+N\+B\+N\+E\+\_\+\+R\+A\+T\+IO} = 16\hypertarget{namespacenegui_1_1pgdriveneestimator_aac916bebf37a7e5e42748fcd7878cdeb}{}\label{namespacenegui_1_1pgdriveneestimator_aac916bebf37a7e5e42748fcd7878cdeb}

\item 
int {\bfseries I\+D\+X\+\_\+\+D\+O\+\_\+\+B\+I\+A\+S\+\_\+\+A\+D\+J\+U\+ST} = 17\hypertarget{namespacenegui_1_1pgdriveneestimator_aedc0eed694d87f02a969994934f884d6}{}\label{namespacenegui_1_1pgdriveneestimator_aedc0eed694d87f02a969994934f884d6}

\item 
int {\bfseries I\+D\+X\+\_\+\+M\+A\+I\+N\+\_\+\+O\+U\+T\+F\+I\+LE} = 18\hypertarget{namespacenegui_1_1pgdriveneestimator_a66f16a5072c0e1dbaacd9e24b6c3eada}{}\label{namespacenegui_1_1pgdriveneestimator_a66f16a5072c0e1dbaacd9e24b6c3eada}

\item 
int {\bfseries I\+D\+X\+\_\+\+S\+E\+C\+O\+N\+D\+A\+R\+Y\+\_\+\+O\+U\+T\+F\+I\+LE} = 19\hypertarget{namespacenegui_1_1pgdriveneestimator_a39d10174aedd519ee696e3ff689d611f}{}\label{namespacenegui_1_1pgdriveneestimator_a39d10174aedd519ee696e3ff689d611f}

\item 
int {\bfseries I\+D\+X\+\_\+\+M\+U\+L\+T\+I\+P\+R\+O\+C\+E\+S\+S\+I\+N\+G\+\_\+\+E\+V\+E\+NT} = 20\hypertarget{namespacenegui_1_1pgdriveneestimator_a372bf29aaa0656694ac089902f5b4a60}{}\label{namespacenegui_1_1pgdriveneestimator_a372bf29aaa0656694ac089902f5b4a60}

\item 
int {\bfseries I\+D\+X\+\_\+\+T\+E\+M\+P\+O\+R\+A\+R\+Y\+\_\+\+D\+I\+R\+E\+C\+T\+O\+RY} = 21\hypertarget{namespacenegui_1_1pgdriveneestimator_af19da5657c153c4214b53f5cad68be3c}{}\label{namespacenegui_1_1pgdriveneestimator_af19da5657c153c4214b53f5cad68be3c}

\item 
int {\bfseries I\+D\+X\+\_\+\+U\+S\+E\+\_\+\+G\+U\+I\+\_\+\+M\+E\+S\+S\+A\+G\+I\+NG} = 22\hypertarget{namespacenegui_1_1pgdriveneestimator_a9f0371fe88418d9b181d7235ca747e73}{}\label{namespacenegui_1_1pgdriveneestimator_a9f0371fe88418d9b181d7235ca747e73}

\item 
{\bfseries I\+D\+X\+\_\+\+L\+A\+S\+T\+\_\+\+C\+O\+N\+S\+O\+L\+E\+\_\+\+A\+RG} = I\+D\+X\+\_\+\+D\+O\+\_\+\+B\+I\+A\+S\+\_\+\+A\+D\+J\+U\+ST\hypertarget{namespacenegui_1_1pgdriveneestimator_a1060cbbc63450dda6b048c88965edadd}{}\label{namespacenegui_1_1pgdriveneestimator_a1060cbbc63450dda6b048c88965edadd}

\item 
list {\bfseries H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS}
\item 
string {\bfseries D\+E\+F\+A\+U\+L\+T\+\_\+\+N\+U\+M\+\_\+\+P\+R\+O\+C\+E\+S\+S\+ES} = \char`\"{}1\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a725197609b22aa42422547cba22744d5}{}\label{namespacenegui_1_1pgdriveneestimator_a725197609b22aa42422547cba22744d5}

\item 
string {\bfseries D\+E\+F\+A\+U\+L\+T\+\_\+\+D\+E\+B\+U\+G\+\_\+\+M\+O\+DE} = \char`\"{}no\+\_\+debug\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a652b7622c1e97a7f7b199266ef9c6f01}{}\label{namespacenegui_1_1pgdriveneestimator_a652b7622c1e97a7f7b199266ef9c6f01}

\item 
string {\bfseries D\+E\+F\+A\+U\+L\+T\+\_\+\+N\+B\+N\+E\+\_\+\+V\+AL} = \char`\"{}None\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a4c995df3f6af3302739c307f3c744bac}{}\label{namespacenegui_1_1pgdriveneestimator_a4c995df3f6af3302739c307f3c744bac}

\item 
int {\bfseries M\+A\+X\+\_\+\+G\+E\+N\+E\+P\+O\+P\+\_\+\+F\+I\+L\+E\+\_\+\+N\+A\+M\+E\+\_\+\+L\+E\+N\+G\+TH} = 310\hypertarget{namespacenegui_1_1pgdriveneestimator_af3d0878b3846d376f070b3c94105ae27}{}\label{namespacenegui_1_1pgdriveneestimator_af3d0878b3846d376f070b3c94105ae27}

\item 
float {\bfseries S\+E\+C\+O\+N\+D\+S\+\_\+\+T\+O\+\_\+\+S\+L\+E\+EP} = 1.\+00\hypertarget{namespacenegui_1_1pgdriveneestimator_a1913968722f6b6ec924ec220421cc2d2}{}\label{namespacenegui_1_1pgdriveneestimator_a1913968722f6b6ec924ec220421cc2d2}

\item 
int {\bfseries P\+O\+O\+L\+\_\+\+B\+A\+T\+C\+H\+\_\+\+S\+I\+ZE} = 100\hypertarget{namespacenegui_1_1pgdriveneestimator_aadadcf84aaf075e62f2a0720ef29ea11}{}\label{namespacenegui_1_1pgdriveneestimator_aadadcf84aaf075e62f2a0720ef29ea11}

\item 
float {\bfseries P\+R\+O\+P\+\_\+\+T\+O\+T\+A\+L\+\_\+\+F\+I\+L\+E\+\_\+\+B\+Y\+T\+E\+S\+\_\+\+U\+S\+E\+D\+\_\+\+B\+Y\+\_\+\+G\+P\+\_\+\+M\+A\+N\+A\+G\+ER} = 0.\+12\hypertarget{namespacenegui_1_1pgdriveneestimator_a89e1e433a1e13de143a0ba9eb5f67c4e}{}\label{namespacenegui_1_1pgdriveneestimator_a89e1e433a1e13de143a0ba9eb5f67c4e}

\item 
int {\bfseries M\+A\+X\+\_\+\+M\+E\+G\+A\+B\+Y\+T\+E\+S\+\_\+\+A\+L\+L\+O\+W\+E\+D\+\_\+\+F\+O\+R\+\_\+\+G\+P\+\_\+\+O\+B\+J\+E\+C\+TS} = 3000\hypertarget{namespacenegui_1_1pgdriveneestimator_ad3cf3c0a3d4e6985fa514fd4a5f31462}{}\label{namespacenegui_1_1pgdriveneestimator_ad3cf3c0a3d4e6985fa514fd4a5f31462}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+N\+O\+NE} = \char`\"{}none\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a4907a8a6c190c39d4a349740c1ba54e5}{}\label{namespacenegui_1_1pgdriveneestimator_a4907a8a6c190c39d4a349740c1ba54e5}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+P\+E\+R\+C\+E\+N\+T\+A\+G\+ES} = \char`\"{}percent\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_ad560d78f911e3ffb994db493a2ab89b2}{}\label{namespacenegui_1_1pgdriveneestimator_ad560d78f911e3ffb994db493a2ab89b2}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+R\+E\+M\+O\+V\+AL} = \char`\"{}remove\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a3fa91319aa849a007dd7cfd4e22213f9}{}\label{namespacenegui_1_1pgdriveneestimator_a3fa91319aa849a007dd7cfd4e22213f9}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+C\+R\+I\+T\+E\+R\+IA} = \char`\"{}criteria\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a3732dae3a6f631c08a94fd98ca308e84}{}\label{namespacenegui_1_1pgdriveneestimator_a3732dae3a6f631c08a94fd98ca308e84}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+C\+O\+H\+O\+RT} = \char`\"{}cohorts\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_ab5cd7cf71765d0fd1ace691d10631239}{}\label{namespacenegui_1_1pgdriveneestimator_ab5cd7cf71765d0fd1ace691d10631239}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+C\+O\+H\+O\+R\+T\+\_\+\+P\+E\+R\+C\+E\+N\+T\+A\+GE} = \char`\"{}cohortsperc\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_abf385f828ea73fe77218cc695ed31ca6}{}\label{namespacenegui_1_1pgdriveneestimator_abf385f828ea73fe77218cc695ed31ca6}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+R\+E\+L\+A\+T\+E\+DS} = \char`\"{}relateds\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a5867272ead84c59e21ac0db9bf0a08f7}{}\label{namespacenegui_1_1pgdriveneestimator_a5867272ead84c59e21ac0db9bf0a08f7}

\item 
list {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA}
\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+L\+O\+C\+I\+\_\+\+S\+C\+H\+E\+M\+E\+\_\+\+N\+O\+NE} = \char`\"{}none\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8dfd8edb967f95bb041e347135a33547}{}\label{namespacenegui_1_1pgdriveneestimator_a8dfd8edb967f95bb041e347135a33547}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+L\+O\+C\+I\+\_\+\+S\+C\+H\+E\+M\+E\+\_\+\+P\+E\+R\+C\+E\+NT} = \char`\"{}percent\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a142c5458e517f1c3bcfcc990b04bb025}{}\label{namespacenegui_1_1pgdriveneestimator_a142c5458e517f1c3bcfcc990b04bb025}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+L\+O\+C\+I\+\_\+\+C\+O\+N\+S\+T\+A\+N\+T\+\_\+\+T\+O\+T\+AL} = \char`\"{}total\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_af0086ffbe3400aa8ef6234dd375701bd}{}\label{namespacenegui_1_1pgdriveneestimator_af0086ffbe3400aa8ef6234dd375701bd}

\item 
string {\bfseries O\+U\+T\+P\+U\+T\+\_\+\+D\+E\+L\+I\+M\+I\+T\+ER} = \char`\"{}\textbackslash{}t\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_abcd9551489a7ca0281f03b3b0236e8e2}{}\label{namespacenegui_1_1pgdriveneestimator_abcd9551489a7ca0281f03b3b0236e8e2}

\item 
string {\bfseries E\+N\+D\+L\+I\+N\+E\+\_\+\+S\+EQ} = \char`\"{}\textbackslash{}n\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a5de353f2cfacdd1ffb10c1eae6ccdd8c}{}\label{namespacenegui_1_1pgdriveneestimator_a5de353f2cfacdd1ffb10c1eae6ccdd8c}

\item 
string {\bfseries O\+U\+T\+P\+U\+T\+\_\+\+E\+N\+D\+L\+I\+NE} = \char`\"{}\textbackslash{}n\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_aa40c91635517374f437b206fa5c12070}{}\label{namespacenegui_1_1pgdriveneestimator_aa40c91635517374f437b206fa5c12070}

\item 
string {\bfseries R\+E\+P\+L\+I\+C\+A\+T\+E\+\_\+\+D\+E\+L\+I\+M\+I\+T\+ER} = \char`\"{},\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a1193fcffbc197c97223fc41a886901f5}{}\label{namespacenegui_1_1pgdriveneestimator_a1193fcffbc197c97223fc41a886901f5}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+\_\+\+P\+A\+R\+A\+M\+\_\+\+D\+E\+L\+I\+M\+I\+T\+ER} = \char`\"{},\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_afb91e259587f0231f20d658ce5cc6039}{}\label{namespacenegui_1_1pgdriveneestimator_afb91e259587f0231f20d658ce5cc6039}

\item 
list {\bfseries N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP} = \mbox{[} \char`\"{}case\+\_\+number\char`\"{} \mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_ae8a64dd815aff6a97f4651a4525c449e}{}\label{namespacenegui_1_1pgdriveneestimator_ae8a64dd815aff6a97f4651a4525c449e}

\item 
list {\bfseries N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+A\+N\+D\+\_\+\+L\+D\+N\+E\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+T\+A\+GS} = \mbox{[} \char`\"{}No\+Dat.\+txt\char`\"{}, \char`\"{}x\+Tp.\+txt\char`\"{}, \char`\"{}ne\+\_\+run.\+txt\char`\"{}, \char`\"{}x.\+txt\char`\"{} \mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_a1849dccaf78707bab233fa6a57a29fe0}{}\label{namespacenegui_1_1pgdriveneestimator_a1849dccaf78707bab233fa6a57a29fe0}

\item 
string {\bfseries N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+A\+N\+D\+\_\+\+L\+D\+N\+E\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+L\+E\+S\+\_\+\+G\+E\+N\+E\+R\+A\+L\+\_\+\+T\+AG} = \char`\"{}\+\_\+g\+\_\+\mbox{[}0-\/9\mbox{]}\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a2476a4171478c9b09a27a1287121a7cc}{}\label{namespacenegui_1_1pgdriveneestimator_a2476a4171478c9b09a27a1287121a7cc}

\item 
string {\bfseries G\+L\+O\+B\+\_\+\+I\+N\+T\+E\+R\+M\+E\+D\+I\+A\+T\+E\+\_\+\+G\+E\+N\+E\+P\+O\+P\+\_\+\+F\+I\+L\+E\+\_\+\+T\+A\+GS} = \char`\"{}$\ast$\+\_\+r\+\_\+$\ast$\+\_\+g\+\_\+$\ast$\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_ae1d8c4de97e82bc6f70762621e961790}{}\label{namespacenegui_1_1pgdriveneestimator_ae1d8c4de97e82bc6f70762621e961790}

\item 
string {\bfseries I\+N\+P\+U\+T\+\_\+\+F\+I\+L\+E\+\_\+\+D\+O\+T\+\_\+\+C\+H\+A\+R\+\_\+\+R\+E\+P\+L\+A\+C\+E\+M\+E\+NT} = \char`\"{}\+\_\+\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_aefc749d7448354bfc67a0db4e559934e}{}\label{namespacenegui_1_1pgdriveneestimator_aefc749d7448354bfc67a0db4e559934e}

\item 
list {\bfseries M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS}
\item 
string {\bfseries C\+O\+L\+\_\+\+N\+A\+M\+E\+\_\+\+N\+B\+N\+E\+\_\+\+R\+A\+T\+I\+O\+\_\+\+F\+O\+R\+\_\+\+B\+I\+A\+S\+\_\+\+A\+D\+J\+U\+ST} = \char`\"{}nbne\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a32fa9f2d3c5f9ddea3310478836c4be3}{}\label{namespacenegui_1_1pgdriveneestimator_a32fa9f2d3c5f9ddea3310478836c4be3}

\item 
string {\bfseries C\+O\+L\+\_\+\+N\+A\+M\+E\+\_\+\+B\+I\+A\+S\+\_\+\+A\+D\+J\+U\+S\+T\+E\+D\+\_\+\+L\+D\+NE} = \char`\"{}ne\+\_\+est\+\_\+adj\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a7b3f5bb6283c6e9d280f6be70e7a594f}{}\label{namespacenegui_1_1pgdriveneestimator_a7b3f5bb6283c6e9d280f6be70e7a594f}

\item 
string {\bfseries C\+O\+L\+\_\+\+N\+A\+M\+E\+\_\+\+N\+E\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+TE} = \char`\"{}est\+\_\+ne\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a25446e91c2df65da0b18fa46f83eec48}{}\label{namespacenegui_1_1pgdriveneestimator_a25446e91c2df65da0b18fa46f83eec48}

\item 
int {\bfseries R\+E\+L\+T\+OL} = 1\hypertarget{namespacenegui_1_1pgdriveneestimator_a1d5b194ec442b286efc83303817a8123}{}\label{namespacenegui_1_1pgdriveneestimator_a1d5b194ec442b286efc83303817a8123}

\item 
{\bfseries o\+\_\+parser} = ap.\+Argument\+Parser()\hypertarget{namespacenegui_1_1pgdriveneestimator_a5d6d3756a8ddff6ece719654fb970636}{}\label{namespacenegui_1_1pgdriveneestimator_a5d6d3756a8ddff6ece719654fb970636}

\item 
{\bfseries o\+\_\+arglist} = o\+\_\+parser.\+add\+\_\+argument\+\_\+group( \char`\"{}args\char`\"{} )\hypertarget{namespacenegui_1_1pgdriveneestimator_a22a84c6d9023007cf69fd2e3d62dcb86}{}\label{namespacenegui_1_1pgdriveneestimator_a22a84c6d9023007cf69fd2e3d62dcb86}

\item 
{\bfseries i\+\_\+total\+\_\+nonopt} = len( L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+S\+H\+O\+R\+T\+\_\+\+R\+E\+Q\+U\+I\+R\+ED )\hypertarget{namespacenegui_1_1pgdriveneestimator_aa4e6982ff99236ac13e0af4c4d6e8839}{}\label{namespacenegui_1_1pgdriveneestimator_aa4e6982ff99236ac13e0af4c4d6e8839}

\item 
{\bfseries help}\hypertarget{namespacenegui_1_1pgdriveneestimator_ad407ca10f30961eabc0630c21fe8da45}{}\label{namespacenegui_1_1pgdriveneestimator_ad407ca10f30961eabc0630c21fe8da45}

\item 
{\bfseries required}\hypertarget{namespacenegui_1_1pgdriveneestimator_ab5c3d005ba8659dc07a5f3171c7214df}{}\label{namespacenegui_1_1pgdriveneestimator_ab5c3d005ba8659dc07a5f3171c7214df}

\item 
{\bfseries i\+\_\+total\+\_\+opt} = len( L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+S\+H\+O\+R\+T\+\_\+\+O\+P\+T\+I\+O\+N\+AL )\hypertarget{namespacenegui_1_1pgdriveneestimator_a3c6b1b9d9500e05eca62433ac13c8692}{}\label{namespacenegui_1_1pgdriveneestimator_a3c6b1b9d9500e05eca62433ac13c8692}

\item 
{\bfseries o\+\_\+args} = o\+\_\+parser.\+parse\+\_\+args()\hypertarget{namespacenegui_1_1pgdriveneestimator_abff5b751b69197ab0c78f53cf34e3a05}{}\label{namespacenegui_1_1pgdriveneestimator_abff5b751b69197ab0c78f53cf34e3a05}

\item 
list {\bfseries ls\+\_\+args\+\_\+passed} = \mbox{[}$\,$\mbox{]}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8bd374d85a2f54f7d6086c3560fba734}{}\label{namespacenegui_1_1pgdriveneestimator_a8bd374d85a2f54f7d6086c3560fba734}

\item 
{\bfseries s\+\_\+argname} = s\+\_\+flag.\+replace( \char`\"{}-\/-\/\char`\"{}, \char`\"{}\char`\"{} )\hypertarget{namespacenegui_1_1pgdriveneestimator_af8fac37bc1806ab9f054b8c4754e669e}{}\label{namespacenegui_1_1pgdriveneestimator_af8fac37bc1806ab9f054b8c4754e669e}

\item 
{\bfseries s\+\_\+val} = getattr( o\+\_\+args, s\+\_\+argname )\hypertarget{namespacenegui_1_1pgdriveneestimator_a801907b0dbfc713d2a817781f2b3c7d4}{}\label{namespacenegui_1_1pgdriveneestimator_a801907b0dbfc713d2a817781f2b3c7d4}

\item 
{\bfseries s\+\_\+temp\+\_\+dir} = pgut.\+get\+\_\+temporary\+\_\+directory()\hypertarget{namespacenegui_1_1pgdriveneestimator_a660fa9f1c3276022986893efbe9a3c76}{}\label{namespacenegui_1_1pgdriveneestimator_a660fa9f1c3276022986893efbe9a3c76}

\item 
string {\bfseries s\+\_\+msg} = \char`\"{}In pgdriveneestimator main section, \char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a167b017a9b7a41a2002208686cc6955e}{}\label{namespacenegui_1_1pgdriveneestimator_a167b017a9b7a41a2002208686cc6955e}

\item 
float {\bfseries P\+R\+O\+P\+O\+R\+T\+I\+O\+N\+\_\+\+F\+I\+L\+E\+\_\+\+B\+Y\+T\+E\+S\+\_\+\+U\+S\+E\+D\+\_\+\+B\+Y\+\_\+\+O\+B\+J\+E\+CT} = 0.\+12\hypertarget{namespacenegui_1_1pgdriveneestimator_aac0c88c001b2a957689cba925d9de85a}{}\label{namespacenegui_1_1pgdriveneestimator_aac0c88c001b2a957689cba925d9de85a}

\item 
float {\bfseries P\+R\+O\+P\+O\+R\+T\+I\+O\+N\+\_\+\+G\+P\+\_\+\+O\+B\+J\+E\+C\+T\+S\+\_\+\+P\+E\+R\+\_\+\+A\+D\+D\+I\+T\+I\+O\+N\+A\+L\+\_\+\+P\+R\+O\+C\+E\+SS} = 0.\+3\hypertarget{namespacenegui_1_1pgdriveneestimator_a257f80975601ec1f262e61da94aab595}{}\label{namespacenegui_1_1pgdriveneestimator_a257f80975601ec1f262e61da94aab595}

\item 
float {\bfseries P\+E\+R\+C\+\_\+\+A\+V\+A\+I\+L\+\_\+\+R\+A\+M\+\_\+\+T\+O\+\_\+\+A\+C\+C\+E\+SS} = 0.\+7\hypertarget{namespacenegui_1_1pgdriveneestimator_a20fbee9f220ed1031975161c1f2a679e}{}\label{namespacenegui_1_1pgdriveneestimator_a20fbee9f220ed1031975161c1f2a679e}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+C\+O\+H\+O\+R\+T\+\_\+\+C\+O\+U\+NT} = \char`\"{}cohortscount\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a40ebaca8593fd1dbd1ad94f1718f4e28}{}\label{namespacenegui_1_1pgdriveneestimator_a40ebaca8593fd1dbd1ad94f1718f4e28}

\item 
string {\bfseries S\+A\+M\+P\+L\+E\+\_\+\+B\+Y\+\_\+\+C\+O\+H\+O\+R\+T\+\_\+\+C\+O\+U\+N\+T\+\_\+\+M\+AX} = \char`\"{}cohortsmax\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_ae8fc9fdec47a0df16a055fdf4a9f693f}{}\label{namespacenegui_1_1pgdriveneestimator_ae8fc9fdec47a0df16a055fdf4a9f693f}

\item 
string {\bfseries C\+O\+L\+\_\+\+N\+A\+M\+E\+\_\+\+H\+E\+T\+\_\+\+V\+A\+L\+UE} = \char`\"{}mean\+\_\+het\char`\"{}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8dcfc49c319774f0b7850d01d806781b}{}\label{namespacenegui_1_1pgdriveneestimator_a8dcfc49c319774f0b7850d01d806781b}

\end{DoxyCompactItemize}


\subsection{Detailed Description}
\begin{DoxyVerb}Description

given 
    -->a list of genepop files, 
    -->a list of percentages of the total indiv
       to subsample
    --> a number giving a minimum pop size 
       after subsampling
    -->a minimum allele frequency (single float)
    --> a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 

    subsamples and then runs the Ne estimator for each combination of 
    genepop_file,percentage,replicate_number. 
    
    As of Fri May 13 19:01:06 MDT 2016, still not sure
    how best to get the results to file or stream.

Mon Aug 15 20:04:00 MDT 2016
    revised so that now args list includes
    -->a list of genepop files, 
    -->one of [percent|remove] (names a subsampling scheme, to be enlarged)
    -->a list of percentages, for percent samplin, or N-values for removing N
from the individuals    
    --> a number giving a minimum pop size 
       after subsampling (fewer than the min count of indiv, and the pop is skipped)
    -->a minimum allele frequency (single float)
    --> a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 
    -->optional the number of processes to create in performing the
estimates in parallele, on process for each population (as divided up inside a file,
or as given by a signle genepop file with only one population listed)
    
    Also revised to allow the improtation of this mod, so that the same operations
    can be intitiated by a call to "mymain".  The importation method allows for 
    passing in two file objects, open for writing.  The console method, that satisfies
    "if __name__== "__main__", does not allow the file object args and prints to stdout and stderr.

Wed Sep  7 18:12:26 MDT 2016
    revised so that now args list includes
    -->a list of genepop files, 
    -->one of [percent|remove|criteria] (names a subsampling scheme, to be enlarged)
    -->a list of comma-delimited percentages, for percent samplin, or N-values for removing N
from the individuals, or, for criteria, a comma-delimited list consisting of:
        1. list of semicolong-delimited genepop indiv id fields (example "numberage;sex;parent")
        2. list of python types corresponding to each field (example "int;int;string")
        3. one or more test expressions using field names 
            surrounded by percentage signs.  For example, to include 
            2 criteria: "%age%>3 and %age%<6 and %sex% == 3","%parent% <= "3992"
    --> a number giving a minimum pop size 
       after subsampling (fewer than the min count of indiv, and the pop is skipped)
    -->a hypnated pair of numbers giving the minimum and maximum i where
i gives the ith population listed in any one of the genepop files. For
example to run the estimator only on the 2nd through the 10th population,
the arg would be: "2-10" (all subject to minimum pop size).  
Using "all" for this arg will result in all pops being included.
    -->a minimum allele frequency (single float)
    --> a number giving the total replicate NeEstimator runs 
to be executed on each file at each subsample percentage 
    -->optional the number of processes to create in performing the
estimates in parallele, on process for each population (as divided up inside a file,
or as given by a signle genepop file with only one population listed)

2016_09_24
    Revisions to add new sampling schemes (No-subsampling, and AgeStructureNe schemes to get
    cohorts and siblings (Relateds).  In addition to the percent, remove, and
    criteria schemes, we add "none", "cohorts", and "relateds" as options for arg 2.
    For arg 3 we now add, if cohorts is the second arg value, comma-delimited
    fields (see mod genepopfilesampler.py, class 
    GenepopFileSamplerParamsAgeStructureCohorts):
    --> int, max individual age
    And, if "relateds" is the 2nd arg value (see mod genepopfilesampler.py, class
    GenepopFileSamplerIndividualsAgeStructureRelateds):
    1. float, percent of relateds each gen
    Also added max pop size as new arg, now arg #
2016_10_20
    Adding to Loci scheme "percent," to subsample a percentage
    of the loci (as circumscribed by existing params min loci 
    position and max loci position.
\end{DoxyVerb}
 

\subsection{Function Documentation}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate@{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate}}
\index{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate@{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate(o\+\_\+genepopfile, s\+\_\+sample\+\_\+scheme, f\+\_\+min\+\_\+allele\+\_\+freq, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, o\+\_\+debug\+\_\+mode, llv\+\_\+args\+\_\+each\+\_\+process, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+I\+P, o\+\_\+secondary\+\_\+outfile, f\+\_\+nbne\+\_\+ratio, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count)}{add_to_set_of_calls_to_do_estimate(o_genepopfile, s_sample_scheme, f_min_allele_freq, i_min_pop_size, i_max_pop_size, o_debug_mode, llv_args_each_process, IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, o_secondary_outfile, f_nbne_ratio, s_temporary_directory, i_genepop_file_count)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{s\+\_\+sample\+\_\+scheme, }
\item[{}]{f\+\_\+min\+\_\+allele\+\_\+freq, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+size, }
\item[{}]{o\+\_\+debug\+\_\+mode, }
\item[{}]{llv\+\_\+args\+\_\+each\+\_\+process, }
\item[{}]{I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, }
\item[{}]{o\+\_\+secondary\+\_\+outfile, }
\item[{}]{f\+\_\+nbne\+\_\+ratio, }
\item[{}]{s\+\_\+temporary\+\_\+directory, }
\item[{}]{i\+\_\+genepop\+\_\+file\+\_\+count}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a0f589286698490f5054d972daa7c9160}{}\label{namespacenegui_1_1pgdriveneestimator_a0f589286698490f5054d972daa7c9160}
\begin{DoxyVerb}This def creates ne-estimator caller object and adds it to list of args for a single call
to def do_estimate.  The call is then appended to llv_args_each_process
\end{DoxyVerb}
 

Definition at line 2041 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
2041                                             i\_genepop\_file\_count ):
2042     \textcolor{stringliteral}{'''     }
2043 \textcolor{stringliteral}{    This def creates ne-estimator caller object and adds it to list of args for a single call}
2044 \textcolor{stringliteral}{    to def do\_estimate.  The call is then appended to llv\_args\_each\_process}
2045 \textcolor{stringliteral}{    '''}
2046 
2047     \textcolor{comment}{#all these GenepopFileManager objects should now have}
2048     \textcolor{comment}{#a single population sample name:}
2049     s\_population\_sample\_name=o\_genepopfile.population\_subsample\_tags[ 0 ]
2050 
2051     ls\_indiv\_sample\_names=o\_genepopfile.indiv\_subsample\_tags
2052 
2053     li\_population\_numbers=o\_genepopfile.getListPopulationNumbers( s\_population\_sample\_name )
2054     \textcolor{comment}{#for each sample (i.e. for each replicate of a sampling with a given set of params:}
2055 
2056     \textcolor{comment}{#Added this to help create a unique, temporary gp file name}
2057     \textcolor{comment}{#using file number individ sample, and loci sample counts:}
2058     i\_individ\_sample\_count=0
2059 
2060     \textcolor{keywordflow}{for} s\_indiv\_sample \textcolor{keywordflow}{in} ls\_indiv\_sample\_names:
2061 
2062         i\_individ\_sample\_count+=1
2063 
2064         \textcolor{comment}{#we also check that at least one pop has at least one individual}
2065         \textcolor{comment}{#listed for this subsample}
2066         li\_indiv\_counts\_for\_this\_sample=o\_genepopfile.getIndividualCounts( s\_indiv\_subsample\_tag=
      s\_indiv\_sample )
2067 
2068         \textcolor{keywordflow}{if} sum( li\_indiv\_counts\_for\_this\_sample ) == 0:
2069             s\_sample\_value, i\_replicate\_count=get\_sample\_val\_and\_rep\_number\_from\_sample\_name( \(\backslash\)
2070                                             s\_indiv\_sample, s\_sample\_scheme )
2071             \textcolor{comment}{#we only want to report the first replicate}
2072             \textcolor{comment}{#(by inference of course all replicates will be skipped):}
2073             \textcolor{keywordflow}{if} i\_replicate\_count == 0:
2074                 s\_msg=\textcolor{stringliteral}{"Skipping samples from: "} \(\backslash\)
2075                         + o\_genepopfile.original\_file\_name \(\backslash\)
2076                         + \textcolor{stringliteral}{".  In pgdriveneestimator.py, all populations "} \(\backslash\)
2077                         + \textcolor{stringliteral}{"have zero individuals when sampled using "} \(\backslash\)
2078                         + \textcolor{stringliteral}{"scheme: "} + s\_sample\_scheme \(\backslash\)
2079                         + \textcolor{stringliteral}{", and with sample parameter, "} + s\_sample\_value + \textcolor{stringliteral}{"."}
2080                 o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2081             \textcolor{comment}{#end if first replicate, report}
2082             \textcolor{keywordflow}{continue}
2083         \textcolor{comment}{#end skip this subsample if no pops have any individuals}
2084 
2085         \textcolor{stringliteral}{'''}
2086 \textcolor{stringliteral}{        For percent and remove-N schemes, the min/max pop size params are not }
2087 \textcolor{stringliteral}{        enforced in the sampler, and, as of 2016\_09\_27, are to be applied}
2088 \textcolor{stringliteral}{        before sampling.  Thus, for these scheme we skip populations if their }
2089 \textcolor{stringliteral}{        original, un-subsampled size is outside the pop size range.  Other schemes}
2090 \textcolor{stringliteral}{        should have applied the criteria when subsampling.  For the latter, this}
2091 \textcolor{stringliteral}{        would not be an appropriate test::}
2092 \textcolor{stringliteral}{        '''}
2093 
2094         li\_pops\_with\_invalid\_size=[]
2095         
2096         \textcolor{keywordflow}{if} s\_sample\_scheme \textcolor{keywordflow}{in} SAMPLE\_SCHEMES\_NON\_CRITERIA:
2097             li\_pops\_with\_invalid\_size=\hyperlink{namespacenegui_1_1pgdriveneestimator_a06645f9b380f736368e27585331ec22f}{get\_pops\_in\_file\_outside\_valid\_size\_range}
      ( o\_genepopfile, 
2098                                                                                 i\_min\_pop\_size,
2099                                                                                 i\_max\_pop\_size )
2100         \textcolor{comment}{#end if we need to check original (non-sampled) pop sizes}
2101 
2102         ls\_loci\_subsample\_tags\_associated\_with\_this\_indiv\_sample= \(\backslash\)
2103                 get\_loci\_subsample\_tags\_for\_this\_indiv\_subsample( o\_genepopfile,
2104                                                                     s\_indiv\_sample )
2105     
2106         i\_loci\_subsample\_count=0
2107 
2108         \textcolor{keywordflow}{for} s\_loci\_subsample\_tag \textcolor{keywordflow}{in} ls\_loci\_subsample\_tags\_associated\_with\_this\_indiv\_sample:
2109 
2110             \textcolor{stringliteral}{'''}
2111 \textcolor{stringliteral}{            To most-efficiently use multi-processes, that is, to assign many}
2112 \textcolor{stringliteral}{            processes a single file with many populations, we do an}
2113 \textcolor{stringliteral}{            estimate on each pop in the file separately, and divvy}
2114 \textcolor{stringliteral}{            the pops up among processes:}
2115 \textcolor{stringliteral}{            '''}
2116             \textcolor{keywordflow}{for} i\_population\_number \textcolor{keywordflow}{in} li\_population\_numbers:
2117 
2118                 i\_loci\_subsample\_count+=1
2119 
2120                 \textcolor{keywordflow}{if} i\_population\_number \textcolor{keywordflow}{in} li\_pops\_with\_invalid\_size:
2121                     s\_msg= \textcolor{stringliteral}{"In pgdriveneestimator.py, def add\_to\_set\_of\_calls\_to\_do\_estimate, "} \(\backslash\)
2122                                     + \textcolor{stringliteral}{" in file, "} + o\_genepopfile.original\_file\_name \(\backslash\)
2123                                     + \textcolor{stringliteral}{", skipping pop number, "}  + str( i\_population\_number ) \(\backslash\)
2124                                     + \textcolor{stringliteral}{".  It has an indiv count outside of valid "} \(\backslash\)
2125                                     + \textcolor{stringliteral}{"range, "} + str( i\_min\_pop\_size ) \(\backslash\)
2126                                     + \textcolor{stringliteral}{" - "} + str( i\_max\_pop\_size ) + \textcolor{stringliteral}{"."}
2127                     o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2128                     \textcolor{keywordflow}{continue}
2129                 \textcolor{comment}{#end if i\_population\_number is}
2130 
2131                 \textcolor{stringliteral}{'''}
2132 \textcolor{stringliteral}{                We send to def do\_estimate the genepop file object, o\_genepopfile, and the subsample tag}
2133 \textcolor{stringliteral}{                that identifies the correct subsample calculated and stored in the object, so we can write}
2134 \textcolor{stringliteral}{                each subsample genepop file in the same process that will use it as input to the the
       estimator.}
2135 \textcolor{stringliteral}{                This should limit the number of input files existing concurrenty in the directory to the }
2136 \textcolor{stringliteral}{                number of processes in use.  We also send allele freq, and the original genepop file name }
2137 \textcolor{stringliteral}{                inside the genepop file manager object):}
2138 \textcolor{stringliteral}{                '''}
2139                 s\_sample\_value, s\_replicate\_number = \(\backslash\)
2140                         get\_sample\_val\_and\_rep\_number\_from\_sample\_name( s\_indiv\_sample, s\_sample\_scheme )
2141 
2142                 s\_loci\_sample\_value, s\_loci\_replicate\_number = \(\backslash\)
2143                         \hyperlink{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}{get\_loci\_sample\_val\_and\_rep\_number\_from\_loci\_sample\_tag}
      ( s\_loci\_subsample\_tag )
2144                 
2145 
2146                 i\_tot\_indivs\_this\_subsample=o\_genepopfile.getIndividualCount( i\_population\_number, 
2147                                                                                         s\_indiv\_sample )
2148 
2149                 \textcolor{keywordflow}{if} i\_tot\_indivs\_this\_subsample == 0:
2150                     s\_msg= \textcolor{stringliteral}{"In pgdriveneestimator.py, def add\_to\_set\_of\_calls\_to\_do\_estimate, "} \(\backslash\)
2151                                     + \textcolor{stringliteral}{" in file, "} + o\_genepopfile.original\_file\_name \(\backslash\)
2152                                     + \textcolor{stringliteral}{", with sample value: "} + s\_sample\_value \(\backslash\)
2153                                     + \textcolor{stringliteral}{", and loci sample value: "} + s\_loci\_sample\_value \(\backslash\)
2154                                     + \textcolor{stringliteral}{", skipping pop number, "}  + str( i\_population\_number ) \(\backslash\)
2155                                     + \textcolor{stringliteral}{".  It has a zero indiv count."}
2156                     o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2157                     \textcolor{keywordflow}{continue}
2158                 \textcolor{comment}{# end if this sample for this pop has zero individuals}
2159 
2160                 \textcolor{comment}{#for downstream analysis, we get the census of total indiv}
2161                 \textcolor{comment}{#non-sampled, for this pop section:}
2162                 s\_census=str( o\_genepopfile.getIndividualCount( i\_population\_number ) )
2163 
2164                 \textcolor{comment}{#we make population subsample list consisting of only this populations number:}
2165                 s\_this\_pop\_number=str( i\_population\_number ) 
2166 
2167                 o\_genepopfile.subsamplePopulationsByList( [ i\_population\_number ], s\_this\_pop\_number )
2168 
2169                 \textcolor{keywordflow}{if} o\_debug\_mode.isSet( DebugMode.PRINT\_REPLICATE\_SELECTIONS ):
2170                     print\_test\_list\_replicate\_selection\_indices( o\_genepopfile, 
2171                                             s\_sample\_scheme,    
2172                                             s\_indiv\_sample, 
2173                                             s\_loci\_sample\_value,
2174                                             s\_loci\_replicate\_number,
2175                                             o\_secondary\_outfile, 
2176                                             s\_population\_subsample\_tag=s\_this\_pop\_number )
2177                 \textcolor{comment}{#end if we're testing, print }
2178 
2179                 \textcolor{stringliteral}{'''}
2180 \textcolor{stringliteral}{                2017\_04\_03.  A Windows limit on path length motivates new, shorter names}
2181 \textcolor{stringliteral}{                for the intermediate genepop files.  Now we simply use the i\_genepop\_file\_count}
2182 \textcolor{stringliteral}{                number passed to this def from def drive\_estimator, and the subsample count}
2183 \textcolor{stringliteral}{                to get a uniq identifier for this genepop file vs. all others in this run.}
2184 \textcolor{stringliteral}{                For now we retain the original code remm'd out, and construct the shorter}
2185 \textcolor{stringliteral}{                name in a new def:}
2186 \textcolor{stringliteral}{                '''}
2187                 \textcolor{comment}{#in naming a the genepop file for the subsampled pop,}
2188                 \textcolor{comment}{#some processing is needed to avoid NeEstimator file naming errors, and}
2189                 \textcolor{comment}{#path mangling, so we made a separate def:}
2190 \textcolor{comment}{#               s\_genepop\_file\_subsample=get\_subsample\_genepop\_file\_name( o\_genepopfile.original\_file\_name,
       }
2191 \textcolor{comment}{#                                                       s\_loci\_subsample\_tag, }
2192 \textcolor{comment}{#                                                       s\_this\_pop\_number,}
2193 \textcolor{comment}{#                                                       s\_temporary\_directory )}
2194                 
2195                 s\_genepop\_file\_subsample=\hyperlink{namespacenegui_1_1pgdriveneestimator_a9c650be4ed8ae5fa5e599523c22273bd}{make\_subsample\_genepop\_file\_name}( 
      o\_genepopfile.original\_file\_name,
2196                                                                                             
      s\_temporary\_directory, 
2197                                                                                             
      i\_genepop\_file\_count, 
2198                                                                                                 
      i\_individ\_sample\_count,
2199                                                                                                 
      i\_loci\_subsample\_count )
2200 
2201                 s\_run\_output\_file=s\_genepop\_file\_subsample + \textcolor{stringliteral}{"\_ne\_run.txt"}
2202 
2203                 o\_neinput=pgin.PGInputNeEstimator( s\_genepop\_file\_subsample )
2204 
2205                 o\_neinput.run\_params=\{ \textcolor{stringliteral}{"crits"}:[ f\_min\_allele\_freq ]  \}
2206 
2207                 o\_neoutput=pgout.PGOutputNeEstimator( s\_genepop\_file\_subsample, 
2208                                                         s\_run\_output\_file, 
2209                                                         s\_estimator\_to\_use=ESTIMATOR\_TO\_USE )   
2210                 
2211                 o\_ne\_estimator=pgne.PGOpNeEstimator( o\_neinput, 
2212                                                         o\_neoutput, 
2213                                                         s\_estimator\_name=ESTIMATOR\_TO\_USE,
2214                                                         s\_parent\_dir\_for\_workspace=s\_temporary\_directory ) 
2215                 
2216                 \textcolor{keywordflow}{if} ESTIMATOR\_TO\_USE == LDNE2 \textcolor{keywordflow}{and} PATH\_TO\_LDNE2 \textcolor{keywordflow}{is} \textcolor{keywordflow}{not} \textcolor{keywordtype}{None}:
2217                     o\_ne\_estimator.ldne\_path=PATH\_TO\_LDNE2
2218                 \textcolor{comment}{#end if we are using LDNe2 and we have a path to the executable}
2219 
2220                 lv\_these\_args = [ o\_genepopfile,  
2221                                     o\_ne\_estimator, 
2222                                     s\_sample\_value, 
2223                                     s\_loci\_sample\_value,
2224                                     f\_min\_allele\_freq, 
2225                                     s\_indiv\_sample, 
2226                                     s\_loci\_subsample\_tag,
2227                                     s\_this\_pop\_number, 
2228                                     s\_census,
2229                                     s\_replicate\_number, 
2230                                     s\_loci\_replicate\_number,
2231                                     o\_debug\_mode,
2232                                     IDX\_NE\_ESTIMATOR\_OUTPUT\_FIELDS\_TO\_SKIP,
2233                                     f\_nbne\_ratio ]
2234 
2235                 llv\_args\_each\_process.append( lv\_these\_args )
2236             \textcolor{comment}{#end for each population}
2237         \textcolor{comment}{#end for each loci sample name assoc with this indiv sample name}
2238     \textcolor{comment}{#end for each indiv sample name}
2239     \textcolor{keywordflow}{return}
2240 \textcolor{comment}{#end add\_to\_set\_of\_calls\_to\_do\_estimate}
2241 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate@{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate}}
\index{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate@{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate(o\+\_\+genepopfile, s\+\_\+sample\+\_\+scheme, f\+\_\+min\+\_\+allele\+\_\+freq, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, o\+\_\+debug\+\_\+mode, llv\+\_\+args\+\_\+each\+\_\+process, I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+I\+P, o\+\_\+secondary\+\_\+outfile, f\+\_\+nbne\+\_\+ratio, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count)}{add_to_set_of_calls_to_do_estimate(o_genepopfile, s_sample_scheme, f_min_allele_freq, i_min_pop_size, i_max_pop_size, i_min_loci_position, i_max_loci_position, o_debug_mode, llv_args_each_process, IDX_NE_ESTIMATOR_OUTPUT_FIELDS_TO_SKIP, o_secondary_outfile, f_nbne_ratio, s_temporary_directory, i_genepop_file_count)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+add\+\_\+to\+\_\+set\+\_\+of\+\_\+calls\+\_\+to\+\_\+do\+\_\+estimate (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{s\+\_\+sample\+\_\+scheme, }
\item[{}]{f\+\_\+min\+\_\+allele\+\_\+freq, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+min\+\_\+loci\+\_\+position, }
\item[{}]{i\+\_\+max\+\_\+loci\+\_\+position, }
\item[{}]{o\+\_\+debug\+\_\+mode, }
\item[{}]{llv\+\_\+args\+\_\+each\+\_\+process, }
\item[{}]{I\+D\+X\+\_\+\+N\+E\+\_\+\+E\+S\+T\+I\+M\+A\+T\+O\+R\+\_\+\+O\+U\+T\+P\+U\+T\+\_\+\+F\+I\+E\+L\+D\+S\+\_\+\+T\+O\+\_\+\+S\+K\+IP, }
\item[{}]{o\+\_\+secondary\+\_\+outfile, }
\item[{}]{f\+\_\+nbne\+\_\+ratio, }
\item[{}]{s\+\_\+temporary\+\_\+directory, }
\item[{}]{i\+\_\+genepop\+\_\+file\+\_\+count}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a4d13e12b51231deb46355be190102cbc}{}\label{namespacenegui_1_1pgdriveneestimator_a4d13e12b51231deb46355be190102cbc}
\begin{DoxyVerb}This def creates ne-estimator caller object and adds it to list of args for a single call
to def do_estimate.  The call is then appended to llv_args_each_process
\end{DoxyVerb}
 

Definition at line 2333 of file pgdriveneestimator.\+py.


\begin{DoxyCode}
2333                                             i\_genepop\_file\_count ):
2334     \textcolor{stringliteral}{'''     }
2335 \textcolor{stringliteral}{    This def creates ne-estimator caller object and adds it to list of args for a single call}
2336 \textcolor{stringliteral}{    to def do\_estimate.  The call is then appended to llv\_args\_each\_process}
2337 \textcolor{stringliteral}{    '''}
2338 
2339     \textcolor{comment}{#all these GenepopFileManager objects should now have}
2340     \textcolor{comment}{#a single population sample name:}
2341     s\_population\_sample\_name=o\_genepopfile.population\_subsample\_tags[ 0 ]
2342 
2343     ls\_indiv\_sample\_names=o\_genepopfile.indiv\_subsample\_tags
2344 
2345     li\_population\_numbers=o\_genepopfile.getListPopulationNumbers( s\_population\_sample\_name )
2346     \textcolor{comment}{#for each sample (i.e. for each replicate of a sampling with a given set of params:}
2347 
2348     \textcolor{comment}{#Added this to help create a unique, temporary gp file name}
2349     \textcolor{comment}{#using file number individ sample, and loci sample counts:}
2350     i\_individ\_sample\_count=0
2351 
2352     \textcolor{keywordflow}{for} s\_indiv\_sample \textcolor{keywordflow}{in} ls\_indiv\_sample\_names:
2353 
2354         i\_individ\_sample\_count+=1
2355 
2356         \textcolor{comment}{#we also check that at least one pop has at least one individual}
2357         \textcolor{comment}{#listed for this subsample}
2358         li\_indiv\_counts\_for\_this\_sample=o\_genepopfile.getIndividualCounts( s\_indiv\_subsample\_tag=
      s\_indiv\_sample )
2359 
2360         \textcolor{keywordflow}{if} sum( li\_indiv\_counts\_for\_this\_sample ) == 0:
2361             s\_sample\_value, i\_replicate\_count=get\_sample\_val\_and\_rep\_number\_from\_sample\_name( \(\backslash\)
2362                                             s\_indiv\_sample, s\_sample\_scheme )
2363             \textcolor{comment}{#we only want to report the first replicate}
2364             \textcolor{comment}{#(by inference of course all replicates will be skipped):}
2365             \textcolor{keywordflow}{if} i\_replicate\_count == 0:
2366                 s\_msg=\textcolor{stringliteral}{"Skipping samples from: "} \(\backslash\)
2367                         + o\_genepopfile.original\_file\_name \(\backslash\)
2368                         + \textcolor{stringliteral}{".  In pgdriveneestimator.py, all populations "} \(\backslash\)
2369                         + \textcolor{stringliteral}{"have zero individuals when sampled using "} \(\backslash\)
2370                         + \textcolor{stringliteral}{"scheme: "} + s\_sample\_scheme \(\backslash\)
2371                         + \textcolor{stringliteral}{", and with sample parameter, "} + s\_sample\_value + \textcolor{stringliteral}{"."}
2372                 o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2373             \textcolor{comment}{#end if first replicate, report}
2374             \textcolor{keywordflow}{continue}
2375         \textcolor{comment}{#end skip this subsample if no pops have any individuals}
2376 
2377         \textcolor{stringliteral}{'''}
2378 \textcolor{stringliteral}{        For percent and remove-N schemes, the min/max pop size params are not }
2379 \textcolor{stringliteral}{        enforced in the sampler, and, as of 2016\_09\_27, are to be applied}
2380 \textcolor{stringliteral}{        before sampling.  Thus, for these scheme we skip populations if their }
2381 \textcolor{stringliteral}{        original, un-subsampled size is outside the pop size range.  Other schemes}
2382 \textcolor{stringliteral}{        should have applied the criteria when subsampling.  For the latter, this}
2383 \textcolor{stringliteral}{        would not be an appropriate test::}
2384 \textcolor{stringliteral}{        '''}
2385 
2386         li\_pops\_with\_invalid\_size=[]
2387         
2388         \textcolor{keywordflow}{if} s\_sample\_scheme \textcolor{keywordflow}{in} SAMPLE\_SCHEMES\_NON\_CRITERIA:
2389             li\_pops\_with\_invalid\_size=\hyperlink{namespacenegui_1_1pgdriveneestimator_a06645f9b380f736368e27585331ec22f}{get\_pops\_in\_file\_outside\_valid\_size\_range}
      ( o\_genepopfile, 
2390                                                                                 i\_min\_pop\_size,
2391                                                                                 i\_max\_pop\_size )
2392         \textcolor{comment}{#end if we need to check original (non-sampled) pop sizes}
2393 
2394         ls\_loci\_subsample\_tags\_associated\_with\_this\_indiv\_sample= \(\backslash\)
2395                 get\_loci\_subsample\_tags\_for\_this\_indiv\_subsample( o\_genepopfile,
2396                                                                     s\_indiv\_sample )
2397     
2398         i\_loci\_subsample\_count=0
2399 
2400         \textcolor{keywordflow}{for} s\_loci\_subsample\_tag \textcolor{keywordflow}{in} ls\_loci\_subsample\_tags\_associated\_with\_this\_indiv\_sample:
2401 
2402             \textcolor{stringliteral}{'''}
2403 \textcolor{stringliteral}{            To most-efficiently use multi-processes, that is, to assign many}
2404 \textcolor{stringliteral}{            processes a single file with many populations, we do an}
2405 \textcolor{stringliteral}{            estimate on each pop in the file separately, and divvy}
2406 \textcolor{stringliteral}{            the pops up among processes:}
2407 \textcolor{stringliteral}{            '''}
2408             \textcolor{keywordflow}{for} i\_population\_number \textcolor{keywordflow}{in} li\_population\_numbers:
2409 
2410                 i\_loci\_subsample\_count+=1
2411 
2412                 \textcolor{keywordflow}{if} i\_population\_number \textcolor{keywordflow}{in} li\_pops\_with\_invalid\_size:
2413                     s\_msg= \textcolor{stringliteral}{"In pgdriveneestimator.py, def add\_to\_set\_of\_calls\_to\_do\_estimate, "} \(\backslash\)
2414                                     + \textcolor{stringliteral}{" in file, "} + o\_genepopfile.original\_file\_name \(\backslash\)
2415                                     + \textcolor{stringliteral}{", skipping pop number, "}  + str( i\_population\_number ) \(\backslash\)
2416                                     + \textcolor{stringliteral}{".  It has an indiv count outside of valid "} \(\backslash\)
2417                                     + \textcolor{stringliteral}{"range, "} + str( i\_min\_pop\_size ) \(\backslash\)
2418                                     + \textcolor{stringliteral}{" - "} + str( i\_max\_pop\_size ) + \textcolor{stringliteral}{"."}
2419                     o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2420                     \textcolor{keywordflow}{continue}
2421                 \textcolor{comment}{#end if i\_population\_number is}
2422 
2423                 \textcolor{stringliteral}{'''}
2424 \textcolor{stringliteral}{                We send to def do\_estimate the genepop file object, o\_genepopfile, and the subsample tag}
2425 \textcolor{stringliteral}{                that identifies the correct subsample calculated and stored in the object, so we can write}
2426 \textcolor{stringliteral}{                each subsample genepop file in the same process that will use it as input to the the
       estimator.}
2427 \textcolor{stringliteral}{                This should limit the number of input files existing concurrenty in the directory to the }
2428 \textcolor{stringliteral}{                number of processes in use.  We also send allele freq, and the original genepop file name }
2429 \textcolor{stringliteral}{                inside the genepop file manager object):}
2430 \textcolor{stringliteral}{                '''}
2431                 s\_sample\_value, s\_replicate\_number = \(\backslash\)
2432                         get\_sample\_val\_and\_rep\_number\_from\_sample\_name( s\_indiv\_sample, s\_sample\_scheme )
2433 
2434                 s\_loci\_sample\_value, s\_loci\_replicate\_number = \(\backslash\)
2435                         \hyperlink{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}{get\_loci\_sample\_val\_and\_rep\_number\_from\_loci\_sample\_tag}
      ( s\_loci\_subsample\_tag )
2436                 
2437 
2438                 i\_tot\_indivs\_this\_subsample=o\_genepopfile.getIndividualCount( i\_population\_number, 
2439                                                                                         s\_indiv\_sample )
2440 
2441                 \textcolor{keywordflow}{if} i\_tot\_indivs\_this\_subsample == 0:
2442 
2443                     s\_msg= \textcolor{stringliteral}{"In pgdriveneestimator.py, def add\_to\_set\_of\_calls\_to\_do\_estimate, "} \(\backslash\)
2444                                     + \textcolor{stringliteral}{" no estimation was done on pop section number "}  \(\backslash\)
2445                                     + str( i\_population\_number ) \(\backslash\)
2446                                     + \textcolor{stringliteral}{" in file, "} + o\_genepopfile.original\_file\_name \(\backslash\)
2447                                     + \textcolor{stringliteral}{".  The parameters for population or loci "} \(\backslash\)
2448                                     + \textcolor{stringliteral}{"resulted in no individuals being sampled."}
2449 
2450                     o\_secondary\_outfile.write( s\_msg + \textcolor{stringliteral}{"\(\backslash\)n"} )
2451                     \textcolor{keywordflow}{continue}
2452                 \textcolor{comment}{# end if this sample for this pop has zero individuals}
2453 
2454                 \textcolor{comment}{#for downstream analysis, we get the census of total indiv}
2455                 \textcolor{comment}{#non-sampled, for this pop section:}
2456                 s\_census=str( o\_genepopfile.getIndividualCount( i\_population\_number ) )
2457 
2458                 \textcolor{comment}{#we make population subsample list consisting of only this populations number:}
2459                 s\_this\_pop\_number=str( i\_population\_number ) 
2460 
2461                 o\_genepopfile.subsamplePopulationsByList( [ i\_population\_number ], s\_this\_pop\_number )
2462 
2463                 \textcolor{keywordflow}{if} o\_debug\_mode.isSet( DebugMode.PRINT\_REPLICATE\_SELECTIONS ):
2464                     print\_test\_list\_replicate\_selection\_indices( o\_genepopfile, 
2465                                             s\_sample\_scheme,    
2466                                             s\_indiv\_sample, 
2467                                             s\_loci\_sample\_value,
2468                                             s\_loci\_replicate\_number,
2469                                             o\_secondary\_outfile, 
2470                                             s\_population\_subsample\_tag=s\_this\_pop\_number )
2471                 \textcolor{comment}{#end if we're testing, print }
2472 
2473                 \textcolor{stringliteral}{'''}
2474 \textcolor{stringliteral}{                2017\_04\_03.  A Windows limit on path length motivates new, shorter names}
2475 \textcolor{stringliteral}{                for the intermediate genepop files.  Now we simply use the i\_genepop\_file\_count}
2476 \textcolor{stringliteral}{                number passed to this def from def drive\_estimator, and the subsample count}
2477 \textcolor{stringliteral}{                to get a uniq identifier for this genepop file vs. all others in this run.}
2478 \textcolor{stringliteral}{                For now we retain the original code remm'd out, and construct the shorter}
2479 \textcolor{stringliteral}{                name in a new def:}
2480 \textcolor{stringliteral}{                '''}
2481                 \textcolor{comment}{#in naming a the genepop file for the subsampled pop,}
2482                 \textcolor{comment}{#some processing is needed to avoid NeEstimator file naming errors, and}
2483                 \textcolor{comment}{#path mangling, so we made a separate def:}
2484 \textcolor{comment}{#               s\_genepop\_file\_subsample=get\_subsample\_genepop\_file\_name( o\_genepopfile.original\_file\_name,
       }
2485 \textcolor{comment}{#                                                       s\_loci\_subsample\_tag, }
2486 \textcolor{comment}{#                                                       s\_this\_pop\_number,}
2487 \textcolor{comment}{#                                                       s\_temporary\_directory )}
2488                 
2489                 s\_genepop\_file\_subsample=\hyperlink{namespacenegui_1_1pgdriveneestimator_a9c650be4ed8ae5fa5e599523c22273bd}{make\_subsample\_genepop\_file\_name}( 
      o\_genepopfile.original\_file\_name,
2490                                                                                             
      s\_temporary\_directory, 
2491                                                                                                 
      i\_genepop\_file\_count, 
2492                                                                                                     
      i\_individ\_sample\_count,
2493                                                                                                         
      i\_loci\_subsample\_count )
2494 
2495                 s\_run\_output\_file=s\_genepop\_file\_subsample + \textcolor{stringliteral}{"\_ne\_run.txt"}
2496 
2497                 o\_neinput=pgin.PGInputNeEstimator( s\_genepop\_file\_subsample )
2498 
2499                 o\_neinput.run\_params=\{ \textcolor{stringliteral}{"crits"}:[ f\_min\_allele\_freq ]  \}
2500 
2501                 o\_neoutput=pgout.PGOutputNeEstimator( s\_genepop\_file\_subsample, 
2502                                                         s\_run\_output\_file, 
2503                                                         s\_estimator\_to\_use=ESTIMATOR\_TO\_USE )   
2504                 
2505                 o\_ne\_estimator=pgne.PGOpNeEstimator( o\_neinput, 
2506                                                         o\_neoutput, 
2507                                                         s\_estimator\_name=ESTIMATOR\_TO\_USE,
2508                                                         s\_parent\_dir\_for\_workspace=s\_temporary\_directory ) 
2509                 
2510                 \textcolor{keywordflow}{if} ESTIMATOR\_TO\_USE == LDNE2 \textcolor{keywordflow}{and} PATH\_TO\_LDNE2 \textcolor{keywordflow}{is} \textcolor{keywordflow}{not} \textcolor{keywordtype}{None}:
2511                     o\_ne\_estimator.ldne\_path=PATH\_TO\_LDNE2
2512                 \textcolor{comment}{#end if we are using LDNe2 and we have a path to the executable}
2513 
2514                 lv\_these\_args = [ o\_genepopfile,  
2515                                     o\_ne\_estimator, 
2516                                     s\_sample\_value, 
2517                                     s\_loci\_sample\_value,
2518                                     f\_min\_allele\_freq, 
2519                                     s\_indiv\_sample, 
2520                                     s\_loci\_subsample\_tag,
2521                                     s\_this\_pop\_number, 
2522                                     s\_census,
2523                                     s\_replicate\_number, 
2524                                     s\_loci\_replicate\_number,
2525                                     o\_debug\_mode,
2526                                     IDX\_NE\_ESTIMATOR\_OUTPUT\_FIELDS\_TO\_SKIP,
2527                                     f\_nbne\_ratio,
2528                                     i\_min\_loci\_position,
2529                                     i\_max\_loci\_position ]
2530 
2531                 llv\_args\_each\_process.append( lv\_these\_args )
2532             \textcolor{comment}{#end for each population}
2533         \textcolor{comment}{#end for each loci sample name assoc with this indiv sample name}
2534     \textcolor{comment}{#end for each indiv sample name}
2535     \textcolor{keywordflow}{return}
2536 \textcolor{comment}{#end add\_to\_set\_of\_calls\_to\_do\_estimate}
2537 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!do\+\_\+estimate@{do\+\_\+estimate}}
\index{do\+\_\+estimate@{do\+\_\+estimate}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{do\+\_\+estimate(xxx\+\_\+todo\+\_\+changeme)}{do_estimate(xxx_todo_changeme)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+do\+\_\+estimate (
\begin{DoxyParamCaption}
\item[{xxx\+\_\+todo\+\_\+changeme}]{}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_aeb13bee850fbafd8116a6b7e8048b98c}{}\label{namespacenegui_1_1pgdriveneestimator_aeb13bee850fbafd8116a6b7e8048b98c}
\begin{DoxyVerb}2017_07_07.  The paramaters min_loci_position
and max_loci_position are added to the arglist,
so that this def can compute the heterozygosity
only using the loci in the range specified by the
caller.
\end{DoxyVerb}
 

Definition at line 1105 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1105 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_aeb13bee850fbafd8116a6b7e8048b98c}{do\_estimate}(xxx\_todo\_changeme ):
1106 
1107     ( o\_genepopfile, o\_ne\_estimator, 
1108                     s\_sample\_param\_val, s\_loci\_sample\_value,
1109                     f\_min\_allele\_freq, s\_subsample\_tag, 
1110                     s\_loci\_sample\_tag, s\_population\_number, 
1111                     s\_census, i\_replicate\_number, 
1112                     s\_loci\_replicate\_number, o\_debug\_mode, 
1113                         IDX\_NE\_ESTIMATOR\_OUTPUT\_FIELDS\_TO\_SKIP,
1114                         f\_nbne\_ratio ) = xxx\_todo\_changeme
1115     \textcolor{keywordflow}{try}:
1116 
1117         s\_genepop\_file\_subsample=o\_ne\_estimator.input.genepop\_file
1118 
1119         o\_genepopfile.writeGenePopFile( s\_genepop\_file\_subsample, 
1120                                 s\_indiv\_subsample\_tag=s\_subsample\_tag,
1121                                 s\_pop\_subsample\_tag=s\_population\_number,
1122                                 s\_loci\_subsample\_tag=s\_loci\_sample\_tag ) 
1123 
1124         o\_ne\_estimator.doOp()
1125 
1126         llv\_output=o\_ne\_estimator.deliverResults()
1127 
1128         li\_sample\_indiv\_count=o\_genepopfile.getIndividualCounts( s\_indiv\_subsample\_tag = s\_subsample\_tag, 
1129                                                                     s\_pop\_subsample\_tag = 
      s\_population\_number )
1130 
1131         s\_sample\_indiv\_count = str( li\_sample\_indiv\_count[ 0 ] )
1132 
1133         ls\_runinfo=[ o\_genepopfile.original\_file\_name, 
1134                                     s\_population\_number, 
1135                                     s\_census,
1136                                     s\_sample\_indiv\_count, 
1137                                     s\_sample\_param\_val, 
1138                                     str( i\_replicate\_number), 
1139                                     s\_loci\_sample\_value,
1140                                     s\_loci\_replicate\_number,
1141                                     str( f\_min\_allele\_freq ) ]
1142         
1143         ls\_stdout=[]
1144 
1145         \textcolor{keywordflow}{for} lv\_output \textcolor{keywordflow}{in} llv\_output:
1146 
1147             ls\_fields\_to\_report=[]  
1148 
1149             ls\_output\_vals\_as\_strings=[ str( v\_val ) \textcolor{keywordflow}{for} v\_val \textcolor{keywordflow}{in} lv\_output ]
1150 
1151             \textcolor{keywordflow}{for} idx \textcolor{keywordflow}{in} range( len( ls\_output\_vals\_as\_strings ) ):
1152                 \textcolor{keywordflow}{if} idx \textcolor{keywordflow}{not} \textcolor{keywordflow}{in} IDX\_NE\_ESTIMATOR\_OUTPUT\_FIELDS\_TO\_SKIP:
1153                     ls\_fields\_to\_report.append( ls\_output\_vals\_as\_strings[ idx ] )
1154                 \textcolor{comment}{#end if idx is not for a field we're skipping}
1155             \textcolor{comment}{#end for each field's index}
1156 
1157             \textcolor{stringliteral}{'''}
1158 \textcolor{stringliteral}{            2017\_04\_17.  The bias adjustment code in this def is now simplified,}
1159 \textcolor{stringliteral}{            so that it is passed an f\_nbne\_ratio, that is either None, or a float,}
1160 \textcolor{stringliteral}{            so that this def can simply pass it and the objects needed to get the}
1161 \textcolor{stringliteral}{            un-adjusted estimate.  Formerly, we passed the genepop file object to a}
1162 \textcolor{stringliteral}{            allow the called def to search  the header for a ratio, if a flag}
1163 \textcolor{stringliteral}{            (also used to be passed to this def) was true.  Now this ratio finding}
1164 \textcolor{stringliteral}{            is done on a per-original-genepop file basis in def run\_estimator.}
1165 \textcolor{stringliteral}{            '''}
1166 
1167             s\_ratio\_used\_for\_adjustment, s\_bias\_adjusted\_value=\(\backslash\)
1168                         \hyperlink{namespacenegui_1_1pgdriveneestimator_a7df54bcca5c33c80e648c440527a85ce}{get\_string\_values\_ldne\_ratio\_and\_bias\_adjustment}
      ( \(\backslash\)
1169                                                                 o\_ne\_estimator,
1170                                                                 lv\_output,
1171                                                                 f\_nbne\_ratio )
1172 
1173             
1174             ls\_fields\_to\_report.append( s\_ratio\_used\_for\_adjustment )
1175             ls\_fields\_to\_report.append( s\_bias\_adjusted\_value )
1176 
1177             ls\_stdout.append( OUTPUT\_DELIMITER.join(  ls\_runinfo + ls\_fields\_to\_report )  )
1178 
1179         \textcolor{comment}{#end for each line of parsed NeEstimator Output}
1180         
1181         \textcolor{comment}{#Make one long string, endline-delimited, for the stdout.}
1182         \textcolor{comment}{#Note:  if the inpout genepop file had only one}
1183         \textcolor{comment}{#population, this should be a single line of text.}
1184 
1185         s\_stdout=ENDLINE\_SEQ.join( ls\_stdout )
1186 
1187         s\_stderr=\textcolor{keywordtype}{None}
1188 
1189         ls\_indiv\_list=\textcolor{keywordtype}{None}
1190 
1191         s\_nodat\_file=o\_ne\_estimator.output.getNoDatFileName()
1192 
1193         \textcolor{keywordflow}{if} o\_debug\_mode.isSet( DebugMode.MAKE\_INDIV\_TABLE ):
1194             ls\_indiv\_list = o\_genepopfile.getListIndividuals( 
1195                     i\_pop\_number=int( s\_population\_number ), 
1196                     s\_indiv\_subsample\_tag = s\_subsample\_tag )   
1197         \textcolor{comment}{#end if return list indiv}
1198         
1199         \textcolor{keywordflow}{if} not( o\_debug\_mode.isSet( DebugMode.KEEP\_ESTIMATOR\_FILES) ):
1200             \textcolor{keywordflow}{if} os.path.exists( o\_ne\_estimator.output.run\_output\_file ):
1201                 os.remove( o\_ne\_estimator.output.run\_output\_file )
1202             \textcolor{comment}{#end if NeEstimator's orig output file exists, delete it}
1203         \textcolor{comment}{#end if flag to keep output file is false}
1204 
1205         \textcolor{keywordflow}{if} not( o\_debug\_mode.isSet( DebugMode.KEEP\_REPLICATE\_GENEPOP\_FILES ) ):
1206             \textcolor{keywordflow}{if} os.path.exists( o\_ne\_estimator.output.run\_input\_file ):
1207                 os.remove( o\_ne\_estimator.output.run\_input\_file )
1208             \textcolor{comment}{#end if subsample genepop file exists, delete it}
1209         \textcolor{comment}{#end if flag to keep input file is false}
1210 
1211         \textcolor{keywordflow}{if} not( o\_debug\_mode.isSet( DebugMode.KEEP\_NODAT\_FILES ) ):
1212             s\_nodat\_file=o\_ne\_estimator.output.getNoDatFileName()
1213             \textcolor{keywordflow}{if} s\_nodat\_file \textcolor{keywordflow}{is} \textcolor{keywordflow}{not} \textcolor{keywordtype}{None}:
1214                 os.remove( s\_nodat\_file )
1215             \textcolor{comment}{#end for each nodatfile}
1216         \textcolor{comment}{#end if debug mode says remove nodats}
1217 
1218         \textcolor{keywordflow}{if} not( o\_debug\_mode.isSet( DebugMode.KEEP\_NE\_EXTRA\_FILES) ):
1219 
1220             ls\_extra\_files=glob.glob( \(\backslash\)
1221                     o\_ne\_estimator.output.run\_output\_file \(\backslash\)
1222                     + \textcolor{stringliteral}{"*x??.txt"} )
1223 
1224             \textcolor{stringliteral}{'''}
1225 \textcolor{stringliteral}{            2017\_03\_21. To accomodate the use of the  LDNe2 estimator, }
1226 \textcolor{stringliteral}{            to cleanup we add to glob the LDNe2 tabular output}
1227 \textcolor{stringliteral}{            file extention, "x.txt".}
1228 \textcolor{stringliteral}{            '''}
1229             ls\_extra\_files+=glob.glob( \(\backslash\)
1230                     o\_ne\_estimator.output.run\_output\_file \(\backslash\)
1231                     + \textcolor{stringliteral}{"x.txt"} )
1232 
1233             \textcolor{keywordflow}{for} o\_extra\_file \textcolor{keywordflow}{in} ls\_extra\_files:
1234                 os.remove( o\_extra\_file )
1235             \textcolor{comment}{#end for each nextpfile}
1236         \textcolor{comment}{#end if debug mode says remove nodats}
1237 
1238         
1239     \textcolor{keywordflow}{except} Exception \textcolor{keyword}{as} oex:
1240         o\_traceback=sys.exc\_info()[ 2 ]
1241         s\_trace\_msg=pgut.get\_traceback\_info\_about\_offending\_code( o\_traceback ) 
1242         \textcolor{keywordflow}{raise} Exception( \textcolor{stringliteral}{"An exception was raised in "} \(\backslash\)
1243                              + \textcolor{stringliteral}{"module pgdriveneestimator.py, "} \(\backslash\)
1244                              + \textcolor{stringliteral}{"def do\_estimate, with message: "} \(\backslash\)
1245                              + str( oex ) + \textcolor{stringliteral}{"\(\backslash\)n"} + \textcolor{stringliteral}{"Exeption origin from "} \(\backslash\)
1246                              + s\_trace\_msg )
1247     \textcolor{comment}{#end try...except}
1248     \textcolor{keywordflow}{return} \{ \textcolor{stringliteral}{"for\_stdout"} : s\_stdout, \textcolor{stringliteral}{"for\_stderr"} : s\_stderr, 
1249             \textcolor{stringliteral}{"for\_indiv\_table"}: \textcolor{keywordtype}{None} \textcolor{keywordflow}{if} ls\_indiv\_list \textcolor{keywordflow}{is} \textcolor{keywordtype}{None} \(\backslash\)
1250                     \textcolor{keywordflow}{else} \{ \textcolor{stringliteral}{"file"} : o\_genepopfile.original\_file\_name, 
1251                         \textcolor{stringliteral}{"sample\_val"}: s\_sample\_param\_val,
1252                         \textcolor{stringliteral}{"loci\_val"}:s\_loci\_sample\_value,
1253                         \textcolor{stringliteral}{"rep"} : i\_replicate\_number,
1254                         \textcolor{stringliteral}{"loci\_rep"}:s\_loci\_replicate\_number,
1255                         \textcolor{stringliteral}{"list"} : ls\_indiv\_list,
1256                         \textcolor{stringliteral}{"pop"} : s\_population\_number \} \}
1257 
1258 \textcolor{comment}{#end do\_estimate}
1259 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!do\+\_\+ldne\+\_\+bias\+\_\+adjustment@{do\+\_\+ldne\+\_\+bias\+\_\+adjustment}}
\index{do\+\_\+ldne\+\_\+bias\+\_\+adjustment@{do\+\_\+ldne\+\_\+bias\+\_\+adjustment}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{do\+\_\+ldne\+\_\+bias\+\_\+adjustment(f\+\_\+ldne\+\_\+estimate, f\+\_\+nbne\+\_\+ratio)}{do_ldne_bias_adjustment(f_ldne_estimate, f_nbne_ratio)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+do\+\_\+ldne\+\_\+bias\+\_\+adjustment (
\begin{DoxyParamCaption}
\item[{}]{f\+\_\+ldne\+\_\+estimate, }
\item[{}]{f\+\_\+nbne\+\_\+ratio}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a99dfc5e3dc217054199cd4f693e26236}{}\label{namespacenegui_1_1pgdriveneestimator_a99dfc5e3dc217054199cd4f693e26236}
\begin{DoxyVerb}2017_02_11. Implements, through the LDNENbBiasAdjustor
class, the bias adjustment to LDNE estimates
via eq (8) in Waples 2014 (see class description).
\end{DoxyVerb}
 

Definition at line 1095 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1095 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a99dfc5e3dc217054199cd4f693e26236}{do\_ldne\_bias\_adjustment}( f\_ldne\_estimate, f\_nbne\_ratio ):
1096     \textcolor{stringliteral}{'''}
1097 \textcolor{stringliteral}{    2017\_02\_11. Implements, through the LDNENbBiasAdjustor}
1098 \textcolor{stringliteral}{    class, the bias adjustment to LDNE estimates}
1099 \textcolor{stringliteral}{    via eq (8) in Waples 2014 (see class description).}
1100 \textcolor{stringliteral}{    '''}
1101     o\_adjustor=LDNENbBiasAdjustor( f\_ldne\_estimate, f\_nbne\_ratio )
1102     \textcolor{keywordflow}{return} o\_adjustor.adjusted\_nb
1103 \textcolor{comment}{#end def do\_ldne\_bias\_adjustment}
1104 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!do\+\_\+sample@{do\+\_\+sample}}
\index{do\+\_\+sample@{do\+\_\+sample}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{do\+\_\+sample(o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+range, i\+\_\+max\+\_\+pop\+\_\+range, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, i\+\_\+replicates, s\+\_\+loci\+\_\+sampling\+\_\+scheme, v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, i\+\_\+min\+\_\+total\+\_\+loci, i\+\_\+max\+\_\+total\+\_\+loci, i\+\_\+loci\+\_\+replicates)}{do_sample(o_genepopfile, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_replicates, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, i_loci_replicates)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+do\+\_\+sample (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+range, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+range, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+size, }
\item[{}]{s\+\_\+sample\+\_\+scheme, }
\item[{}]{lv\+\_\+sample\+\_\+values, }
\item[{}]{i\+\_\+replicates, }
\item[{}]{s\+\_\+loci\+\_\+sampling\+\_\+scheme, }
\item[{}]{v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, }
\item[{}]{i\+\_\+min\+\_\+loci\+\_\+position, }
\item[{}]{i\+\_\+max\+\_\+loci\+\_\+position, }
\item[{}]{i\+\_\+min\+\_\+total\+\_\+loci, }
\item[{}]{i\+\_\+max\+\_\+total\+\_\+loci, }
\item[{}]{i\+\_\+loci\+\_\+replicates}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a74f962e4a6815d8acc2e23c9ad146d0d}{}\label{namespacenegui_1_1pgdriveneestimator_a74f962e4a6815d8acc2e23c9ad146d0d}
\begin{DoxyVerb}2017_05_29.  We add a return value, so that the calling
def (drive_estimator), will know whether any pops were
sampled. If the call to do_sample_individuals has no
pops in range we return 0, otherwise, we do the loci
sampling and return the value returned, which will be
the pop list length, or zero if no loci are in range,
as given by min_loci_number and max_loci_number.
\end{DoxyVerb}
 

Definition at line 1556 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1556                 i\_loci\_replicates ):
1557     \textcolor{stringliteral}{'''}
1558 \textcolor{stringliteral}{    2017\_05\_29.  We add a return value, so that the calling}
1559 \textcolor{stringliteral}{    def (drive\_estimator), will know whether any pops were}
1560 \textcolor{stringliteral}{    sampled. If the call to do\_sample\_individuals has no}
1561 \textcolor{stringliteral}{    pops in range we return 0, otherwise, we do the loci}
1562 \textcolor{stringliteral}{    sampling and return the value returned, which will be}
1563 \textcolor{stringliteral}{    the pop list length, or zero if no loci are in range,}
1564 \textcolor{stringliteral}{    as given by min\_loci\_number and max\_loci\_number.}
1565 \textcolor{stringliteral}{    '''}
1566 
1567     li\_population\_list=\hyperlink{namespacenegui_1_1pgdriveneestimator_aa663cfeded8b8a3710d1a11192f12581}{do\_sample\_individuals}( o\_genepopfile, 
1568                     i\_min\_pop\_range,
1569                     i\_max\_pop\_range,
1570                     i\_min\_pop\_size,
1571                     i\_max\_pop\_size,
1572                     s\_sample\_scheme, 
1573                     lv\_sample\_values, 
1574                     i\_replicates )
1575 
1576     \textcolor{keywordflow}{if} li\_population\_list \textcolor{keywordflow}{is} \textcolor{keywordtype}{None}:
1577         \textcolor{keywordflow}{return} 0
1578     \textcolor{comment}{#end if no pops were sampled (i.e. min pop > total pops).  }
1579 
1580     \textcolor{comment}{#we also want to sample loci one loci sampling}
1581     \textcolor{comment}{#replicate for each each individual replicate,}
1582     \textcolor{comment}{#the latter now included in the genepopfile object:}
1583     i\_total\_pops\_sampled=\hyperlink{namespacenegui_1_1pgdriveneestimator_a408b2c39ad32bd720e35f2eace94509c}{do\_sample\_loci}( o\_genepopfile,
1584                     s\_loci\_sampling\_scheme,
1585                     v\_loci\_sampling\_scheme\_param,
1586                     i\_min\_loci\_position,
1587                     i\_max\_loci\_position,
1588                     i\_min\_total\_loci,
1589                     i\_max\_total\_loci,
1590                     li\_population\_list,
1591                     i\_loci\_replicates )
1592 
1593     \textcolor{keywordflow}{return} i\_total\_pops\_sampled
1594 \textcolor{comment}{#end do\_sample}
1595 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!do\+\_\+sample\+\_\+individuals@{do\+\_\+sample\+\_\+individuals}}
\index{do\+\_\+sample\+\_\+individuals@{do\+\_\+sample\+\_\+individuals}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{do\+\_\+sample\+\_\+individuals(o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+range, i\+\_\+max\+\_\+pop\+\_\+range, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size, s\+\_\+sample\+\_\+scheme, lv\+\_\+sample\+\_\+values, i\+\_\+replicates)}{do_sample_individuals(o_genepopfile, i_min_pop_range, i_max_pop_range, i_min_pop_size, i_max_pop_size, s_sample_scheme, lv_sample_values, i_replicates)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+do\+\_\+sample\+\_\+individuals (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+range, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+range, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+size, }
\item[{}]{s\+\_\+sample\+\_\+scheme, }
\item[{}]{lv\+\_\+sample\+\_\+values, }
\item[{}]{i\+\_\+replicates}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_aa663cfeded8b8a3710d1a11192f12581}{}\label{namespacenegui_1_1pgdriveneestimator_aa663cfeded8b8a3710d1a11192f12581}
\begin{DoxyVerb}do_sample adds subsample info to the o_genepopfile GenepopFileManager object,
and so changes it in place, returns no value.  The o_genepopfile can then
be accessed to write a new genepop file that contains the subsampled data.
sampler requires a population list (in case we want to sample fewer than
all the pops in the file). If the client passed in "all" parse_args
will have assigned None to both min and max.

2017_05_29.  We add check to see if the min pop range number is >
the total pops int he genepop file, in which case we return None.
\end{DoxyVerb}
 

Definition at line 1603 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1603                 i\_replicates ):
1604 
1605     \textcolor{stringliteral}{'''}
1606 \textcolor{stringliteral}{    do\_sample adds subsample info to the o\_genepopfile GenepopFileManager object,}
1607 \textcolor{stringliteral}{    and so changes it in place, returns no value.  The o\_genepopfile can then}
1608 \textcolor{stringliteral}{    be accessed to write a new genepop file that contains the subsampled data.}
1609 \textcolor{stringliteral}{    sampler requires a population list (in case we want to sample fewer than}
1610 \textcolor{stringliteral}{    all the pops in the file). If the client passed in "all" parse\_args}
1611 \textcolor{stringliteral}{    will have assigned None to both min and max.}
1612 \textcolor{stringliteral}{}
1613 \textcolor{stringliteral}{    2017\_05\_29.  We add check to see if the min pop range number is >}
1614 \textcolor{stringliteral}{    the total pops int he genepop file, in which case we return None.}
1615 \textcolor{stringliteral}{    '''}
1616 
1617     i\_total\_pops\_in\_file=o\_genepopfile.pop\_total
1618     li\_population\_list=\textcolor{keywordtype}{None}
1619 
1620     \textcolor{keywordflow}{if} i\_min\_pop\_range \textcolor{keywordflow}{is} \textcolor{keywordtype}{None}:
1621         \textcolor{comment}{#make sure both are None, else}
1622         \textcolor{comment}{#arg-parsing error:}
1623         \textcolor{keywordflow}{if} i\_max\_pop\_range \textcolor{keywordflow}{is} \textcolor{keywordtype}{None}:
1624             li\_population\_list=list( range( 1, i\_total\_pops\_in\_file + 1 ) )
1625         \textcolor{keywordflow}{else}:
1626             s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def do\_sample, "} \(\backslash\)
1627                         + \textcolor{stringliteral}{" found arg min pop range is None, "} \(\backslash\)
1628                         + \textcolor{stringliteral}{" but max pop range is not None, "} \(\backslash\)
1629                         + str( i\_max\_pop\_range ) \(\backslash\)
1630                         + \textcolor{stringliteral}{".  Invalid argument pair."}
1631             \textcolor{keywordflow}{raise} Exception( s\_msg )
1632         \textcolor{comment}{#end if max is None, then get all pops into list, else error}
1633     \textcolor{keywordflow}{elif} i\_min\_pop\_range < 1 \textcolor{keywordflow}{or} i\_min\_pop\_range > i\_max\_pop\_range:
1634             s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def do\_sample, "} \(\backslash\)
1635                         + \textcolor{stringliteral}{"min pop number range, "} \(\backslash\)
1636                         + str( i\_min\_pop\_range ) \(\backslash\)
1637                         + \textcolor{stringliteral}{", is less than one or "} \(\backslash\)
1638                         + \textcolor{stringliteral}{"greaterh than max pop number, "} \(\backslash\)
1639                         + str( i\_max\_pop\_range ) + \textcolor{stringliteral}{"."} 
1640             \textcolor{keywordflow}{raise} Exception( s\_msg )
1641         \textcolor{comment}{#end if min or max out of range}
1642     \textcolor{keywordflow}{elif} i\_min\_pop\_range > i\_total\_pops\_in\_file:
1643         \textcolor{keywordflow}{return} \textcolor{keywordtype}{None}
1644     \textcolor{keywordflow}{else}:
1645 
1646         i\_max\_pop\_range=min( i\_max\_pop\_range, i\_total\_pops\_in\_file )
1647         li\_population\_list=list(range( i\_min\_pop\_range, i\_max\_pop\_range + 1))
1648     \textcolor{comment}{#end if min pop range is None, else if range invalid, else if out of range, }
1649     \textcolor{comment}{#else good range, get corresponding pop number list }
1650 
1651     o\_sampler=\textcolor{keywordtype}{None}  
1652 
1653     s\_population\_subsample\_tag=s\_sample\_scheme
1654 
1655     \textcolor{keywordflow}{if} s\_sample\_scheme==SAMPLE\_BY\_NONE:
1656         o\_sample\_params=gps.GenepopFileSampleParamsNone( li\_population\_numbers=li\_population\_list,
1657                                                             i\_min\_pop\_size=i\_min\_pop\_size,
1658                                                             i\_max\_pop\_size=i\_max\_pop\_size,
1659                                                             i\_replicates=i\_replicates,
1660                                                             s\_population\_subsample\_name=
      s\_population\_subsample\_tag )    
1661 
1662         o\_sampler=gps.GenepopFileSamplerNone( o\_genepopfile, o\_sample\_params )
1663 
1664     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_PERCENTAGES:
1665 
1666         \textcolor{comment}{#our "proportion" sampler object needs proportions instead of percentages:}
1667 
1668         o\_sample\_params=gps.GenepopFileSampleParamsProportion( li\_population\_numbers=li\_population\_list,
1669                                                 lf\_proportions=lv\_sample\_values,
1670                                                 i\_replicates=i\_replicates,
1671                                                 s\_population\_subsample\_name=s\_population\_subsample\_tag )
1672 
1673         o\_sampler=gps.GenepopFileSamplerIndividualsByProportion( o\_genepopfile, o\_sample\_params )
1674 
1675     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_REMOVAL:
1676 
1677         o\_sample\_params=gps.GenepopFileSampleParamsRemoval( li\_population\_numbers=li\_population\_list,
1678                                                 li\_n\_to\_remove=lv\_sample\_values,
1679                                                 i\_replicates=i\_replicates,
1680                                                 s\_population\_subsample\_name=s\_population\_subsample\_tag )
1681 
1682         o\_sampler=gps.GenepopFileSamplerIndividualsByRemoval( o\_genepopfile, o\_sample\_params )
1683 
1684     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_CRITERIA:
1685 
1686         o\_indiv\_fields, o\_criteria = get\_criteria\_sampling\_params( lv\_sample\_values )
1687 
1688         o\_sample\_params=gps.GenepopFileSampleParamsCriteria( o\_genepop\_indiv\_id\_fields=o\_indiv\_fields,
1689                                                                     o\_genepop\_indiv\_id\_critera=o\_criteria,
1690                                                                     li\_population\_numbers=
      li\_population\_list,
1691                                                                     i\_min\_sampled\_pop\_size=i\_min\_pop\_size,
1692                                                                     i\_max\_sampled\_pop\_size=i\_max\_pop\_size,
1693                                                                     s\_population\_subsample\_name = 
      s\_population\_subsample\_tag,
1694                                                                     i\_replicates=i\_replicates )
1695 
1696         o\_sampler=gps.GenepopFileSamplerIndividualsByCriteria( o\_genepopfile, o\_sample\_params )
1697 
1698     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_COHORT:
1699 
1700         ( o\_indiv\_fields,
1701             i\_max\_age ) = get\_cohorts\_sampling\_params( lv\_sample\_values )
1702 
1703         o\_sample\_params=gps.GenepopFileSampleParamsAgeStructureCohorts( \(\backslash\)
1704                                         o\_genepop\_indiv\_id\_fields = o\_indiv\_fields,
1705                                         li\_population\_numbers = li\_population\_list,
1706                                         i\_max\_age=i\_max\_age,
1707                                         i\_min\_individuals\_per\_gen=i\_min\_pop\_size,
1708                                         i\_max\_individuals\_per\_gen=i\_max\_pop\_size,
1709                                         i\_replicates=i\_replicates )
1710 
1711         o\_sampler=gps.GenepopFileSamplerIndividualsAgeStructureCohorts( o\_genepopfile,
1712                                                                     o\_sample\_params )
1713     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_COHORT\_PERCENTAGE:
1714 
1715         \textcolor{comment}{#this scheme has params identical to the regular "cohorts" scheme,}
1716         \textcolor{comment}{#with a hyphenated list percentage values appended to the max-age param, }
1717         \textcolor{comment}{#so we extract the last two values of the last param list:  }
1718         
1719         COHORT\_PARAM\_LIST\_DELIMITER=\textcolor{stringliteral}{";"}
1720         COHORT\_PERC\_DELIM=\textcolor{stringliteral}{"-"}
1721         IDX\_MAX\_AGE=0
1722         IDX\_PERC=1
1723 
1724         s\_last\_param=lv\_sample\_values[ -1 ]
1725 
1726         ls\_age\_and\_percentage=s\_last\_param.split( \(\backslash\)
1727                         COHORT\_PARAM\_LIST\_DELIMITER )
1728 
1729         s\_max\_age=ls\_age\_and\_percentage[ IDX\_MAX\_AGE ]
1730         s\_percentages=ls\_age\_and\_percentage[ IDX\_PERC ]
1731         
1732         \textcolor{comment}{#we reassemble the list to match the expected}
1733         \textcolor{comment}{#cohort sampling params (without the percentage):}
1734         lv\_cohort\_param\_fields=lv\_sample\_values[ 0:len( lv\_sample\_values ) - 1 ] \(\backslash\)
1735                                                             + [ s\_max\_age ]     
1736         \textcolor{comment}{#Convert the percentages to a list of floats giving proportion:}
1737         lf\_proportions=[ old\_div(float( s\_percentage ),100.0) \textcolor{keywordflow}{for} s\_percentage  \(\backslash\)
1738                             \textcolor{keywordflow}{in} s\_percentages.split( COHORT\_PERC\_DELIM ) ]
1739 
1740         ( o\_indiv\_fields,
1741                 i\_max\_age ) = get\_cohorts\_sampling\_params( lv\_cohort\_param\_fields )
1742 
1743         o\_sample\_params=gps.GenepopFileSampleParamsAgeStructureCohorts( \(\backslash\)
1744                                             o\_genepop\_indiv\_id\_fields = o\_indiv\_fields,
1745                                             li\_population\_numbers = li\_population\_list,
1746                                             i\_max\_age=i\_max\_age,
1747                                             lf\_proportions=lf\_proportions,
1748                                             i\_min\_individuals\_per\_gen=i\_min\_pop\_size,
1749                                             i\_max\_individuals\_per\_gen=i\_max\_pop\_size,
1750                                             i\_replicates=i\_replicates )
1751 
1752         o\_sampler=gps.GenepopFileSamplerIndividualsAgeStructureCohortsPercentage( o\_genepopfile,
1753                                                                                     o\_sample\_params )
1754                                 
1755     \textcolor{keywordflow}{elif} s\_sample\_scheme==SAMPLE\_BY\_RELATEDS:
1756 
1757             ( o\_indiv\_fields,
1758                 f\_percent\_relateds) = get\_relateds\_sampling\_params( lv\_sample\_values )
1759 
1760             o\_sample\_params=gps.GenepopFileSampleParamsAgeStructureRelateds( \(\backslash\)
1761                                         o\_genepop\_indiv\_id\_fields = o\_indiv\_fields,
1762                                         li\_population\_numbers = li\_population\_list,
1763                                         f\_percent\_relateds\_per\_gen=f\_percent\_relateds,
1764                                         i\_min\_individuals\_per\_gen=i\_min\_pop\_size,
1765                                         i\_max\_individuals\_per\_gen=i\_max\_pop\_size,
1766                                         i\_replicates=i\_replicates )
1767 
1768             o\_sampler=gps.GenepopFileSamplerIndividualsAgeStructureRelateds( o\_genepopfile,
1769                                                                     o\_sample\_params )
1770     \textcolor{keywordflow}{else}:
1771 
1772         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def do\_sample, "} \(\backslash\)
1773                 + \textcolor{stringliteral}{"unknown sampling scheme: "} \(\backslash\)
1774                 + s\_sample\_scheme + \textcolor{stringliteral}{"."}
1775         \textcolor{keywordflow}{raise} Exception( s\_msg )
1776     \textcolor{comment}{#end if percent, else removal, else criteria, else cohort, else relateds, else error}
1777 
1778     o\_sampler.doSample()
1779 
1780     \textcolor{keywordflow}{return} li\_population\_list
1781 \textcolor{comment}{#end do\_sample\_individuals}
1782 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!do\+\_\+sample\+\_\+loci@{do\+\_\+sample\+\_\+loci}}
\index{do\+\_\+sample\+\_\+loci@{do\+\_\+sample\+\_\+loci}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{do\+\_\+sample\+\_\+loci(o\+\_\+genepopfile, s\+\_\+loci\+\_\+sampling\+\_\+scheme, v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, i\+\_\+min\+\_\+loci\+\_\+position, i\+\_\+max\+\_\+loci\+\_\+position, i\+\_\+min\+\_\+total\+\_\+loci, i\+\_\+max\+\_\+total\+\_\+loci, li\+\_\+population\+\_\+list, i\+\_\+loci\+\_\+replicates=1)}{do_sample_loci(o_genepopfile, s_loci_sampling_scheme, v_loci_sampling_scheme_param, i_min_loci_position, i_max_loci_position, i_min_total_loci, i_max_total_loci, li_population_list, i_loci_replicates=1)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+do\+\_\+sample\+\_\+loci (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{s\+\_\+loci\+\_\+sampling\+\_\+scheme, }
\item[{}]{v\+\_\+loci\+\_\+sampling\+\_\+scheme\+\_\+param, }
\item[{}]{i\+\_\+min\+\_\+loci\+\_\+position, }
\item[{}]{i\+\_\+max\+\_\+loci\+\_\+position, }
\item[{}]{i\+\_\+min\+\_\+total\+\_\+loci, }
\item[{}]{i\+\_\+max\+\_\+total\+\_\+loci, }
\item[{}]{li\+\_\+population\+\_\+list, }
\item[{}]{i\+\_\+loci\+\_\+replicates = {\ttfamily 1}}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a408b2c39ad32bd720e35f2eace94509c}{}\label{namespacenegui_1_1pgdriveneestimator_a408b2c39ad32bd720e35f2eace94509c}
\begin{DoxyVerb}As of 2016_10_06, we have no loci sampling scheme
but we still subsample by loci order, start index, stop index
and a max total -- randomly drawn if the start/stop range
results in total loci over the max. 

Further, we currently only permit one loci subsample for
each individual subsampling session -- i.e. for all 
of a given sample replicate.

As of 2016_10_20, we've added new scheme "percent", but
still doing one sample in the genepop file per call
to this def, which now still means one loci sample will
be applied to all replicate indiv samples.

2016_11_06, revising to do multiple loci sampling.  The
scheme is, for each existing individual subsample,
we subsample the loci for each loci-subsampling parameter
value, and for each of these, for the number of loci
replicates given by def param i_loci_replicates.

2017_05_29.  We add a check at the start of this def
to validate the min and max loci position arguments.
if the range is invalid, we throw and error.  If the
min is greater than the total number of loci, we return
a 0, to indicate no pops were sampled.  Otherwise, 
we return the number of items in the li_population_numbers.\end{DoxyVerb}
 

Definition at line 1791 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1791                         i\_loci\_replicates=1 ):
1792         
1793     \textcolor{stringliteral}{'''}
1794 \textcolor{stringliteral}{    As of 2016\_10\_06, we have no loci sampling scheme}
1795 \textcolor{stringliteral}{    but we still subsample by loci order, start index, stop index}
1796 \textcolor{stringliteral}{    and a max total -- randomly drawn if the start/stop range}
1797 \textcolor{stringliteral}{    results in total loci over the max. }
1798 \textcolor{stringliteral}{}
1799 \textcolor{stringliteral}{    Further, we currently only permit one loci subsample for}
1800 \textcolor{stringliteral}{    each individual subsampling session -- i.e. for all }
1801 \textcolor{stringliteral}{    of a given sample replicate.}
1802 \textcolor{stringliteral}{}
1803 \textcolor{stringliteral}{    As of 2016\_10\_20, we've added new scheme "percent", but}
1804 \textcolor{stringliteral}{    still doing one sample in the genepop file per call}
1805 \textcolor{stringliteral}{    to this def, which now still means one loci sample will}
1806 \textcolor{stringliteral}{    be applied to all replicate indiv samples.}
1807 \textcolor{stringliteral}{}
1808 \textcolor{stringliteral}{    2016\_11\_06, revising to do multiple loci sampling.  The}
1809 \textcolor{stringliteral}{    scheme is, for each existing individual subsample,}
1810 \textcolor{stringliteral}{    we subsample the loci for each loci-subsampling parameter}
1811 \textcolor{stringliteral}{    value, and for each of these, for the number of loci}
1812 \textcolor{stringliteral}{    replicates given by def param i\_loci\_replicates.}
1813 \textcolor{stringliteral}{}
1814 \textcolor{stringliteral}{    2017\_05\_29.  We add a check at the start of this def}
1815 \textcolor{stringliteral}{    to validate the min and max loci position arguments.}
1816 \textcolor{stringliteral}{    if the range is invalid, we throw and error.  If the}
1817 \textcolor{stringliteral}{    min is greater than the total number of loci, we return}
1818 \textcolor{stringliteral}{    a 0, to indicate no pops were sampled.  Otherwise, }
1819 \textcolor{stringliteral}{    we return the number of items in the li\_population\_numbers.}
1820 \textcolor{stringliteral}{}
1821 \textcolor{stringliteral}{    '''}
1822 
1823     \textcolor{keywordflow}{if} i\_min\_loci\_position < 1  \(\backslash\)
1824             \textcolor{keywordflow}{or} i\_min\_loci\_position > i\_max\_loci\_position:
1825         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def do\_sample\_loci, "} \(\backslash\)
1826                         + \textcolor{stringliteral}{"min loci number range, "} \(\backslash\)
1827                         + str( i\_min\_loci\_position ) \(\backslash\)
1828                         + \textcolor{stringliteral}{", is less than one or "} \(\backslash\)
1829                         + \textcolor{stringliteral}{"greater than max pop number, "} \(\backslash\)
1830                         + str( i\_max\_loci\_position ) + \textcolor{stringliteral}{"."} 
1831         \textcolor{keywordflow}{raise} Exception( s\_msg )
1832     \textcolor{keywordflow}{elif} i\_min\_loci\_position > o\_genepopfile.loci\_total:
1833             \textcolor{keywordflow}{return} 0
1834     \textcolor{comment}{#end if invalid range, else no loci in range}
1835 
1836 
1837     \textcolor{keywordflow}{for} s\_indiv\_subsample\_tag \textcolor{keywordflow}{in} o\_genepopfile.indiv\_subsample\_tags:
1838 
1839         \textcolor{keywordflow}{if} s\_loci\_sampling\_scheme == SAMPLE\_LOCI\_SCHEME\_NONE:
1840 
1841             \textcolor{stringliteral}{'''}
1842 \textcolor{stringliteral}{            Loci subsample tags consist of an individual subsample tag}
1843 \textcolor{stringliteral}{            followed by "\_l\_t\_v\_r" where t gives the scheme type, v gives}
1844 \textcolor{stringliteral}{            the salient param value for the scheme, and r gives the replicate.}
1845 \textcolor{stringliteral}{}
1846 \textcolor{stringliteral}{            }
1847 \textcolor{stringliteral}{            2017\_05\_24. Bug fix.  The subsampling for this scheme was being}
1848 \textcolor{stringliteral}{            done in a loop, for idx in range( i\_replicates), but the}
1849 \textcolor{stringliteral}{            i\_replicates parameter in the sampler was hard-coded to 1.}
1850 \textcolor{stringliteral}{            Hence only a single replicate was being done, with one}
1851 \textcolor{stringliteral}{            loci subsample tag, but overwritten i\_loci\_replicates}
1852 \textcolor{stringliteral}{            times.  The tsv file then had only one replicate of the sampling,}
1853 \textcolor{stringliteral}{            corresponding to the last sampling.  Since the instance of }
1854 \textcolor{stringliteral}{            GenepopFileSamplerLociByRangeAndTotal, in its doSample()}
1855 \textcolor{stringliteral}{            def, if passed i\_loci\_replicates by its GenepopFileSampleParamsLoci}
1856 \textcolor{stringliteral}{            object, will do each replicate and give it a uniq tag (as with}
1857 \textcolor{stringliteral}{            other schemes).}
1858 \textcolor{stringliteral}{            '''}
1859             s\_loci\_subsample\_tag\_base=s\_indiv\_subsample\_tag
1860 
1861             o\_sample\_params=gps.GenepopFileSampleParamsLoci( \(\backslash\)
1862                                         li\_population\_numbers=li\_population\_list,
1863                                         i\_min\_loci\_position=i\_min\_loci\_position,
1864                                         i\_max\_loci\_position=i\_max\_loci\_position,
1865                                         i\_min\_total\_loci=i\_min\_total\_loci,
1866                                         i\_max\_total\_loci=i\_max\_total\_loci,
1867                                         i\_replicates=i\_loci\_replicates,
1868                                         s\_sample\_tag\_base=s\_loci\_subsample\_tag\_base )
1869 
1870             o\_sampler=gps.GenepopFileSamplerLociByRangeAndTotal( o\_genepopfile,
1871                                                                     o\_sample\_params )
1872 
1873             o\_sampler.doSample()
1874 
1875         \textcolor{keywordflow}{elif} s\_loci\_sampling\_scheme == SAMPLE\_LOCI\_SCHEME\_PERCENT:
1876 
1877             s\_loci\_subsample\_tag\_base=s\_indiv\_subsample\_tag
1878 
1879             o\_sample\_params=gps.GenepopFileSampleParamsLoci( \(\backslash\)
1880                                                     li\_population\_numbers=li\_population\_list,
1881                                                     i\_min\_loci\_position=i\_min\_loci\_position,
1882                                                     i\_max\_loci\_position=i\_max\_loci\_position,
1883                                                     i\_min\_total\_loci=i\_min\_total\_loci,
1884                                                     i\_max\_total\_loci=i\_max\_total\_loci,
1885                                                     lf\_proportions=v\_loci\_sampling\_scheme\_param,
1886                                                     i\_replicates=i\_loci\_replicates,
1887                                                     s\_sample\_tag\_base=s\_loci\_subsample\_tag\_base )
1888 
1889             o\_sampler=gps.GenepopFileSamplerLociByRangeAndPercentage( o\_genepopfile,
1890                                                                     o\_sample\_params )
1891             o\_sampler.doSample()
1892 
1893         \textcolor{keywordflow}{elif} s\_loci\_sampling\_scheme == SAMPLE\_LOCI\_CONSTANT\_TOTAL:
1894 
1895             s\_loci\_subsample\_tag\_base=s\_indiv\_subsample\_tag
1896         
1897             \textcolor{stringliteral}{'''}
1898 \textcolor{stringliteral}{            For the constant sampling totals, the sampler ignores }
1899 \textcolor{stringliteral}{            the passed values #for min and max loci total, and }
1900 \textcolor{stringliteral}{            substitutes the current total for both min and max. Thus,}
1901 \textcolor{stringliteral}{            for each N in the list of Ns, (and also for each replicate),  }
1902 \textcolor{stringliteral}{            a random N will be drawn from the loci (positions). }
1903 \textcolor{stringliteral}{            '''}
1904 
1905             o\_sample\_params=gps.GenepopFileSampleParamsLoci( \(\backslash\)
1906                                         li\_population\_numbers=li\_population\_list,
1907                                         i\_min\_loci\_position=i\_min\_loci\_position,
1908                                         i\_max\_loci\_position=i\_max\_loci\_position,
1909                                         i\_min\_total\_loci=i\_min\_total\_loci,
1910                                         i\_max\_total\_loci=i\_max\_total\_loci,
1911                                         li\_sample\_totals=v\_loci\_sampling\_scheme\_param,
1912                                         i\_replicates=i\_loci\_replicates,
1913                                         s\_sample\_tag\_base=s\_loci\_subsample\_tag\_base )
1914 
1915             o\_sampler=gps.GenepopFileSamplerLociByRangeAndSampleTotalList( o\_genepopfile,
1916                                                                             o\_sample\_params )
1917             o\_sampler.doSample()
1918         \textcolor{keywordflow}{else}:
1919             s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def do\_sample\_loci, "} \(\backslash\)
1920                         + \textcolor{stringliteral}{"unknown loci sampling scheme: "} \(\backslash\)
1921                         + s\_loci\_sampling\_scheme + \textcolor{stringliteral}{"."}
1922             \textcolor{keywordflow}{raise} Exception( s\_msg )
1923         \textcolor{comment}{#end if sampling scheme none else unknown}
1924     \textcolor{comment}{#end for each individual subsample tag}
1925 
1926     \textcolor{keywordflow}{return} len( li\_population\_list )
1927 \textcolor{comment}{#end do\_sample\_loci}
1928 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+count\+\_\+estimator\+\_\+fields@{get\+\_\+count\+\_\+estimator\+\_\+fields}}
\index{get\+\_\+count\+\_\+estimator\+\_\+fields@{get\+\_\+count\+\_\+estimator\+\_\+fields}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+count\+\_\+estimator\+\_\+fields()}{get_count_estimator_fields()}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+count\+\_\+estimator\+\_\+fields (
\begin{DoxyParamCaption}
{}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_aaf461586686c9d7f0bf90188ab241dcc}{}\label{namespacenegui_1_1pgdriveneestimator_aaf461586686c9d7f0bf90188ab241dcc}
\begin{DoxyVerb}2017_06_22.   This def is called by do_estimate 
when the estimator produces no output, so that
an entry can be made in the tsv file (ex. with NA
values for estimator values) with correct field count.
\end{DoxyVerb}
 

Definition at line 2822 of file pgdriveneestimator.\+py.


\begin{DoxyCode}
2822 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_aaf461586686c9d7f0bf90188ab241dcc}{get\_count\_estimator\_fields}():
2823     \textcolor{stringliteral}{'''}
2824 \textcolor{stringliteral}{    2017\_06\_22.   This def is called by do\_estimate }
2825 \textcolor{stringliteral}{    when the estimator produces no output, so that}
2826 \textcolor{stringliteral}{    an entry can be made in the tsv file (ex. with NA}
2827 \textcolor{stringliteral}{    values for estimator values) with correct field count.}
2828 \textcolor{stringliteral}{    '''}
2829     i\_count=0
2830     \textcolor{keywordflow}{for} s\_field \textcolor{keywordflow}{in} pgout.PGOutputNeEstimator.OUTPUT\_FIELDS:
2831         \textcolor{keywordflow}{if} s\_field \textcolor{keywordflow}{not} \textcolor{keywordflow}{in} NE\_ESTIMATOR\_OUTPUT\_FIELDS\_TO\_SKIP:
2832             i\_count += 1
2833         \textcolor{comment}{#end if not a skipped field}
2834     \textcolor{comment}{#end for each output field}
2835     \textcolor{keywordflow}{return} i\_count
2836 \textcolor{comment}{#end get\_count\_estimator\_fields}
2837 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+genepop\+\_\+file\+\_\+list@{get\+\_\+genepop\+\_\+file\+\_\+list}}
\index{get\+\_\+genepop\+\_\+file\+\_\+list@{get\+\_\+genepop\+\_\+file\+\_\+list}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+genepop\+\_\+file\+\_\+list(s\+\_\+genepop\+\_\+files\+\_\+arg)}{get_genepop_file_list(s_genepop_files_arg)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+genepop\+\_\+file\+\_\+list (
\begin{DoxyParamCaption}
\item[{}]{s\+\_\+genepop\+\_\+files\+\_\+arg}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_ad0c4544bb6ce5bc184615cfade0f8495}{}\label{namespacenegui_1_1pgdriveneestimator_ad0c4544bb6ce5bc184615cfade0f8495}
\begin{DoxyVerb}if invoked from the console the arg is a glob expression
if invoked from the gui interface in pgguineestimator.py,
the arg is a list, such that eval will genreate a python
list of strings.

This def assumes that no user will invoke this mod from 
the command line with a glob expression, includeing a file name,
that can be evaluated as a python list (i.e. "[ \"item\", \"item\" ... ]\end{DoxyVerb}


\begin{DoxyVerb}if invoked from the console the arg is a glob expression, but
but if invoked from the gui interface in pgguineestimator.py,
the arg is a list, such that eval will genreate a python
list of strings.

This def assumes that no user will invoke this mod from 
the command line with a glob expression, includeing a file name,
that can be evaluated as a python list (i.e. "[ \"item\", \"item\" ... ]\end{DoxyVerb}
 

Definition at line 709 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
709 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_ad0c4544bb6ce5bc184615cfade0f8495}{get\_genepop\_file\_list}( s\_genepop\_files\_arg ):
710     \textcolor{stringliteral}{'''}
711 \textcolor{stringliteral}{    if invoked from the console the arg is a glob expression}
712 \textcolor{stringliteral}{    if invoked from the gui interface in pgguineestimator.py,}
713 \textcolor{stringliteral}{    the arg is a list, such that eval will genreate a python}
714 \textcolor{stringliteral}{    list of strings.}
715 \textcolor{stringliteral}{}
716 \textcolor{stringliteral}{    This def assumes that no user will invoke this mod from }
717 \textcolor{stringliteral}{    the command line with a glob expression, includeing a file name,}
718 \textcolor{stringliteral}{    that can be evaluated as a python list (i.e. "[ \(\backslash\)"item\(\backslash\)", \(\backslash\)"item\(\backslash\)" ... ]}
719 \textcolor{stringliteral}{    }
720 \textcolor{stringliteral}{    '''}
721 
722     ls\_files=\textcolor{keywordtype}{None}
723 
724     \textcolor{keywordflow}{try}:
725 
726         ls\_files=eval( s\_genepop\_files\_arg )
727 
728         \textcolor{keywordflow}{if} type( ls\_files ) != list:
729             \textcolor{comment}{#still could be a glob expression, for example,}
730             \textcolor{comment}{#of a file name that python can eval as an int or float,}
731             \textcolor{comment}{#so we'll just glob the orig:}
732 
733             ls\_files=glob.glob( s\_genepop\_files\_arg )
734         \textcolor{keywordflow}{else}:
735             \textcolor{comment}{#make sure it's a list of strings}
736             \textcolor{keywordflow}{if} not( file\_name\_list\_is\_all\_strings( ls\_files ) ):
737                 s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def get\_genepop\_file\_list, "} \(\backslash\)
738                         + \textcolor{stringliteral}{"genepop file arg evaluated as a list, but, not composed "} \(\backslash\)
739                         + \textcolor{stringliteral}{"of strings: "} + str( list ) + \textcolor{stringliteral}{"."}
740                 \textcolor{keywordflow}{raise} Exception( s\_msg )
741             \textcolor{comment}{#end if not all strings in list}
742     \textcolor{keywordflow}{except} ( NameError, SyntaxError ) \textcolor{keyword}{as} evalfail:
743 
744         \textcolor{stringliteral}{'''}
745 \textcolor{stringliteral}{        this is caught if the eval fails due to }
746 \textcolor{stringliteral}{        the arg being a string.  Further, a syntax}
747 \textcolor{stringliteral}{        error results if the arg begins/ends in a wildcard}
748 \textcolor{stringliteral}{        "*" char -- as will happen often with a glob.}
749 \textcolor{stringliteral}{}
750 \textcolor{stringliteral}{        So for these eval failures we assume the arg }
751 \textcolor{stringliteral}{        is a glob expression}
752 \textcolor{stringliteral}{        '''}
753 
754         ls\_files=glob.glob( s\_genepop\_files\_arg )
755 
756     \textcolor{keywordflow}{except} Exception \textcolor{keyword}{as} unknown\_fail:
757         s\_errortype=type( unknown\_fail ).\_\_name\_\_
758         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def get\_genepop\_file\_list, "} \(\backslash\)
759                         + \textcolor{stringliteral}{"gene pop file arg, "} \(\backslash\)
760                         + str( s\_genepop\_files\_arg ) \(\backslash\)
761                         + \textcolor{stringliteral}{", eval operation produced "} \(\backslash\)
762                         + \textcolor{stringliteral}{"error of type "} + s\_errortype \(\backslash\)
763                         + \textcolor{stringliteral}{", with message, "} + str( unknown\_fail ) + \textcolor{stringliteral}{"."} 
764         \textcolor{keywordflow}{raise} Exception( s\_msg )
765     \textcolor{comment}{#end try eval, except name error, except other}
766 
767     \textcolor{keywordflow}{return} ls\_files
768 \textcolor{comment}{#end get\_genepop\_file\_list}
769 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+invalid\+\_\+file\+\_\+names@{get\+\_\+invalid\+\_\+file\+\_\+names}}
\index{get\+\_\+invalid\+\_\+file\+\_\+names@{get\+\_\+invalid\+\_\+file\+\_\+names}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+invalid\+\_\+file\+\_\+names(ls\+\_\+file\+\_\+names)}{get_invalid_file_names(ls_file_names)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+invalid\+\_\+file\+\_\+names (
\begin{DoxyParamCaption}
\item[{}]{ls\+\_\+file\+\_\+names}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8972bdaeb431f941619300e9e63a0732}{}\label{namespacenegui_1_1pgdriveneestimator_a8972bdaeb431f941619300e9e63a0732}
\begin{DoxyVerb}Currently only check is for file name
length (see comments above declaration of const 
MAX_GENEPOP_FILE_NAME_LENGTH.
\end{DoxyVerb}
 

Definition at line 668 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
668 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a8972bdaeb431f941619300e9e63a0732}{get\_invalid\_file\_names}( ls\_file\_names ):
669     \textcolor{stringliteral}{'''}
670 \textcolor{stringliteral}{    Currently only check is for file name}
671 \textcolor{stringliteral}{    length (see comments above declaration of const }
672 \textcolor{stringliteral}{    MAX\_GENEPOP\_FILE\_NAME\_LENGTH.}
673 \textcolor{stringliteral}{    '''}
674     MSG\_FILE\_TOO\_LONG=\textcolor{stringliteral}{"File name longer than "} \(\backslash\)
675             + str( MAX\_GENEPOP\_FILE\_NAME\_LENGTH ) \(\backslash\)
676             + \textcolor{stringliteral}{" characters."}
677 
678     ls\_invalid\_file\_names\_and\_messages=[]
679 
680     \textcolor{keywordflow}{for} s\_file\_name \textcolor{keywordflow}{in} ls\_file\_names:
681         s\_file\_name\_no\_path=os.path.basename( s\_file\_name )
682         i\_len\_file\_name=len( s\_file\_name\_no\_path )
683 
684         \textcolor{keywordflow}{if} i\_len\_file\_name > MAX\_GENEPOP\_FILE\_NAME\_LENGTH:
685             ls\_invalid\_file\_names\_and\_messages.append( \(\backslash\)
686                     \textcolor{stringliteral}{", "}.join( [ s\_file\_name, MSG\_FILE\_TOO\_LONG ] ) )
687         \textcolor{comment}{#and if file name too long}
688     \textcolor{comment}{#end for each file name}
689 
690     \textcolor{keywordflow}{return} ls\_invalid\_file\_names\_and\_messages
691 \textcolor{comment}{#end get\_invalid\_file\_names}
692 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag@{get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag}}
\index{get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag@{get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag(s\+\_\+loci\+\_\+subsample\+\_\+tag)}{get_loci_sample_val_and_rep_number_from_loci_sample_tag(s_loci_subsample_tag)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+loci\+\_\+sample\+\_\+val\+\_\+and\+\_\+rep\+\_\+number\+\_\+from\+\_\+loci\+\_\+sample\+\_\+tag (
\begin{DoxyParamCaption}
\item[{}]{s\+\_\+loci\+\_\+subsample\+\_\+tag}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}{}\label{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}
\begin{DoxyVerb}The loci info in the tag follows an indiv tag prefix,
and has form "_l_t_v_r", with t giving the scheme,
v giving the scheme salient param value (ex percent val),
and r giving the replicate number.
\end{DoxyVerb}
 

Definition at line 2242 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
2242 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_ac6e046ac333dcd975329026d752ea2d9}{get\_loci\_sample\_val\_and\_rep\_number\_from\_loci\_sample\_tag}
      ( s\_loci\_subsample\_tag ):
2243     \textcolor{stringliteral}{'''}
2244 \textcolor{stringliteral}{    The loci info in the tag follows an indiv tag prefix,}
2245 \textcolor{stringliteral}{    and has form "\_l\_t\_v\_r", with t giving the scheme,}
2246 \textcolor{stringliteral}{    v giving the scheme salient param value (ex percent val),}
2247 \textcolor{stringliteral}{    and r giving the replicate number.}
2248 \textcolor{stringliteral}{    '''}
2249     \textcolor{comment}{#Sample param val is 2nd to last item}
2250     \textcolor{comment}{#in the tag fields, while the replicate}
2251     \textcolor{comment}{#number is the last:}
2252     IDX\_LOCI\_SAMPLE\_VAL=-2
2253     IDX\_LOCI\_REP=-1
2254 
2255     s\_sample\_val=\textcolor{keywordtype}{None}
2256     s\_replicate\_number=\textcolor{keywordtype}{None}
2257 
2258     s\_tag\_delim=gps.TAG\_DELIMITER
2259     
2260     ls\_tag\_fields=s\_loci\_subsample\_tag.split( s\_tag\_delim )
2261 
2262     s\_sample\_val=ls\_tag\_fields[ IDX\_LOCI\_SAMPLE\_VAL ]
2263     s\_replicate\_number=ls\_tag\_fields[ IDX\_LOCI\_REP ]
2264 
2265     \textcolor{keywordflow}{return} s\_sample\_val, s\_replicate\_number
2266 \textcolor{comment}{#end get\_loci\_sample\_val\_and\_rep\_number\_from\_loci\_sample\_tag}
2267 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs@{get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs}}
\index{get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs@{get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs(o\+\_\+genepopfile, s\+\_\+pop\+\_\+subsample\+\_\+tag=\+None, s\+\_\+individual\+\_\+subsample\+\_\+tag=\+None, s\+\_\+loci\+\_\+subsample\+\_\+tag=\+None)}{get_mean_het_based_on_allele_freqs(o_genepopfile, s_pop_subsample_tag=None, s_individual_subsample_tag=None, s_loci_subsample_tag=None)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+mean\+\_\+het\+\_\+based\+\_\+on\+\_\+allele\+\_\+freqs (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{s\+\_\+pop\+\_\+subsample\+\_\+tag = {\ttfamily None}, }
\item[{}]{s\+\_\+individual\+\_\+subsample\+\_\+tag = {\ttfamily None}, }
\item[{}]{s\+\_\+loci\+\_\+subsample\+\_\+tag = {\ttfamily None}}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a810fb4b95084358f69aafdc1e048451d}{}\label{namespacenegui_1_1pgdriveneestimator_a810fb4b95084358f69aafdc1e048451d}
\begin{DoxyVerb}2017_06_30.  Adds the mean heterozygosity calculated using allele frequencies of all loci,
for the pop (or pops) given by the pop subsample tag, which, as of this date, will always
be just a single pop, as this is the unit of estimation used in def do_estimate, and all loci
will be called using a subsample of the loci number ranges that the caller provided.
\end{DoxyVerb}
 

Definition at line 1201 of file pgdriveneestimator.\+py.


\begin{DoxyCode}
1201                                                 s\_loci\_subsample\_tag=\textcolor{keywordtype}{None} ):
1202 
1203     \textcolor{stringliteral}{'''}
1204 \textcolor{stringliteral}{    2017\_06\_30.  Adds the mean heterozygosity calculated using allele frequencies of all loci,}
1205 \textcolor{stringliteral}{    for the pop (or pops) given by the pop subsample tag, which, as of this date, will always}
1206 \textcolor{stringliteral}{    be just a single pop, as this is the unit of estimation used in def do\_estimate, and all loci}
1207 \textcolor{stringliteral}{    will be called using a subsample of the loci number ranges that the caller provided.}
1208 \textcolor{stringliteral}{    '''}
1209     df\_mean\_hets\_by\_pop=\{\}
1210 
1211     o\_loci\_info=gpl.GenepopFileLociInfo( o\_genepopfile, 
1212                                             s\_population\_subsample\_tag=s\_pop\_subsample\_tag,
1213                                             s\_individual\_subsample\_tag=s\_individual\_subsample\_tag,
1214                                             s\_loci\_subsample\_tag=s\_loci\_subsample\_tag )
1215 
1216     ddf\_het\_per\_pop\_per\_loci=o\_loci\_info.getHeterozygosity()
1217 
1218     \textcolor{keywordflow}{for} i\_pop \textcolor{keywordflow}{in} ddf\_het\_per\_pop\_per\_loci:
1219 
1220         lf\_het\_values\_this\_pop=list( ddf\_het\_per\_pop\_per\_loci[ i\_pop ].values() )
1221         f\_mean\_het\_value\_this\_pop=np.mean( lf\_het\_values\_this\_pop )
1222 
1223         df\_mean\_hets\_by\_pop[ i\_pop ]=f\_mean\_het\_value\_this\_pop
1224         
1225     \textcolor{comment}{#end for this pop}
1226 
1227     \textcolor{keywordflow}{return} df\_mean\_hets\_by\_pop
1228 
1229 \textcolor{comment}{#end get\_mean\_het\_based\_on\_allele\_freqs}
1230 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range@{get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range}}
\index{get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range@{get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range(o\+\_\+genepopfile, i\+\_\+min\+\_\+pop\+\_\+size, i\+\_\+max\+\_\+pop\+\_\+size)}{get_pops_in_file_outside_valid_size_range(o_genepopfile, i_min_pop_size, i_max_pop_size)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+pops\+\_\+in\+\_\+file\+\_\+outside\+\_\+valid\+\_\+size\+\_\+range (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile, }
\item[{}]{i\+\_\+min\+\_\+pop\+\_\+size, }
\item[{}]{i\+\_\+max\+\_\+pop\+\_\+size}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a06645f9b380f736368e27585331ec22f}{}\label{namespacenegui_1_1pgdriveneestimator_a06645f9b380f736368e27585331ec22f}
\begin{DoxyVerb}checks size of every pop in the file
in case we need to skip some pops and 
accept others.

Returns a list of pop numbers, 1-based
indexed (as caller expects).
\end{DoxyVerb}
 

Definition at line 1015 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1015                                                 i\_max\_pop\_size ):
1016     \textcolor{stringliteral}{'''}
1017 \textcolor{stringliteral}{    checks size of every pop in the file}
1018 \textcolor{stringliteral}{    in case we need to skip some pops and }
1019 \textcolor{stringliteral}{    accept others.}
1020 \textcolor{stringliteral}{}
1021 \textcolor{stringliteral}{    Returns a list of pop numbers, 1-based}
1022 \textcolor{stringliteral}{    indexed (as caller expects).}
1023 \textcolor{stringliteral}{    '''}
1024 
1025     li\_popsizes=o\_genepopfile.getIndividualCounts()
1026 
1027     li\_pops\_outside\_valid\_size\_range = []
1028     \textcolor{keywordflow}{for} idx\_pop \textcolor{keywordflow}{in} range( len ( li\_popsizes ) ):
1029         i\_size\_of\_pop=li\_popsizes[ idx\_pop ] 
1030 
1031         \textcolor{keywordflow}{if} i\_size\_of\_pop < i\_min\_pop\_size \(\backslash\)
1032                 \textcolor{keywordflow}{or} i\_size\_of\_pop > i\_max\_pop\_size:
1033         
1034             li\_pops\_outside\_valid\_size\_range.append( idx\_pop + 1 )
1035 
1036         \textcolor{comment}{#end if subsample size too small}
1037     \textcolor{comment}{#end for each pop number}
1038     \textcolor{keywordflow}{return} li\_pops\_outside\_valid\_size\_range
1039 \textcolor{comment}{#end get\_pops\_in\_file\_outside\_valid\_range}
1040 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment@{get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment}}
\index{get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment@{get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment(o\+\_\+ne\+\_\+estimator, lv\+\_\+results\+\_\+list, f\+\_\+nbne\+\_\+ratio)}{get_string_values_ldne_ratio_and_bias_adjustment(o_ne_estimator, lv_results_list, f_nbne_ratio)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+string\+\_\+values\+\_\+ldne\+\_\+ratio\+\_\+and\+\_\+bias\+\_\+adjustment (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+ne\+\_\+estimator, }
\item[{}]{lv\+\_\+results\+\_\+list, }
\item[{}]{f\+\_\+nbne\+\_\+ratio}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a7df54bcca5c33c80e648c440527a85ce}{}\label{namespacenegui_1_1pgdriveneestimator_a7df54bcca5c33c80e648c440527a85ce}
\begin{DoxyVerb}Revised on 2017_04_17, so that the nbne ratio passed here has 
already been vetted in run_estimator, so the we can test it
and, if it is None, we return "None" and the original LDNe
estimate.  Otherwise, we 
\end{DoxyVerb}
 

Definition at line 1043 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1043                                                             f\_nbne\_ratio ):
1044     \textcolor{stringliteral}{'''}
1045 \textcolor{stringliteral}{    Revised on 2017\_04\_17, so that the nbne ratio passed here has }
1046 \textcolor{stringliteral}{    already been vetted in run\_estimator, so the we can test it}
1047 \textcolor{stringliteral}{    and, if it is None, we return "None" and the original LDNe}
1048 \textcolor{stringliteral}{    estimate.  Otherwise, we }
1049 \textcolor{stringliteral}{    '''}
1050 
1051     s\_bias\_adjusted\_value=\textcolor{stringliteral}{"None"}
1052 
1053 
1054     \textcolor{stringliteral}{'''}
1055 \textcolor{stringliteral}{    We need the original Ne (Nb)  estimate}
1056 \textcolor{stringliteral}{    in order to do the bias adjustment, or,}
1057 \textcolor{stringliteral}{    if we cannot do the bias adjustment, }
1058 \textcolor{stringliteral}{    (that is, we don't have an Nb/Ne ratio),}
1059 \textcolor{stringliteral}{    then to write the original to the ne\_est\_adj}
1060 \textcolor{stringliteral}{    column.  This allows the Viz plotter to}
1061 \textcolor{stringliteral}{    use just the one column to plot the estimate,}
1062 \textcolor{stringliteral}{    whether or not a bias adjustment was performed.}
1063 \textcolor{stringliteral}{    '''}
1064     i\_col\_ne\_est=\(\backslash\)
1065                 o\_ne\_estimator.getOutputColumnNumberForFieldName( \(\backslash\)
1066                                             COL\_NAME\_NEESTIMATOR\_NE\_ESTIMATE )
1067 
1068     f\_unadjusted\_estimate=float( lv\_results\_list[ i\_col\_ne\_est ] )
1069 
1070     \textcolor{keywordflow}{if} f\_nbne\_ratio \textcolor{keywordflow}{is} \textcolor{keywordflow}{not} \textcolor{keywordtype}{None}:
1071         f\_adj\_estimate=\hyperlink{namespacenegui_1_1pgdriveneestimator_a99dfc5e3dc217054199cd4f693e26236}{do\_ldne\_bias\_adjustment}( f\_unadjusted\_estimate,
1072                                                             f\_nbne\_ratio )
1073         s\_bias\_adjusted\_value=str( f\_adj\_estimate ) 
1074     \textcolor{keywordflow}{else}:
1075         \textcolor{stringliteral}{'''}
1076 \textcolor{stringliteral}{        2017\_02\_15.  Instead of writing "None" as the bias-adjusted estimate, }
1077 \textcolor{stringliteral}{        when we have no bias adjustment, we'll enter the original estimate.  }
1078 \textcolor{stringliteral}{        This allows Brian T to find the Nb (Ne) values to plot in the new }
1079 \textcolor{stringliteral}{        ne\_est\_adj column instead of the original est\_ne column.  }
1080 \textcolor{stringliteral}{        By using original estimates when no bias adjustment was done,}
1081 \textcolor{stringliteral}{        he does not have to use conditional language to find }
1082 \textcolor{stringliteral}{        his Ne estimate.}
1083 \textcolor{stringliteral}{        '''}
1084         s\_bias\_adjusted\_value=str( f\_unadjusted\_estimate )
1085 
1086     \textcolor{comment}{#end if we have a nb/ne ratio, do bias adjustment }
1087     \textcolor{keywordflow}{return} str( f\_nbne\_ratio ), s\_bias\_adjusted\_value
1088 \textcolor{comment}{#end get\_string\_values\_ldne\_ratio\_and\_bias\_adjustment}
1089 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name@{get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name}}
\index{get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name@{get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name(s\+\_\+original\+\_\+genepop\+\_\+file\+\_\+name, s\+\_\+loci\+\_\+subsample\+\_\+tag, s\+\_\+this\+\_\+pop\+\_\+sample\+\_\+name, s\+\_\+temporary\+\_\+directory)}{get_subsample_genepop_file_name(s_original_genepop_file_name, s_loci_subsample_tag, s_this_pop_sample_name, s_temporary_directory)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+get\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name (
\begin{DoxyParamCaption}
\item[{}]{s\+\_\+original\+\_\+genepop\+\_\+file\+\_\+name, }
\item[{}]{s\+\_\+loci\+\_\+subsample\+\_\+tag, }
\item[{}]{s\+\_\+this\+\_\+pop\+\_\+sample\+\_\+name, }
\item[{}]{s\+\_\+temporary\+\_\+directory}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_af559a2d83516e3fb3fcebbfcdc96881b}{}\label{namespacenegui_1_1pgdriveneestimator_af559a2d83516e3fb3fcebbfcdc96881b}
\begin{DoxyVerb}Was originally simply a matter of replacing, in the original genepop file,
dots with underscores, but Windows revealed a case that needed attention 
(see "replace" below).
\end{DoxyVerb}
 

Definition at line 1932 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1932                                                     s\_temporary\_directory ):
1933     \textcolor{stringliteral}{'''}
1934 \textcolor{stringliteral}{    Was originally simply a matter of replacing, in the original genepop file,}
1935 \textcolor{stringliteral}{    dots with underscores, but Windows revealed a case that needed attention }
1936 \textcolor{stringliteral}{    (see "replace" below).}
1937 \textcolor{stringliteral}{    '''}
1938 
1939     s\_genepop\_file\_subsample=s\_original\_genepop\_file\_name \(\backslash\)
1940                                 + INPUT\_FILE\_DOT\_CHAR\_REPLACEMENT \(\backslash\)
1941                                 +  s\_loci\_subsample\_tag \(\backslash\)
1942                                 + INPUT\_FILE\_DOT\_CHAR\_REPLACEMENT \(\backslash\)
1943                                 + \textcolor{stringliteral}{"g"} + INPUT\_FILE\_DOT\_CHAR\_REPLACEMENT \(\backslash\)
1944                                 + s\_this\_pop\_sample\_name
1945     
1946     \textcolor{stringliteral}{'''}
1947 \textcolor{stringliteral}{    Note: In the original file name, if a period is left to the left of that}
1948 \textcolor{stringliteral}{    used do delimit the file extension, then NeEstimator, when naming}
1949 \textcolor{stringliteral}{    "*NoDat.txt" file, will replace the left-most period and all text that}
1950 \textcolor{stringliteral}{    follows it,  with "NoDat.txt" -- this then removes any percent and replicate}
1951 \textcolor{stringliteral}{    number text I'm using to make files unique, and so creates a non-uniq}
1952 \textcolor{stringliteral}{    *NoDat.txt file name, which, when one process removes it, and another goes}
1953 \textcolor{stringliteral}{    looking for it , throws an exception -- (an exception from the}
1954 \textcolor{stringliteral}{    multiprocessing.Pools async mapping -- file-not-found, but no trace since it}
1955 \textcolor{stringliteral}{    occurs in a different process from the main).  This peculiar mangling of a}
1956 \textcolor{stringliteral}{    users input file -- was preventing me in previous versions, from deleting}
1957 \textcolor{stringliteral}{    NoDat files, and also I think meant that only the last process in a given}
1958 \textcolor{stringliteral}{    group of replicates (share same genepop file and proportion sampled) to use}
1959 \textcolor{stringliteral}{    the NoDAt file got to write it's NoDat info.  Removing all periods from the}
1960 \textcolor{stringliteral}{    input file to the NeEstimator program, I belive, completely solves the}
1961 \textcolor{stringliteral}{    problems.}
1962 \textcolor{stringliteral}{    '''}
1963 
1964     s\_basename=os.path.basename( s\_genepop\_file\_subsample )
1965 
1966     \textcolor{stringliteral}{'''}
1967 \textcolor{stringliteral}{    2017\_03\_27.  We replace the path given by the original}
1968 \textcolor{stringliteral}{    genepop file with a temporary directory, if available.}
1969 \textcolor{stringliteral}{    '''}
1970     s\_dirs=\textcolor{keywordtype}{None}
1971 
1972     \textcolor{keywordflow}{if} s\_temporary\_directory \textcolor{keywordflow}{is} \textcolor{keywordtype}{None}:
1973         s\_dirs=os.path.dirname( s\_genepop\_file\_subsample )
1974     \textcolor{keywordflow}{else}:
1975         s\_dirs=s\_temporary\_directory
1976     \textcolor{comment}{#end if we have no temp directory, else use it instead of genepop file dir}
1977 
1978     s\_subsample\_file\_basename=s\_basename.replace( \textcolor{stringliteral}{"."}, INPUT\_FILE\_DOT\_CHAR\_REPLACEMENT )
1979 
1980     s\_genepop\_file\_subsample=os.path.join( s\_dirs, s\_subsample\_file\_basename )
1981     
1982     \textcolor{comment}{#remmed out -- problems when the orig file has path that includes }
1983     \textcolor{comment}{#directories with dots in their names, so the wholesale replace}
1984     \textcolor{comment}{#then includes directories that are renamed and (no doubt) non-existent}
1985 
1986     \textcolor{comment}{#s\_genepop\_file\_subsample=s\_genepop\_file\_subsample.replace( ".", INPUT\_FILE\_DOT\_CHAR\_REPLACEMENT )}
1987     
1988     \textcolor{stringliteral}{'''}
1989 \textcolor{stringliteral}{    problem seen in Windows, in which when this mod is invoked as \_\_main\_\_ from a shell}
1990 \textcolor{stringliteral}{    then ".\(\backslash\)\(\backslash\)" is appended to (at least) one genepop file name.  This then is mangled by}
1991 \textcolor{stringliteral}{    the above replace, so we restore the leading dot char:}
1992 \textcolor{stringliteral}{    '''}
1993 
1994     \textcolor{keywordflow}{if} s\_genepop\_file\_subsample.startswith(  \textcolor{stringliteral}{"\_"} ):
1995         s\_genepop\_file\_subsample=\textcolor{stringliteral}{"."} + s\_genepop\_file\_subsample[1:] 
1996     \textcolor{comment}{#end if path was mangled by replace}
1997 
1998     \textcolor{keywordflow}{return} s\_genepop\_file\_subsample
1999 \textcolor{comment}{#end get\_subsample\_genepop\_file\_name}
2000 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name@{make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name}}
\index{make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name@{make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name(s\+\_\+original\+\_\+genepop\+\_\+file, s\+\_\+temporary\+\_\+directory, i\+\_\+genepop\+\_\+file\+\_\+count, i\+\_\+indiv\+\_\+sample\+\_\+count, i\+\_\+loci\+\_\+subsample\+\_\+count)}{make_subsample_genepop_file_name(s_original_genepop_file, s_temporary_directory, i_genepop_file_count, i_indiv_sample_count, i_loci_subsample_count)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+make\+\_\+subsample\+\_\+genepop\+\_\+file\+\_\+name (
\begin{DoxyParamCaption}
\item[{}]{s\+\_\+original\+\_\+genepop\+\_\+file, }
\item[{}]{s\+\_\+temporary\+\_\+directory, }
\item[{}]{i\+\_\+genepop\+\_\+file\+\_\+count, }
\item[{}]{i\+\_\+indiv\+\_\+sample\+\_\+count, }
\item[{}]{i\+\_\+loci\+\_\+subsample\+\_\+count}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a9c650be4ed8ae5fa5e599523c22273bd}{}\label{namespacenegui_1_1pgdriveneestimator_a9c650be4ed8ae5fa5e599523c22273bd}
\begin{DoxyVerb}2017_04_03.  This def replaces def get_subsample_genepop_file_name,
in order to shorten the intermediate genepop file names, to address
the path length limitations of Windows.
\end{DoxyVerb}
 

Definition at line 2005 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
2005                                                         i\_loci\_subsample\_count ):
2006         \textcolor{stringliteral}{'''}
2007 \textcolor{stringliteral}{        2017\_04\_03.  This def replaces def get\_subsample\_genepop\_file\_name,}
2008 \textcolor{stringliteral}{        in order to shorten the intermediate genepop file names, to address}
2009 \textcolor{stringliteral}{        the path length limitations of Windows.}
2010 \textcolor{stringliteral}{        '''}
2011         s\_parent\_dir=\textcolor{keywordtype}{None}
2012 
2013         \textcolor{keywordflow}{if} s\_temporary\_directory \textcolor{keywordflow}{is} \textcolor{keywordtype}{None}:
2014             s\_parent\_dir=os.path.dirname( s\_original\_genepop\_file )
2015         \textcolor{keywordflow}{else}:
2016 
2017             s\_parent\_dir=s\_temporary\_directory  
2018         \textcolor{comment}{#end if no temp dir use orig genepop dir, else use temp}
2019 
2020         s\_file\_name=\textcolor{stringliteral}{"f"} + str( i\_genepop\_file\_count ) \(\backslash\)
2021                     + \textcolor{stringliteral}{"i"} + str( i\_indiv\_sample\_count ) \(\backslash\)
2022                     + \textcolor{stringliteral}{"l"} + str( i\_loci\_subsample\_count )
2023 
2024         s\_subsample\_genepop\_file=os.path.join( s\_parent\_dir, s\_file\_name )
2025         
2026         \textcolor{keywordflow}{return} s\_subsample\_genepop\_file
2027 
2028 \textcolor{comment}{#end make\_subsample\_genepop\_file\_name}
2029 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!mymain@{mymain}}
\index{mymain@{mymain}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{mymain(q\+\_\+args)}{mymain(q_args)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+mymain (
\begin{DoxyParamCaption}
\item[{}]{q\+\_\+args}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8344146a59d84da02da349abd68723ad}{}\label{namespacenegui_1_1pgdriveneestimator_a8344146a59d84da02da349abd68723ad}
\begin{DoxyVerb}This code was at mod level under if __name__=="__main__",
but moved inside this def so this mod can be imported
and run inside python with a set of args identical to
those required on the command line.

This was an adaptation to enable PGGuiNeEstimator objects
to import this mod, then call this def after they load 
params (args for this driver) from their interface

2017_05_31. We have added a new final string argument
"True" or "False" to indicate whether to use GUI
messaging. Note that parse_args does not parse this
final arg, but this def evaluates it immediately
in order to propery use error messaging.
\end{DoxyVerb}
\begin{DoxyVerb}The multiprocessing event arg is not yet added to 
the list, so that we'll find the gui messaging flag
at its designated index minus one.
\end{DoxyVerb}
 

Definition at line 3016 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
3016 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a8344146a59d84da02da349abd68723ad}{mymain}( *q\_args ):
3017 
3018     \textcolor{stringliteral}{'''}
3019 \textcolor{stringliteral}{    This code was at mod level under if \_\_name\_\_=="\_\_main\_\_",}
3020 \textcolor{stringliteral}{    but moved inside this def so this mod can be imported}
3021 \textcolor{stringliteral}{    and run inside python with a set of args identical to}
3022 \textcolor{stringliteral}{    those required on the command line.}
3023 \textcolor{stringliteral}{}
3024 \textcolor{stringliteral}{    This was an adaptation to enable PGGuiNeEstimator objects}
3025 \textcolor{stringliteral}{    to import this mod, then call this def after they load }
3026 \textcolor{stringliteral}{    params (args for this driver) from their interface}
3027 \textcolor{stringliteral}{    }
3028 \textcolor{stringliteral}{    2017\_05\_31. We have added a new final string argument}
3029 \textcolor{stringliteral}{    "True" or "False" to indicate whether to use GUI}
3030 \textcolor{stringliteral}{    messaging. Note that parse\_args does not parse this}
3031 \textcolor{stringliteral}{    final arg, but this def evaluates it immediately}
3032 \textcolor{stringliteral}{    in order to propery use error messaging.}
3033 \textcolor{stringliteral}{    '''}
3034 
3035     \textcolor{stringliteral}{'''}
3036 \textcolor{stringliteral}{    The multiprocessing event arg is not yet added to }
3037 \textcolor{stringliteral}{    the list, so that we'll find the gui messaging flag}
3038 \textcolor{stringliteral}{    at its designated index minus one.}
3039 \textcolor{stringliteral}{    '''}
3040     s\_gui\_flag=q\_args[ IDX\_USE\_GUI\_MESSAGING - 1 ]
3041 
3042     \textcolor{keywordflow}{if} s\_gui\_flag \textcolor{keywordflow}{not} \textcolor{keywordflow}{in} [ \textcolor{stringliteral}{"True"}, \textcolor{stringliteral}{"False"} ]:
3043 
3044         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def mymain, "} \(\backslash\)
3045                         + \textcolor{stringliteral}{"the program did not find the "} \(\backslash\)
3046                         + \textcolor{stringliteral}{"\(\backslash\)"True\(\backslash\)" or \(\backslash\)"False\(\backslash\)" value "} \(\backslash\)
3047                         + \textcolor{stringliteral}{"for the parameter that indicates "} \(\backslash\)
3048                         + \textcolor{stringliteral}{"whether to activate GUI messaging.  "} \(\backslash\)
3049                         + \textcolor{stringliteral}{"The value found is: "} + s\_gui\_flag + \textcolor{stringliteral}{"."}
3050         \textcolor{keywordflow}{raise} Exception( s\_msg )
3051     \textcolor{comment}{#end if no flag value}
3052 
3053     \hyperlink{namespacenegui_1_1pgdriveneestimator_a54013fc419eb559e07a5d278baa3fc6d}{set\_messaging\_procedures}( s\_gui\_flag )
3054 
3055     \textcolor{keywordflow}{try}:
3056 
3057         \textcolor{keywordflow}{if} VERY\_VERBOSE:
3058             print( \textcolor{stringliteral}{"In pgdriveneestimator, def mymain"} )
3059         \textcolor{comment}{#end if VERY\_VERBOSE}
3060 
3061         \textcolor{stringliteral}{'''}
3062 \textcolor{stringliteral}{        To allow the pygenomics genomics.ne2.controller modules}
3063 \textcolor{stringliteral}{        Ne2 estimator class object to loop over the paths in the}
3064 \textcolor{stringliteral}{        PATH environmental variable, and be ensured that each}
3065 \textcolor{stringliteral}{        exists.}
3066 \textcolor{stringliteral}{        '''}
3067         pgut.remove\_non\_existent\_paths\_from\_path\_variable()
3068 
3069         \textcolor{stringliteral}{'''}
3070 \textcolor{stringliteral}{        New code to confirm the selected executable,}
3071 \textcolor{stringliteral}{        Now that we are using LDNe2 by default,}
3072 \textcolor{stringliteral}{        with Ne2 as the alternative, as indicated by the}
3073 \textcolor{stringliteral}{        flag constant.}
3074 \textcolor{stringliteral}{        '''}
3075         b\_found\_executable=\textcolor{keyword}{False}
3076 
3077         \textcolor{keywordflow}{if} ESTIMATOR\_TO\_USE==LDNE2:
3078             b\_found\_executable = set\_mod\_level\_var\_to\_ldne2\_executable\_path()
3079         \textcolor{keywordflow}{elif} ESTIMATOR\_TO\_USE==NEESTIMATOR:
3080             b\_found\_executable = did\_find\_ne\_estimator\_executable()
3081         \textcolor{keywordflow}{else}:
3082             s\_msg = \textcolor{stringliteral}{"In pgdriveneestimator.py, def mymain(), "} \(\backslash\)
3083                             + \textcolor{stringliteral}{"checking for selected estimator executable. "} \(\backslash\)
3084                             + \textcolor{stringliteral}{"The selected estimator program is unknown: "} \(\backslash\)
3085                             + str( ESTIMATOR\_TO\_USE ) + \textcolor{stringliteral}{"."}
3086             \textcolor{keywordflow}{raise} Exception( s\_msg )
3087         \textcolor{comment}{#dnd if using ldne, else neestimator else unknown}
3088         
3089         \textcolor{keywordflow}{if} \textcolor{keywordflow}{not} b\_found\_executable:
3090             s\_msg = \textcolor{stringliteral}{"In pgdriveneestimator.py, def mymain(), "} \(\backslash\)
3091                                 + \textcolor{stringliteral}{"did not find NeEstimator executable."} 
3092             \textcolor{keywordflow}{raise} Exception(s\_msg )
3093         \textcolor{comment}{#end if can't find executable}
3094         
3095         \textcolor{stringliteral}{'''}
3096 \textcolor{stringliteral}{        We adapted this def to open and close the main and secondary }
3097 \textcolor{stringliteral}{        output files. Originally, it simply passed open file objects to def,}
3098 \textcolor{stringliteral}{        parse args.  Now we we check for string vs file for these}
3099 \textcolor{stringliteral}{        args, as the console version will still pass stdout and sterr}
3100 \textcolor{stringliteral}{        file objects, but the GUI will call with string filenames.  }
3101 \textcolor{stringliteral}{        We strip off the last 3 items, which now need to be checked }
3102 \textcolor{stringliteral}{        for type string file names and assume, if not string, then }
3103 \textcolor{stringliteral}{        the objects passed are assumed to be (open-for-writing) }
3104 \textcolor{stringliteral}{        file objects.  We also pass along the last arg, the multiprocessing}
3105 \textcolor{stringliteral}{        event.  }
3106 \textcolor{stringliteral}{        '''}
3107         seq\_unaltered\_args=q\_args[0:IDX\_LAST\_CONSOLE\_ARG + 1 ]
3108 
3109         v\_main\_outfile\_arg= q\_args[ IDX\_MAIN\_OUTFILE ]
3110         v\_secondary\_outfile\_arg=q\_args[ IDX\_SECONDARY\_OUTFILE ]
3111 
3112         \textcolor{stringliteral}{'''}
3113 \textcolor{stringliteral}{        New argument, added 2017\_03\_27, to allow def do\_estimate}
3114 \textcolor{stringliteral}{        to write its intermediate files inside a single temp}
3115 \textcolor{stringliteral}{        dir, whose name is known to the caller, so caller can}
3116 \textcolor{stringliteral}{        remove it.  }
3117 \textcolor{stringliteral}{        }
3118 \textcolor{stringliteral}{        Note that we subtract one to get its postiion in the}
3119 \textcolor{stringliteral}{        passed args.  This argument is passed as the last,}
3120 \textcolor{stringliteral}{        but the caller will not have passed the multiprocessing}
3121 \textcolor{stringliteral}{        event argument, so the index for our temp directory is}
3122 \textcolor{stringliteral}{        one less than the final position as passed to }
3123 \textcolor{stringliteral}{        drive\_estimator (and subsequenctly to parse\_args):}
3124 \textcolor{stringliteral}{        '''}
3125         v\_temporary\_directory\_arg=q\_args[ IDX\_TEMPORARY\_DIRECTORY - 1 ]
3126 
3127         \textcolor{stringliteral}{'''}
3128 \textcolor{stringliteral}{        Note that we check for string type, and otherwise}
3129 \textcolor{stringliteral}{        assume open file objects. (In Linux sys.std\{out,err\} return}
3130 \textcolor{stringliteral}{        "file" as type, but in Windows they return iostream or}
3131 \textcolor{stringliteral}{        something similar, but not file object.}
3132 \textcolor{stringliteral}{        '''}
3133         \textcolor{keywordflow}{if} type( v\_main\_outfile\_arg  ) == str:
3134             o\_main\_outfile=open( v\_main\_outfile\_arg, \textcolor{stringliteral}{'w'} )
3135             o\_secondary\_outfile=open( v\_secondary\_outfile\_arg, \textcolor{stringliteral}{'w'} )
3136         \textcolor{keywordflow}{else}:
3137             o\_main\_outfile=v\_main\_outfile\_arg
3138             o\_secondary\_outfile=v\_secondary\_outfile\_arg
3139         \textcolor{comment}{#end if passed non-file object, open file, else pass along file object}
3140 
3141         \textcolor{stringliteral}{'''}
3142 \textcolor{stringliteral}{        Reconstruct our arg list after creating the outfile objects.}
3143 \textcolor{stringliteral}{}
3144 \textcolor{stringliteral}{        2017\_02\_12.  Having replaced the use of the python multiprocessing.Process}
3145 \textcolor{stringliteral}{        object with a Popen call when calling this def from the pgutilities.py, we}
3146 \textcolor{stringliteral}{        no longer have the multiprocessing event available.  For now, rather than}
3147 \textcolor{stringliteral}{        removing this arg, the former multiprocessing event to allow communication}
3148 \textcolor{stringliteral}{        between pgutilities.py and this code, we simply default this arg to None:}
3149 \textcolor{stringliteral}{        '''}
3150         o\_event=\textcolor{keywordtype}{None}
3151 
3152         seq\_args\_to\_parse=seq\_unaltered\_args + ( o\_main\_outfile, 
3153                                                     o\_secondary\_outfile, 
3154                                                                     o\_event, 
3155                                                     v\_temporary\_directory\_arg )
3156 
3157         drive\_estimator( *seq\_args\_to\_parse )
3158 
3159         \textcolor{keywordflow}{if} o\_main\_outfile != sys.stdout:
3160             o\_main\_outfile.close()
3161         \textcolor{comment}{#end if not stdout, close}
3162 
3163         \textcolor{keywordflow}{if} o\_secondary\_outfile != sys.stderr:
3164             o\_secondary\_outfile.close()
3165         \textcolor{comment}{#end if not stderr, close file}
3166 
3167     \textcolor{keywordflow}{except} Exception \textcolor{keyword}{as} oex:
3168 
3169         \textcolor{keywordflow}{if} ACTIVATE\_GUI\_MESSAGING:
3170             s\_trace=o\_traceback=sys.exc\_info()[ 2 ]
3171             s\_trace= \(\backslash\)
3172                 pgut.get\_traceback\_info\_about\_offending\_code( o\_traceback )
3173 
3174             s\_errormsg=str( oex )
3175             GUIERR( \textcolor{keywordtype}{None} , s\_errormsg + \textcolor{stringliteral}{"\(\backslash\)n"} + s\_trace )
3176         \textcolor{comment}{#end if using gui messaging}
3177         
3178         \textcolor{keywordflow}{raise} ( oex )
3179     \textcolor{comment}{#end try ... except}
3180 
3181     \textcolor{keywordflow}{return}
3182 \textcolor{comment}{#end mymain}
3183 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!mymainlongfilelist@{mymainlongfilelist}}
\index{mymainlongfilelist@{mymainlongfilelist}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{mymainlongfilelist(q\+\_\+args)}{mymainlongfilelist(q_args)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+mymainlongfilelist (
\begin{DoxyParamCaption}
\item[{}]{q\+\_\+args}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a4010f65f71a1bce539c5e15317632be2}{}\label{namespacenegui_1_1pgdriveneestimator_a4010f65f71a1bce539c5e15317632be2}
\begin{DoxyVerb}2017_02_21.  This def is added as an alternative call
for pgutilitites run_driveneestimator_in_new_process
and its call to call_driveneestimator_using_subprocess,
to solve a Windows problem.  When calling with a long
list of (many hundreds) of genepop files, we saw a limit
of Windows ability to correctly make the call via pOpen, 
due to a limit on the number of characters it can handle
in a command.  Now the above pgutilities defs will have
tested the command length.  If it exceeds a threshold,
the comma-delimited list of genepop files will be
written to a temporary file.  This def will read in
the list from that file, delete the file, and then
call mymain with the now normalized arg 1.

We assume the list is formatted as it is when mymain
is called directly, and, further, that it is in a single,
first line in the file.
\end{DoxyVerb}
 

Definition at line 2944 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
2944 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a4010f65f71a1bce539c5e15317632be2}{mymainlongfilelist}( *q\_args ):
2945     \textcolor{stringliteral}{'''}
2946 \textcolor{stringliteral}{    2017\_02\_21.  This def is added as an alternative call}
2947 \textcolor{stringliteral}{    for pgutilitites run\_driveneestimator\_in\_new\_process}
2948 \textcolor{stringliteral}{    and its call to call\_driveneestimator\_using\_subprocess,}
2949 \textcolor{stringliteral}{    to solve a Windows problem.  When calling with a long}
2950 \textcolor{stringliteral}{    list of (many hundreds) of genepop files, we saw a limit}
2951 \textcolor{stringliteral}{    of Windows ability to correctly make the call via pOpen, }
2952 \textcolor{stringliteral}{    due to a limit on the number of characters it can handle}
2953 \textcolor{stringliteral}{    in a command.  Now the above pgutilities defs will have}
2954 \textcolor{stringliteral}{    tested the command length.  If it exceeds a threshold,}
2955 \textcolor{stringliteral}{    the comma-delimited list of genepop files will be}
2956 \textcolor{stringliteral}{    written to a temporary file.  This def will read in}
2957 \textcolor{stringliteral}{    the list from that file, delete the file, and then}
2958 \textcolor{stringliteral}{    call mymain with the now normalized arg 1.}
2959 \textcolor{stringliteral}{}
2960 \textcolor{stringliteral}{    We assume the list is formatted as it is when mymain}
2961 \textcolor{stringliteral}{    is called directly, and, further, that it is in a single,}
2962 \textcolor{stringliteral}{    first line in the file.}
2963 \textcolor{stringliteral}{    '''}
2964 
2965     s\_temp\_file\_with\_genepop\_file\_list=q\_args[ IDX\_GENEPOP\_FILES ]
2966 
2967     o\_temp\_file\_with\_genepop\_file=open( s\_temp\_file\_with\_genepop\_file\_list )
2968 
2969     \textcolor{comment}{#Thie file's single line should give the correct first argument of}
2970     \textcolor{comment}{#a list of genepop files, so that mymain can pass it to parse\_args }
2971     \textcolor{comment}{#so that it is parsed into a list of file names.}
2972     s\_genepop\_file\_list=o\_temp\_file\_with\_genepop\_file.readline().strip()
2973     q\_corrected\_args=( s\_genepop\_file\_list, ) + q\_args[ 1 : ]
2974 
2975     o\_temp\_file\_with\_genepop\_file.close()
2976 
2977     \hyperlink{namespacenegui_1_1pgdriveneestimator_a8344146a59d84da02da349abd68723ad}{mymain}( *q\_corrected\_args )
2978 
2979     \textcolor{keywordflow}{return}
2980 
2981 \textcolor{comment}{#end mymainlongfilelist}
2982 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields@{parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields}}
\index{parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields@{parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields(ls\+\_\+sampling\+\_\+list\+\_\+items)}{parse_sampling_list_that_uses_indiv_id_fields(ls_sampling_list_items)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+parse\+\_\+sampling\+\_\+list\+\_\+that\+\_\+uses\+\_\+indiv\+\_\+id\+\_\+fields (
\begin{DoxyParamCaption}
\item[{}]{ls\+\_\+sampling\+\_\+list\+\_\+items}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8918df5a619173d4baa530e79dfe2385}{}\label{namespacenegui_1_1pgdriveneestimator_a8918df5a619173d4baa530e79dfe2385}
\begin{DoxyVerb}We require that we have at least 3 items
to set up  sampling using individual id field
criteria:
--list of field names (space delimited)
    ex: "age parent1 parent2"
--list of field types (python types, space delimited)
    ex: "int str str"
--at least one parameter (not validated here
    -> example in criteria expression  "age > 3"
    ->example, in cohort we expect several params like max. age     
2016_11_10
    ->example, in cohorts by percentages, we'll have the max age
      followed by a hyphented list of percentage numerics.  This
      will not be parsed here, but will be passsed as part of the
      list returned as  ls_params\end{DoxyVerb}
 

Definition at line 1401 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
1401 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a8918df5a619173d4baa530e79dfe2385}{parse\_sampling\_list\_that\_uses\_indiv\_id\_fields}( 
      ls\_sampling\_list\_items ):
1402 
1403     \textcolor{stringliteral}{'''}
1404 \textcolor{stringliteral}{    We require that we have at least 3 items}
1405 \textcolor{stringliteral}{    to set up  sampling using individual id field}
1406 \textcolor{stringliteral}{    criteria:}
1407 \textcolor{stringliteral}{    --list of field names (space delimited)}
1408 \textcolor{stringliteral}{        ex: "age parent1 parent2"}
1409 \textcolor{stringliteral}{    --list of field types (python types, space delimited)}
1410 \textcolor{stringliteral}{        ex: "int str str"}
1411 \textcolor{stringliteral}{    --at least one parameter (not validated here}
1412 \textcolor{stringliteral}{        -> example in criteria expression  "age > 3"}
1413 \textcolor{stringliteral}{        ->example, in cohort we expect several params like max. age     }
1414 \textcolor{stringliteral}{    2016\_11\_10}
1415 \textcolor{stringliteral}{        ->example, in cohorts by percentages, we'll have the max age}
1416 \textcolor{stringliteral}{          followed by a hyphented list of percentage numerics.  This}
1417 \textcolor{stringliteral}{          will not be parsed here, but will be passsed as part of the}
1418 \textcolor{stringliteral}{          list returned as  ls\_params}
1419 \textcolor{stringliteral}{}
1420 \textcolor{stringliteral}{    '''}
1421 
1422     MIN\_NUM\_SAMPLING\_ITEMS=3
1423     IDX\_FIELD\_NAMES=0
1424     IDX\_FIELD\_TYPES=1
1425     IDX\_FIRST\_PARAM=2
1426     FIELD\_DELIMITER=\textcolor{stringliteral}{";"}
1427 
1428     i\_tot\_sampling\_items=len( ls\_sampling\_list\_items )
1429 
1430     \textcolor{keywordflow}{if} i\_tot\_sampling\_items < MIN\_NUM\_SAMPLING\_ITEMS:
1431         s\_msg = \textcolor{stringliteral}{"In pgdriveneestimator.py, def parse\_sampling\_list\_that\_uses\_indiv\_id\_fields, "} \(\backslash\)
1432                     + \textcolor{stringliteral}{"for the sampling criteria list, expecting at 3 items "} \(\backslash\)
1433                     + \textcolor{stringliteral}{"(field names, field types, and one or more "} \(\backslash\)
1434                     + \textcolor{stringliteral}{"parameters.  Found list: "} \(\backslash\)
1435                     + str( ls\_sampling\_list\_items ) + \textcolor{stringliteral}{"."}
1436         \textcolor{keywordflow}{raise} Exception( s\_msg )
1437 
1438     \textcolor{comment}{#end if too few criteria items}
1439     ls\_field\_names=ls\_sampling\_list\_items[ IDX\_FIELD\_NAMES ].split( FIELD\_DELIMITER )
1440     ls\_types=ls\_sampling\_list\_items[ IDX\_FIELD\_TYPES ].split( FIELD\_DELIMITER )
1441     \textcolor{keywordflow}{try}:
1442 
1443         lo\_field\_types=[ eval( s\_type ) \textcolor{keywordflow}{for} s\_type \textcolor{keywordflow}{in} ls\_types ]
1444     
1445     \textcolor{keywordflow}{except} Exception \textcolor{keyword}{as} oex:
1446         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py, def parse\_sampling\_list\_that\_uses\_indiv\_id\_fields, "} \(\backslash\)
1447                     + \textcolor{stringliteral}{"Exception trying to evaluate the field type list. "}\(\backslash\)
1448                     + \textcolor{stringliteral}{"Type list: "} + str( ls\_types ) + \textcolor{stringliteral}{".  Exception raised: "} \(\backslash\)
1449                     + str( oex ) + \textcolor{stringliteral}{"."}
1450         \textcolor{keywordflow}{raise} Exception ( s\_msg )
1451     \textcolor{comment}{#end try ... except}
1452 
1453     ls\_params=[]
1454 
1455     \textcolor{keywordflow}{for} idx \textcolor{keywordflow}{in} range( IDX\_FIRST\_PARAM, i\_tot\_sampling\_items ):
1456         ls\_params.append( ls\_sampling\_list\_items[ idx ] )
1457     \textcolor{comment}{#end for each item with idx>1}
1458 
1459     \textcolor{keywordflow}{return} ls\_field\_names, lo\_field\_types, ls\_params
1460 
1461 \textcolor{comment}{#end parse\_sampling\_list\_that\_uses\_indiv\_id\_fields}
1462 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file@{raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file}}
\index{raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file@{raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file(o\+\_\+genepopfile)}{raise_exception_if_non_single_pop_file(o_genepopfile)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+raise\+\_\+exception\+\_\+if\+\_\+non\+\_\+single\+\_\+pop\+\_\+file (
\begin{DoxyParamCaption}
\item[{}]{o\+\_\+genepopfile}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a4551e387ca20ed229183818ac8ce733c}{}\label{namespacenegui_1_1pgdriveneestimator_a4551e387ca20ed229183818ac8ce733c}
\begin{DoxyVerb}Not currently in use ( 2016_08_10 ) -- was used when
genepop files with >1 population were not accepted.
\end{DoxyVerb}
 

Definition at line 991 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
991 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a4551e387ca20ed229183818ac8ce733c}{raise\_exception\_if\_non\_single\_pop\_file}( o\_genepopfile ):
992     \textcolor{stringliteral}{'''}
993 \textcolor{stringliteral}{    Not currently in use ( 2016\_08\_10 ) -- was used when}
994 \textcolor{stringliteral}{    genepop files with >1 population were not accepted.}
995 \textcolor{stringliteral}{    '''}
996     i\_total\_pops=o\_genepopfile.pop\_total
997 
998     \textcolor{keywordflow}{if} ( i\_total\_pops != 1 ):
999         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator.py:"} + OUTPUT\_ENDLINE
1000         \textcolor{keywordflow}{if} i\_total\_pops < 1:
1001             s\_msg+=\textcolor{stringliteral}{"No \(\backslash\)"pop\(\backslash\)" entry found in file, "} \(\backslash\)
1002                     + o\_genepopfile.original\_file\_name
1003         \textcolor{keywordflow}{else}:
1004             s\_msg+=\textcolor{stringliteral}{"Current implementation of drive accepts genepop "} \(\backslash\)
1005                 + \textcolor{stringliteral}{"files with only a single pop."}
1006         \textcolor{comment}{#end if lt 1 else more}
1007         \textcolor{keywordflow}{raise} Exception( s\_msg )
1008     \textcolor{comment}{#end if non-single pop file}
1009 
1010     \textcolor{keywordflow}{return}
1011 \textcolor{comment}{#end raise\_exception\_non\_single\_pop\_file}
1012 
\end{DoxyCode}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!set\+\_\+messaging\+\_\+procedures@{set\+\_\+messaging\+\_\+procedures}}
\index{set\+\_\+messaging\+\_\+procedures@{set\+\_\+messaging\+\_\+procedures}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{set\+\_\+messaging\+\_\+procedures(s\+\_\+gui\+\_\+flag)}{set_messaging_procedures(s_gui_flag)}}]{\setlength{\rightskip}{0pt plus 5cm}def negui.\+pgdriveneestimator.\+set\+\_\+messaging\+\_\+procedures (
\begin{DoxyParamCaption}
\item[{}]{s\+\_\+gui\+\_\+flag}
\end{DoxyParamCaption}
)}\hypertarget{namespacenegui_1_1pgdriveneestimator_a54013fc419eb559e07a5d278baa3fc6d}{}\label{namespacenegui_1_1pgdriveneestimator_a54013fc419eb559e07a5d278baa3fc6d}
\begin{DoxyVerb}2017_05_31. This def was added in order to
control the GUI messaging, and limit it to
the processes that call def mymain from
the GUI-based program.  Console based calls
will have the flag set to "False."
\end{DoxyVerb}
 

Definition at line 2983 of file pgdriveneestimator.\+experimental.\+py.


\begin{DoxyCode}
2983 \textcolor{keyword}{def }\hyperlink{namespacenegui_1_1pgdriveneestimator_a54013fc419eb559e07a5d278baa3fc6d}{set\_messaging\_procedures}( s\_gui\_flag ):
2984     \textcolor{stringliteral}{'''}
2985 \textcolor{stringliteral}{    2017\_05\_31. This def was added in order to}
2986 \textcolor{stringliteral}{    control the GUI messaging, and limit it to}
2987 \textcolor{stringliteral}{    the processes that call def mymain from}
2988 \textcolor{stringliteral}{    the GUI-based program.  Console based calls}
2989 \textcolor{stringliteral}{    will have the flag set to "False."}
2990 \textcolor{stringliteral}{    '''}
2991 
2992     \textcolor{keyword}{global} ACTIVATE\_GUI\_MESSAGING
2993 
2994     \textcolor{keywordflow}{if} s\_gui\_flag == \textcolor{stringliteral}{"True"}:
2995 
2996         \textcolor{keyword}{global} GUIERR
2997         \textcolor{keyword}{global} GUIINFO
2998 
2999         ACTIVATE\_GUI\_MESSAGING=\textcolor{keyword}{True}
3000         \textcolor{keyword}{from} pgguiutilities \textcolor{keyword}{import} PGGUIErrorMessage
3001         \textcolor{keyword}{from} pgguiutilities \textcolor{keyword}{import} PGGUIInfoMessage
3002         GUIERR=PGGUIErrorMessage
3003         GUIINFO=PGGUIInfoMessage
3004     \textcolor{keywordflow}{elif} s\_gui\_flag == \textcolor{stringliteral}{"False"}:
3005         ACTIVATE\_GUI\_MESSAGING=\textcolor{keyword}{False}
3006     \textcolor{keywordflow}{else}:
3007         s\_msg=\textcolor{stringliteral}{"In pgdriveneestimator, def set\_messaging\_procedures, "} \(\backslash\)
3008                     + \textcolor{stringliteral}{"the program found an unknown value for the "} \(\backslash\)
3009                     + \textcolor{stringliteral}{"flag that indicates whether to use GUI "} \(\backslash\)
3010                     + \textcolor{stringliteral}{"messaging: "} + s\_gui\_flag + \textcolor{stringliteral}{"."}
3011         \textcolor{keywordflow}{raise} Exception( s\_msg )
3012     \textcolor{comment}{#end if we are using gui messaging}
3013     \textcolor{keywordflow}{return}
3014 \textcolor{comment}{#end set\_messaging\_procedures}
3015 
\end{DoxyCode}


\subsection{Variable Documentation}
\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS@{H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS}}
\index{H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS@{H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS}{HIDDEN_ARGS}}]{\setlength{\rightskip}{0pt plus 5cm}list negui.\+pgdriveneestimator.\+H\+I\+D\+D\+E\+N\+\_\+\+A\+R\+GS}\hypertarget{namespacenegui_1_1pgdriveneestimator_a45abba7d5f3257469b6cbfd7392eba09}{}\label{namespacenegui_1_1pgdriveneestimator_a45abba7d5f3257469b6cbfd7392eba09}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = [ \textcolor{stringliteral}{"file\_object\_main"}, \textcolor{stringliteral}{"file\_object\_secondary"}, 
2                     \textcolor{stringliteral}{"multiprocessing\_event"}, \textcolor{stringliteral}{"temporary\_directory"}  ]
\end{DoxyCode}


Definition at line 303 of file pgdriveneestimator.\+experimental.\+py.

\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL@{L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL}}
\index{L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL@{L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL}{LS_ARGS_HELP_OPTIONAL}}]{\setlength{\rightskip}{0pt plus 5cm}list negui.\+pgdriveneestimator.\+L\+S\+\_\+\+A\+R\+G\+S\+\_\+\+H\+E\+L\+P\+\_\+\+O\+P\+T\+I\+O\+N\+AL}\hypertarget{namespacenegui_1_1pgdriveneestimator_ad14f2792e801804d0cd77aa3f72631e4}{}\label{namespacenegui_1_1pgdriveneestimator_ad14f2792e801804d0cd77aa3f72631e4}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = [  \textcolor{stringliteral}{"total processes to use (single integer) Default is 1 process."},
2                 \textcolor{stringliteral}{"\(\backslash\)"no\_debug\(\backslash\)", \(\backslash\)"debug1\(\backslash\)", \(\backslash\)"debug2\(\backslash\)", \(\backslash\)"debug3\(\backslash\)", \(\backslash\)"testserial\(\backslash\)", \(\backslash\)"testmulti\(\backslash\)""} \(\backslash\)
3                 + \textcolor{stringliteral}{".  Indicates a run mode. The default is \(\backslash\)"no\_debug\(\backslash\)", which runs multiplexed "} \(\backslash\)
4                 + \textcolor{stringliteral}{"with standard output.  "} \(\backslash\)
5                 + \textcolor{stringliteral}{"Other modes except \(\backslash\)"testmulti\(\backslash\)", run non-parallelized, with increasing output.  "} \(\backslash\)
6                 + \textcolor{stringliteral}{"Debug 3, for example, adds to the output a table listing, for each indiv. "} \(\backslash\)
7                 + \textcolor{stringliteral}{"in each file, which replicate Ne estimates include the individual.  It also preserves "} 
      \(\backslash\)
8                 + \textcolor{stringliteral}{"the intermediate genepop and NeEstimator output files"},
9                 \textcolor{stringliteral}{"An Nb/Ne ratio, if supplied, will be used in a bias adjustment to the LDNE "} \(\backslash\)
10                 + \textcolor{stringliteral}{"estimates (after Waples, 2014).  Ignored if --donbbiasadjust is \(\backslash\)"False.\(\backslash\)"  "} \(\backslash\)
11                     + \textcolor{stringliteral}{"If you use 0.0 here, the program will look in the genepop file "} \(\backslash\)
12                     + \textcolor{stringliteral}{"header for an entry giving the Nb/Ne ratio as \(\backslash\)"nbne=<value>\(\backslash\)"."}, 
13                 \textcolor{stringliteral}{"True|False, whether to do the Nb bias adjustment. "} \(\backslash\)
14                 + \textcolor{stringliteral}{"If False, the Nb/Ne ratio parameter is ignored."} ]
\end{DoxyCode}


Definition at line 229 of file pgdriveneestimator.\+experimental.\+py.

\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED@{L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}}
\index{L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED@{L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}{LS_FLAGS_LONG_REQUIRED}}]{\setlength{\rightskip}{0pt plus 5cm}list negui.\+pgdriveneestimator.\+L\+S\+\_\+\+F\+L\+A\+G\+S\+\_\+\+L\+O\+N\+G\+\_\+\+R\+E\+Q\+U\+I\+R\+ED}\hypertarget{namespacenegui_1_1pgdriveneestimator_a8b1f5775794f7d387dedf1887a26c4ba}{}\label{namespacenegui_1_1pgdriveneestimator_a8b1f5775794f7d387dedf1887a26c4ba}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = [ \textcolor{stringliteral}{"--gpfiles"}, \textcolor{stringliteral}{"--scheme"}, \textcolor{stringliteral}{"--params"}, \textcolor{stringliteral}{"--minpopsize"}, \textcolor{stringliteral}{"--maxpopsize"}, \textcolor{stringliteral}{"--poprange"}, 
2                         \textcolor{stringliteral}{"--minallelefreq"}, \textcolor{stringliteral}{"--replicates"}, \textcolor{stringliteral}{"--locischeme"}, \textcolor{stringliteral}{"--locischemeparams"},  
3                         \textcolor{stringliteral}{"--mintotalloci"}, \textcolor{stringliteral}{"--maxtotalloci"}, \textcolor{stringliteral}{"--locirange"}, \textcolor{stringliteral}{"--locireplicates"} ]
\end{DoxyCode}


Definition at line 168 of file pgdriveneestimator.\+experimental.\+py.

\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS@{M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS}}
\index{M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS@{M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS}{MAIN_TABLE_RUN_INFO_COLS}}]{\setlength{\rightskip}{0pt plus 5cm}list negui.\+pgdriveneestimator.\+M\+A\+I\+N\+\_\+\+T\+A\+B\+L\+E\+\_\+\+R\+U\+N\+\_\+\+I\+N\+F\+O\+\_\+\+C\+O\+LS}\hypertarget{namespacenegui_1_1pgdriveneestimator_a7d21ac3784419638647683e99b085cc0}{}\label{namespacenegui_1_1pgdriveneestimator_a7d21ac3784419638647683e99b085cc0}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = [ \textcolor{stringliteral}{"original\_file"}, \textcolor{stringliteral}{"pop"}, \textcolor{stringliteral}{"census"},   \textcolor{stringliteral}{"indiv\_count"}, 
2                             \textcolor{stringliteral}{"sample\_value"}, \textcolor{stringliteral}{"replicate\_number"}, \textcolor{stringliteral}{"loci\_sample\_value"}, 
3                             \textcolor{stringliteral}{"loci\_replicate\_number"},\textcolor{stringliteral}{"min\_allele\_freq"} ]
\end{DoxyCode}


Definition at line 433 of file pgdriveneestimator.\+experimental.\+py.

\index{negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}!S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA@{S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA}}
\index{S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA@{S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA}!negui\+::pgdriveneestimator@{negui\+::pgdriveneestimator}}
\subsubsection[{\texorpdfstring{S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA}{SAMPLE_SCHEMES_NON_CRITERIA}}]{\setlength{\rightskip}{0pt plus 5cm}list negui.\+pgdriveneestimator.\+S\+A\+M\+P\+L\+E\+\_\+\+S\+C\+H\+E\+M\+E\+S\+\_\+\+N\+O\+N\+\_\+\+C\+R\+I\+T\+E\+R\+IA}\hypertarget{namespacenegui_1_1pgdriveneestimator_adb33149c6f4e36493056d012e2029270}{}\label{namespacenegui_1_1pgdriveneestimator_adb33149c6f4e36493056d012e2029270}
{\bfseries Initial value\+:}
\begin{DoxyCode}
1 = [ SAMPLE\_BY\_PERCENTAGES, 
2                                     SAMPLE\_BY\_REMOVAL ]
\end{DoxyCode}


Definition at line 374 of file pgdriveneestimator.\+experimental.\+py.

